---
title: "Normalisation"
bibliography:
  bibtex.bib
---

<!-- As HTML pages are built independently, large objects need to be
reimported from disk, and smaller ones quickly recomputed. -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
stopifnot(
  require(RColorBrewer),
  require(scran),
  require(ggplot2)
)
sce.filtered <- readRDS("rds/sce.filtered.rds")
col.time <- brewer.pal(9, "Set3")[c(2,7:8)]
col.infection <- brewer.pal(12, "Paired")[c(9,2,4)]
col.status <- brewer.pal(12, "Paired")[c(9,1,7)]
names(col.time) <- levels(sce.filtered$Time)[1:3]
names(col.infection) <- levels(sce.filtered$Infection)[1:3]
names(col.status) <- levels(sce.filtered$Status)
```

# Sum factors

Having identified in previous sections:

* cells with good quality sequencing data ([here](03_sample_QC.html)),
* features detected in at least one of the experimental groups
  ([here](04_feature_QC.html));

We may now normalise expression levels between
cells using an approach suitable for single cell data sets.

## Unsupervised clusters

Considering the multifactorial experimental design of this data set,
and the considerable transcriptional changes expected over time, and in
response to bacterial stimuli,
let us first identify clusters of transcriptionally similar cells;
those unsupervised clusters will subsequently be used to apply a
stratified normalisation procedure within each cluster, concluded by
a scaling of size factors for comparison between clusters,
as implemented in the [scran](http://bioconductor.org/packages/scran) package.

Importantly, let us define the *smallest cluster* size to the minimum number
of cells in any experimental group (*i.e.*, `r min(table(sce.filtered$Group))`
cells in group `r names(which.min(table(sce.filtered$Group)))`; see this
[earlier section](04_feature_QC.html#selectFeatures) for the count of cells
per group).

**Note:**
The default behaviour of the `quickCluster` function is to
only use *endogenous features* (`get.spikes=FALSE`),
as ERCC spike-in features should be present in equal amounts in all samples:

```{r quickCluster}
clusters <- quickCluster(  sce.filtered, min.size=min(table(sce.filtered$Group)))
sce.filtered$Cluster <- clusters
```

```{r, clusters_table, echo=FALSE, results='asis'}
clusters_table <- data.frame(table(clusters))
colnames(clusters_table) <- c("Cluster", "Cells")
pander::pandoc.table(clusters_table)
```

In this case, four clusters were identified.
Let us examine the distribution of experimental groups across the
identified clusters:

```{r longCluster, warning=FALSE}
longCluster <- reshape2::melt(
  data.frame(colData(sce.filtered)[,c("Cluster","Time","Infection","Status")]),
  measure.vars = c("Time","Infection","Status"),
  variable.name = "Factor", value.name = "Level"
)
longCluster$Level <- with(
  colData(sce.filtered),
  factor(longCluster$Level, rev(c(levels(Time), levels(Infection), levels(Status))))
)
```

```{r longCluster_plot, echo=FALSE, fig.height=6}
ggplot(longCluster) +
  geom_bar(aes(x = interaction(Factor, Level, sep = ": "), fill = Factor)) +
  facet_grid(Cluster ~ ., labeller = label_both) +
  theme_bw() +
  coord_flip() +
  labs(x = "") +
  theme(axis.text.y = element_text(hjust = 0))
```

As expected, the figure above demonstrates that the largest cluster
(cluster `r names(which.max(table(clusters)))`: `r max(table(clusters))` cells)
contains most `2h` cells and `4h:uninfected`
cells (which cannot be found in any other facet).
Clusters 2 and 3 contain uniquely `4h` and `6h` stimulated cells, respectively.
Finally, the smallest cluster (4) contains mostly `6h:uninfected`, with
a handful of other cells.

## Compute size factors

Let us apply the normalisation procedure, deconvolving size factors clusters:

```{r computeSumFactors}
sce.filtered <- computeSumFactors(
  sce.filtered, sizes = c(10, 15), clusters = sce.filtered$Cluster)
ggplot(data.frame(colData(sce.filtered, internal=TRUE))) +
  geom_histogram(
    aes(size_factor, fill=Time),
    binwidth = 0.1, colour="black", position = position_stack()) +
  scale_x_continuous(limits = c(0, 2)) +
  labs(x = "Size factor") +
  theme_bw()
```

## Relation to library size {.tabset}

In the absence of biological effect, size factors are expected to correlate
with library size [@67]. Let us visualise the two values in a scatter plot:

### Scatter: Status

```{r librarySize_Status, echo=FALSE}
ggplot(
  data.frame(
    colData(sce.filtered)[,c("total_counts","Time","Status")],
    sizeFactors = sizeFactors(sce.filtered)),
    aes(total_counts / 1E6, sizeFactors)
  ) +
  stat_smooth(
    aes(colour = Time, linetype = Status),
    method = "lm", se = TRUE, size = 0.5, alpha = 0.1
  ) +
  geom_point(aes(colour = Time, shape = Status)) +
  labs(x = "Library size (millions)", y = "Size factor") +
  theme_minimal()
```

In the figure above, the size factors are somewhat correlated with library size,
while significant scatter is visible around the trend;
augmenting the figure with experimental phenotype information emphasises
how the effect of library size on the size factor of individual cells is
subject to factors related to the experimental design of the experiment.

In particular, the figure emphasises how the effect of library size is
more important in `uninfected` cells, and at the earliest time point
(*i.e.*, `2h`), when the biological effect of infection may not have fully
affect the transcriptome of most cells yet.

In contrast, at later time points (*i.e.*, `4h` and `6h`), trends indicate a
markedly weaker effect of library size on size factor.

This suggest that other sources of systematic differences between cells
arose (*i.e.*, `Status`, `Infection`),
and that differential expression is to be expected
between experimental groups at those time points

### Scatter: Infection

```{r librarySize_Infection, echo=FALSE}
ggplot(
  data.frame(
    colData(sce.filtered)[,c("total_counts","Time","Infection")],
    sizeFactors = sizeFactors(sce.filtered)),
    aes(total_counts / 1E6, sizeFactors)
  ) +
  stat_smooth(
    aes(colour = Time, linetype = Infection),
    method = "lm", se = TRUE, size = 0.5, alpha = 0.1
  ) + geom_point(aes(colour = Time, shape = Infection)) +
  labs(x = "Library size (millions)", y = "Size factor") +
  theme_minimal()
```

## Relation to experimental groups

The same effect to experimental covariates on the distribution of size factors
within each experimental group may also be summarised in box plots as
shown below:

```{r sizeFactor_expFactors, echo=FALSE}
ggplot(
  data.frame(
    colData(sce.filtered)[,c("total_counts","Time","Infection","Status")],
    sizeFactors = sizeFactors(sce.filtered)),
    aes(interaction(Infection, Status, sep = " "), sizeFactors)
  ) +
  geom_boxplot(aes(fill = Infection, linetype = Status)) +
  coord_flip() +
  labs(y = "Size factor", x = "") +
  theme_bw() +
  theme(panel.grid.major.y = element_blank()) +
  facet_grid(Time ~ .) +
  scale_y_continuous(limits = c(0,2))
```

In a different perspective, the above figure emphasises again how
size factors are affected by experimental factors in this data set;
in particular at the later time points,
size factors for cells  stimulated with bacteria
display markedly distinct size factors, irrespective of library size.

Overall, size factors suggest that experimental factors drive more important
systematic differences between cells than library size at those time points,
having likely reshaped the transcriptional profile of stimulated cells.

# Spike factors {#setSpike}

To ensure normalization is performed correctly, let us compute a separate set
of size factors for the spike-in set. This assumes
that none of the spike-in transcripts are differentially expressed.

```{r computeSpikeFactors}
sce.filtered <- computeSpikeFactors(
  sce.filtered, type = "ERCC", general.use = FALSE
)
```

We may then examine the spike factors calculated:

```{r spikeFactors_ggplot, echo=FALSE}
ggplot(
  data.frame(
    colData(sce.filtered)[,c("total_counts","Time","Status")],
    size_factor_ERCC = sizeFactors(sce.filtered, type="ERCC")
  ),
  aes(total_counts / 1E6, size_factor_ERCC)
  ) +
  stat_smooth(method = "lm") +
  geom_point(aes(colour = Time, shape = Status)) +
  theme_minimal() +
  labs(x = "Library size (millions)", y = "Size factor (ERCC)")
```

In contrast to the earlier size factors calculated using endogenous features,
spike factors calculated using ERCC spike-in display a much weaker correlation
with library size, and no trend with experimental factors.

Furthermore, this reinforces the absence of extreme outliers after the removal
of cells with abnormal ERCC content in an [earlier section](#ERCCoutliers);
moreover, it suggests that counts assigned to individual ERCC spike-in features
are similar across all the cells retained after quality control.

# Apply size factors to normalise {#normalize}

The count data are used to compute normalized log-expression values for use in
downstream analyses.
Each value is defined as the log-ratio of each count to the size factor for the
corresponding cell, after adding a prior count of `1` to avoid undefined values
at zero counts.
Division by the size factor ensures that any cell-specific biases are removed.
Spike-in-specific size factors present are applied to normalize the spike-in
transcripts separately from the endogenous features:

```{r normalise}
sce.norm <- normalize(sce.filtered)
assayNames(sce.norm)
```

<!-- As HTML pages are built independently, objects need to be saved to disk
and imported as needed by the other R mardown files -->

```{r saveRDS, include=FALSE}
saveRDS(sce.norm, "rds/sce.norm.rds")
```

# References
