---
title: "QC"
author: Anna Aulicino & KÃ©vin Rue-Albrecht
date: "01/11/2016"
output: ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r check, include=FALSE}
library(scater)
library(S4Vectors)
```

```{r settings, include=FALSE}
if (interactive()){
  workdir <- "."
} else {
  workdir <- ".."
}
folder.rds <- file.path(workdir, "rds")
folder.qc <- file.path(workdir, "qc")
```

## Quantify expression

Prior to this document, quantitation data for all 384 samples
(including single cells, blanks and bulks) was obtained using _Kallisto_ 
as follows:

```
kallisto quant \
    --index=$kallistoIdx \
    --output-dir=$kallistoDir/$cell \
    $fastqFolder/${cell}_1.fastq.gz $fastqFolder/${cell}_2.fastq.gz
```

## Import preprocessed data

The resulting quantitation data was imported into the _R_ environment as a
`SCESet` using the `scater` `readKallistoResults` method,
and preprocessed as follows:

* Import quantitation data as a `SCESet` using `readKallistoResults()`
* Phenotype data was imported from a CSV file and attached to the `SCESet`.
* Transcript-level expression data was aggregated into gene-level
  expression estimates using `getBMFeatureAnnos()` and
  `summariseExprsAcrossFeatures()`
* Gene identifiers were made more intuitive by prepending gene symbol if
  available
* The resulting `SCESet` was serialised and backed up to file.

## Import preprocessed data

```{r preprocessed, echo = TRUE, message=FALSE}
library(scater)
sc_DC_gene <- readRDS(file.path(folder.rds, "sc_DC_gene.rds"))
dim(sc_DC_gene)
names(assayData(sc_DC_gene))
```

## Phenotype data

```{r expdata}
sc_DC_gene$Lane <- as.factor(sc_DC_gene$Lane)
sc_DC_gene$Identifier.number <- as.character(sc_DC_gene$Identifier.number)
summary(pData(sc_DC_gene)[,1:7])
```

## Identify control features and samples

### Features

> - ERCC spike-ins
> - Mitochondrial genes

### Samples

>- Blanks
>- Bulks

## Mitochondrial genes

`TODO:` use `biomaRt`

## Identify control features and samples

```{r, echo=TRUE}
filter.features.ctrl <- FilterRules(list(
        ERCC = function(x) grepl("ERCC-[[:digit:]]", featureNames(x))
))
summary(evalSeparately(filter.features.ctrl, sc_DC_gene))
filter.samples.ctrl <- FilterRules(list(
        BULK = function(x) x$Status == "BULK",
        Blank = function(x) x$Status == "Blank"
))
summary(evalSeparately(filter.samples.ctrl, pData(sc_DC_gene)))
```

## Calculate QC metrics




## R Markdown

This is an R Markdown presentation. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document.

## Slide with Bullets

- Bullet 1
- Bullet 2
- Bullet 3

## Slide with R Output

```{r cars, echo = TRUE}
summary(cars)
```

## Slide with Plot

```{r pressure}
plot(pressure)
```

