---
title: "Import of mini-bulk read counts and metadata"
author: "Kevin Rue-Albrecht"
date: "06/01/2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

<style>
body {
text-align: justify}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
stopifnot(
  require(edgeR),
  require(rtracklayer),
  require(SingleCellExperiment),
  requireNamespace("DT"),
  requireNamespace("xlsx")
)
```

# Phenotype information {#phenotypes}

Let us import experimental information:

```{r importPheno}
pheno <- xlsx::read.xlsx("expdata/Mini_bulk_metadata.xlsx", sheetName = "pheno")
colnames(pheno)
```

To facilitate later use, let us retype some fields, reorder some factor levels,
and add a set of additional fields:

* `Sample`: contains a shortened unique identifier derived from the count
  file name for each sample,
* `Treatment`: contains a factor that combines `Infection` and `Status`
  phenotypes
* `Group`: contains a factor that combines all experimental phenotypes
  (*i.e.*, `Time`, `Infection` and `Status`)

```{r editPheno}
pheno$File <- as.character(pheno$File)
pheno$Infection <-
  factor(pheno$Infection, c("Mock","STM-LT2","STM-D23580"))
pheno$Status <-
  factor(pheno$Status, c("Uninfected","Exposed","Infected"))
pheno$Lane <- as.factor(pheno$Lane)
pheno$Sample <- gsub("WTCHG_[[:digit:]]+_", "Sample_", pheno$File) # character
treatmentLevels <- c(
  "Mock_Uninfected",
  "STM-LT2_Exposed","STM-LT2_Infected",
  "STM-D23580_Exposed","STM-D23580_Infected"
  )
pheno$Treatment <- with(pheno, factor(
  droplevels(interaction(Infection, Status, sep = "_")), treatmentLevels
))
# factor
pheno$Group <- with(pheno, factor(
  droplevels(interaction(pheno$Time, pheno$Treatment, sep = "_"))
))
```

# QC metrics {#QC}

Importantly, [MultiQC](http://multiqc.info) was used to collate metrics
for each sample at each step of preprocessing---from raw reads to assigned
read counts---; those QC metrics provide key information to identify
technical issues and sample outliers.

Let us import the QC metrics produced by [MultiQC](http://multiqc.info):

```{r importMultiQC}
multiqc <- read.delim("MultiQC/mini_bulks_data/multiqc_general_stats.txt")
colnames(multiqc)
```

Let us shorten the sample identifier used by [MultiQC](http://multiqc.info)
to a minimal unique identifier
(as done [above](#phenotypes) for experimental phenotype information):

```{r gsub_Sample}
multiqc$Sample <- gsub("WTCHG_[[:digit:]]+_", "Sample_", multiqc$Sample)
rownames(multiqc) <- multiqc$Sample
colnames(multiqc)
```

Let us append those QC metrics to the phenotype information,
that will [later](#SummarizedExperiment) be attached to the read counts in a `SummarizedExperiment`
([SummarizedExperiment](http://bioconductor.org/packages/SummarizedExperiment)):

```{r addPhenoMultiQC}
pheno <- merge(pheno, multiqc, by = "Sample", sort = FALSE)
```

# Counts {#Counts}

To import count data that is stored in a separate file for each sample,
let us use the [edgeR](http://bioconductor.org/packages/edgeR)
`readDGE` method and the phenotype information prepared [above](#phenotypes):

```{r readDGE}
RG <- with(
  pheno,
  readDGE(File, "expdata/counts", c(1, 3), Group, Sample)
)
dim(RG)
```

# Gene annotations {.tabset}

To ensure coherence between gene annotations and gene expression data,
let us also import the records for genes and ERCC spike-ins molecules
from the same GTF file that was supplied to *featureCounts*,
as `GRanges` of the
[GenomicRanges](http://bioconductor.org/packages/GenomicRanges) package;
in the next section, those annotations will be attached to the corresponding
read counts in a `SummarizedExperiment` object
([SummarizedExperiment](http://bioconductor.org/packages/SummarizedExperiment)).

First we import all the records from the composite GFF file that includes
both endogenous gene features and ERCC spike-in molecules:

```{r import.gff}
gtfData <- import.gff("../expdata/genes_with_ERCC.gtf")
colnames(as.data.frame(gtfData))
```

Finally, let us combine the two types of feature annotations,
and reorder them to ensure that they match the order of features imported
[above](#Counts):

```{r gtfData}
names(gtfData) <- gtfData$gene_id
stopifnot(all(rownames(RG) %in% names(gtfData)))
gtfData <- gtfData[match(rownames(RG), names(gtfData))]
stopifnot(all(rownames(RG) == names(gtfData)))
```

In addition, let us also obtain the identifier of ERCC spike-in features,
to annotate them as spike-in in the following section:

```{r ERCC_ids}
ERCCs <- grep("^ERCC-[[:digit:]]+$", rownames(RG), value = TRUE)
```

# SummarizedExperiment {#SummarizedExperiment .tabset}

Let us assemble the count data, phenotype information, and feature information
prepared above into `SingleCellExperiment` objects
([SingleCellExperiment](http://bioconductor.org/packages/SingleCellExperiment)):

* one object that contains all samples together
* one for the blank samples
* one for the bulk samples
* one for the single cells

**Note:**
This distinction will be important later, to compute and compared QC metrics
within each sets of samples.

First, let us set the `rownames` of the phenotypes information `data.frame` to
the unique sample identifier for consistency with the sample identifier
associated with the count data imported [above](#Counts):

```{r renamePheno}
rownames(pheno) <- pheno$Sample
```

We may then create the four data sets, declaring in each of them the set
of ERCC spike-in features:

## All samples

```{r sce.all}
pd <- DataFrame(pheno)
sce <- SingleCellExperiment(
  assays=list(counts=RG$counts), colData=pd, rowRanges=gtfData)
isSpike(sce, "ERCC") <- ERCCs
dim(sce)
```

<!-- As HTML pages are built independently, objects need to be saved to disk
and imported as needed by the following R mardown files -->

```{r saveRDS, include=FALSE}
saveRDS(sce, "rds/sce.minibulks.rds")
```
