---
title: "Comparison of differential expression with mini-bulks"
author: "Kevin Rue-Albrecht"
date: "06/01/2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
stopifnot(
  require(dplyr),
  require(RColorBrewer),
  require(ggplot2),
  require(ggrepel)
)
sce.endo <- readRDS("rds/sce.endo_clusters.rds")
scde.res <- readRDS("rds/scde_group.rds")
deseq.res <- readRDS("rds/deseq.res.rds") # NOTE: uses "DESeq2-safe" names
outdir <- "006_out"; dir.create(outdir)
col.time <- brewer.pal(9, "Set3")[c(2,7:8)]
col.infection <- brewer.pal(12, "Paired")[c(9,4,2)]
col.status <- brewer.pal(12, "Paired")[c(9,1,7)]
names(col.time) <- levels(sce.endo$Time)[1:3]
names(col.infection) <- levels(sce.endo$Infection)
names(col.status) <- levels(sce.endo$Status)
col.treatment <- RColorBrewer::brewer.pal(9, "Paired")[c(9,3,1,4,2)]
names(col.treatment) <- levels(sce.endo$Treatment)
#
convert.z.score <- function(x, one.sided = NULL) {
  z <- x$Z
  if(is.null(one.sided)) {
    pval = pnorm(-abs(z));
    pval = 2 * pval
  } else if(one.sided=="-") {
    pval = pnorm(z);
  } else {
    pval = pnorm(-z);
  }
    x <- cbind(
      x,
      p.value = pval
  )
  return(x);
}
addGeneName <- function(x){
  x <- cbind(
    gene_name = with(rowData(sce.endo), gene_name[match(rownames(x), gene_id)]),
    x
  )
  return(x)
}
```

# Define filter choices

```{r filters}
max.padj <- 0.01
min.FC <- 0
topNgenes <- 20
```

# Set shared axis scales

```{r fc.scale}
fc.scde <- max(do.call("rbind", lapply(scde.res, function(x){abs(range(x$mle))})))
fc.deseq <- max(do.call("rbind", lapply(deseq.res, function(x){abs(range(x$log2FoldChange))})))
fc.scale <- c(-1,1)*max(c(fc.scde, fc.deseq))
```

# Comparisons relative to uninfected controls

## Combine data from all contrasts

```{r tablesUninfected}
contrastNames <- grep("Uninfected", names(scde.res), value = TRUE)
tablesUninfected <- lapply(
  contrastNames, function(contrastName){
    message(contrastName)
    stopifnot(contrastName %in% names(deseq.res))
    tableSCDE <- convert.z.score(scde.res[[contrastName]])
    tableDESeq <- deseq.res[[contrastName]]
    intersectTests <- intersect(rownames(tableSCDE), rownames(tableDESeq))
    tableIntersect <- cbind(
      tableSCDE[intersectTests, c("mle","p.value")],
      tableDESeq[intersectTests, c("log2FoldChange","padj")]
    )
    tableIntersect <- tableIntersect[with(tableIntersect, order(mle * log2FoldChange, decreasing = TRUE)),]
    tableIntersect$Contrast <- gsub("-([[:digit:]])", "\n\\1", contrastName)
    tableIntersect <- addGeneName(tableIntersect)
    tableIntersect
  }
)
names(tablesUninfected) <- contrastNames
```

```{r tablesMerged}
tablesMerged <- DataFrame(do.call("rbind", tablesUninfected), row.names = NULL)
DE.levels <- c("None", "SCDE", "DESeq2", "Both")
tablesMerged$DE <- with(
  tablesMerged, DE.levels[1 + (p.value < max.padj) + 2*(padj < max.padj)]
)
# tablesMerged <- tablesMerged[with(tablesMerged, order(mle * log2FoldChange, decreasing = TRUE)),]
```

## Label top genes DE both experiments

```{r ggplot_DE_both}
ggplot(subset(as.data.frame(tablesMerged), DE != "None"), aes(mle, log2FoldChange)) +
  facet_wrap(~Contrast, nrow = 3, ncol = 4) +
  geom_point(
    aes(colour = DE),
    alpha = 0.3, size = 0.4
  ) +
  scale_x_continuous(limits = fc.scale) +
  scale_y_continuous(limits = fc.scale) +
  theme_bw() +
  geom_text_repel(
    aes(label = gene_name, color = DE),
    head(
      dplyr::arrange(subset(as.data.frame(tablesMerged), DE == "Both"), desc(abs(mle))),
      topNgenes
    ),
    min.segment.length = unit(0, "lines"),
    size = 2, alpha = 0.9
  )
```

## Label top genes DE in single-cell only

```{r ggplot_DE_scde}
ggplot(subset(as.data.frame(tablesMerged), DE != "None"), aes(mle, log2FoldChange)) +
  facet_wrap(~Contrast, nrow = 3, ncol = 4) +
  geom_point(
    aes(colour = DE),
    alpha = 0.3, size = 0.4
  ) +
  scale_x_continuous(limits = fc.scale) +
  scale_y_continuous(limits = fc.scale) +
  theme_bw() +
  geom_text_repel(
    aes(label = gene_name, color = DE),
    head(
      dplyr::arrange(subset(as.data.frame(tablesMerged), DE == "SCDE"), desc(abs(mle))),
      topNgenes
    ),
    min.segment.length = unit(0, "lines"),
    size = 3, alpha = 0.9
  )
```

## Label MARCH1

```{r ggplot_MARCH1}
ggplot(subset(as.data.frame(tablesMerged), DE != "None"), aes(mle, log2FoldChange)) +
  facet_wrap(~Contrast, nrow = 3, ncol = 4) +
  geom_point(
    aes(colour = DE),
    alpha = 0.3, size = 0.4
  ) +
  scale_x_continuous(limits = fc.scale) +
  scale_y_continuous(limits = fc.scale) +
  theme_bw() +
  geom_text_repel(
    aes(label = gene_name, color = DE),
    subset(as.data.frame(tablesMerged), gene_name == "MARCH1"),
    min.segment.length = unit(0, "lines"),
    size = 4, alpha = 0.8
  )
```

# Volcano plots ----

## 4h_STM-D23580_Infected-4h_STM-D23580_Exposed

```{r volcano_4h_D23580_top}
contrastName <- "4h_STM-D23580_Infected-4h_STM-D23580_Exposed"
tmpTable <- addGeneName(deseq.res[[contrastName]])
tmpTable <- tmpTable[order(tmpTable$pvalue),]
ggplot(tmpTable, aes(log2FoldChange, -log10(pvalue))) +
  geom_point(
    aes(colour = padj < 0.01), alpha = 0.5
  ) +
  geom_text_repel(
    aes(label = gene_name), min.segment.length = unit(0, "mm"), size = 3,
    head(tmpTable, 40)
  ) +
  labs(title = contrastName) +
  theme_bw()
```

```{r volcano_4h_D23580_curated}
tmpTable <- addGeneName(deseq.res[[contrastName]])
tmpTable <- tmpTable[order(tmpTable$pvalue),]
gg <- ggplot(tmpTable, aes(log2FoldChange, -log10(pvalue))) +
  geom_point(
    aes(colour = padj < 0.01), alpha = 0.3, size = 0.5
  ) +
  geom_text_repel(
    aes(label = gene_name), min.segment.length = unit(0, "mm"),
    subset(tmpTable, gene_name %in% c(
      "CTSD","LGMN","CCL7","CCL26","MS4A4A","LAMP1","APOE",
      "IL1B","PLAT","CD1A","CD1C","TIFAB","CCL1","TRAF4","CD83"
    ))
  ) +
  labs(title = contrastName) +
  theme_bw()
ggsave(
  file.path(outdir, sprintf("volcano_%s.pdf", contrastName)), gg,
  width = 6, height = 4
)
```

## 4h_STM-LT2_Infected-4h_STM-LT2_Exposed

```{r volcano_4h_LT2_top}
contrastName <- "4h_STM-LT2_Infected-4h_STM-LT2_Exposed"
tmpTable <- addGeneName(deseq.res[[contrastName]])
tmpTable <- tmpTable[order(tmpTable$pvalue),]
ggplot(tmpTable, aes(log2FoldChange, -log10(pvalue))) +
  geom_point(
    aes(colour = padj < 0.01), alpha = 0.5
  ) +
  geom_text_repel(
    aes(label = gene_name), min.segment.length = unit(0, "mm"), size = 3,
    head(tmpTable, 40)
  ) +
  labs(title = contrastName) +
  theme_bw()
```

```{r volcano_4h_LT2_curated}
tmpTable <- addGeneName(deseq.res[[contrastName]])
tmpTable <- tmpTable[order(tmpTable$pvalue),]
gg <- ggplot(tmpTable, aes(log2FoldChange, -log10(pvalue))) +
  geom_point(
    aes(colour = padj < 0.01), alpha = 0.3, size = 0.5
  ) +
  geom_text_repel(
    aes(label = gene_name), min.segment.length = unit(0, "mm"),
    subset(tmpTable, gene_name %in% c(
      "CTSD","CCL26","CCL7","LGMN","MS4A4A","LAMP1",
      "CCL1","CD1C","CD1A","PLAT","CLEC7A","REL","IL23A"
    ))
  ) +
  labs(title = contrastName) +
  theme_bw()
ggsave(
  file.path(outdir, sprintf("volcano_%s.pdf", contrastName)), gg,
  width = 6, height = 4
)
```

