---
title: "Preprocessing to RNA-Seq read counts"
editor_options: 
  chunk_output_type: console
---

# Overview {#Overview}

Prior to this vignette, paired-end reads (**2x 75 bp**) were processed as
follows:

* Quality control using
  [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc)
  (version **0.11.4**),
* Alignment to reference genome using
  [HISAT2](https://ccb.jhu.edu/software/hisat2) (version **2.0.3b**),
* Assigned to annotated features using
  [featureCounts](http://bioinf.wehi.edu.au/featureCounts)
  (*subread* version **1.5.0-p2**).

**Note:**
Adapter and quality trimming using
[Trim Galore!](http://www.bioinformatics.babraham.ac.uk/projects/trim_galore)
(version **0.4.1**; using [cutadapt](https://pypi.python.org/pypi/cutadapt)
version **1.10**) was tested;
however, the high quality of the raw data, and the marginal improvement of
alignment rate (< 1%) did not justify additional processing time and disk space.

# Composite genome (*Ensembl*)

Composite genome Fasta and GTF files were formed by concatenating the
primary assembly of the *Homo sapiens* genome build **GRCh38**
(excluding haplotypes and patches) and the ERCC RNA Spike-In Mix:

```
cat $genomeFasta $erccFasta > $compositeFasta
cat $genomeGTF $erccGTF > $compositeGTF
```

The Fasta and GTF files for the human genome were obtained from the
*Ensembl FTP site* [release 86](ftp://ftp.ensembl.org/pub//release-86).
Namely, the downloaded files are:

* *Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz*
* *Homo_sapiens.GRCh38.86.chr.gtf.gz*

The ERCC Fasta and GTF files were obtained from
[ThermoFisher](https://www.thermofisher.com/order/catalog/product/4456739).

# Genome index (*HISAT2*)

The composite genome was indexed as follows:

```
hisat2-build $compositeFasta $hisatIdx
```

# Read quality control (*FastQC*)

Raw sequenced reads were quality-controlled using
[FastQC](#http://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
as follows:

```
fastqc \
  --outdir $fastqcFolder \
  --nogroup \
  $fastqFile
```

An example *FastQC* report is embedded below, and
linked [here](fastqc/WTCHG_305265_201231_2_fastqc.html) for full-screen layout.

<iframe height="500" width="100%"
src="fastqc/WTCHG_305265_201231_2_fastqc.html"></iframe>

# Alignment (*HISAT2*)

Raw (*i.e.*, not trimmed; see [above](#Overview)) read pairs were aligned as
follows:

```
hisat2 \
    -x ${hisatIdx} \
    -1 ${fastqFolder}/${cell}_1.fastq.gz \
    -2 ${fastqFolder}/${cell}_2.fastq.gz \
    -S $hisatDir/${cell}.sam \
    1> $logDir/${cell}.o 2> $logDir/${cell}.e
```

The resulting SAM files were compressed to BAM format as follows:

```
samtools view -hb -o $hisatDir/${cell}.bam $hisatDir/${cell}.sam
```

Finally, SAM files were then removed as follows:

```
rm -f $hisatDir/${cell}.sam
```

# Read assignment (*featureCounts*) {#featureCounts}

Uniquely aligned read pairs were assigned to annotated genomic feaures and
ERCC RNA spike-in features, and counted in an unstranded manner as follows:

```
featureCounts \
    --primary \
    --ignoreDup \
    -p \
    -a $compositeGTF \
    -o $countDir/$cell \
    $bamDir/${cell}.bam \
    1> $logDir/$cell.out 2> $logDir/$cell.err
```

# *MultiQC*

Reports of the various preprocessing steps above were collated using [MultiQC](http://multiqc.info). The overall report in embedded below, and
linked [here](MultiQC/final_pipeline.html) for full-screen layout.

<iframe height="500" width="100%" src="MultiQC/final_pipeline.html"></iframe>
