---
title: "Feature selection"
bibliography:
  bibtex.bib
---

<!-- As HTML pages are built independently, large objects need to be
reimported from disk, and smaller ones quickly recomputed. -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
stopifnot(
  require(SingleCellExperiment),
  require(RColorBrewer),
  require(ggplot2),
  requireNamespace("edgeR"),
  require(scater),
  requireNamespace("limma")
)
sce.pass <- readRDS("rds/sce.pass.rds")
sce.bulk <- readRDS("rds/sce.bulk.SCE.rds")
mixC <- readRDS("rds/mixC.rds")
ERCCs <- grep("^ERCC-[[:digit:]]+$", rownames(sce.pass), value = TRUE)
MTs <- names(rowRanges(sce.pass)[seqnames(sce.pass) == "MT",])
col.time <- brewer.pal(9, "Set3")[c(2,7:8)]
col.infection <- brewer.pal(12, "Paired")[c(9,4,2)]
col.status <- brewer.pal(12, "Paired")[c(9,1,7)]
names(col.time) <- levels(sce.pass$Time)[1:3]
names(col.infection) <- levels(sce.pass$Infection)[1:3]
names(col.status) <- levels(sce.pass$Status)
dir.create("03_out", showWarnings = FALSE)
```

# Preliminary notes

This section uses objects defined in previous sections of the website, namely:

* `mixC`: concentrations of ERCC spike-in molecules in mix
  ([ERCC content](01_ERCC.html#import)),
* `ERCCs`: names of the ERCC spike-in features in the count matrix
  ([Sample QC#blanks](02_QC_sample.html#blanks)),
* `MTs`: names of the mitochondrial gene features in the count matrix
  ([Sample QC#calculateQCMetrics](02_QC_sample.html#calculateQCMetrics)).

# Correlation of bulks and single cell average

## Preprocessing

Let us first calculate the log~2~-transformed count per million (CPM) after
adding a default $0.25$ count to all genes, to avoid computing the logarithm
of $0$:

```{r bulkSCdata_init}
logCPM.cells <- edgeR::cpm(counts(sce.pass), log=TRUE, prior.count=0.25)
logCPM.bulk <- edgeR::cpm(counts(sce.bulk), log=TRUE, prior.count=0.25)
```

Let us progressively collect data for each experimental group:

```{r bulkSCdata_fill, message=FALSE}
bulkSCdata <- data.frame()
for (x_group in levels(sce.pass$Group)){
  message(x_group)
  time <- gsub("^(.*h)_.*", "\\1", x_group)
  infection <- gsub(".*_(.*)_.*", "\\1", x_group)
  status <- gsub(".*_.*_(.*)", "\\1", x_group)
  treatment <- gsub(".*_(.*_.*)", "\\1", x_group)
  cellIndex <- (sce.pass$Group == x_group)
  featureIndex <- (
    rowSums(counts(sce.pass)) > 0 |
      rowSums(counts(sce.bulk)) > 0)
  cellAvg <- rowMeans(logCPM.cells[featureIndex, cellIndex])
  bulkExpr <- logCPM.bulk[
    featureIndex,
    which(sce.bulk$Time == time & sce.bulk$Infection == infection)
  ]
  geneNames <- names(cellAvg)
  bulkSCdata <- rbind(
    bulkSCdata,
    data.frame(
      Cells = cellAvg,
      Bulk = bulkExpr[geneNames], # ensure order identical to single cells
      Gene = geneNames,
      Group = x_group, Treatment = treatment,
      Time = time, Infection = infection, Status = status
    )
  )
}
```

Stimulated bulk samples are expected to be a mixture of
`infected` (minority; < 5%) and `exposed` cells (vast majority; > 95%).
We may therefore compare the average expression
in each subset of cells against the bulk average, both in log~2~(CPM),
to identify potential differences.
For reference, let us also indicate the identity line ($y=x$) in grey:

```{r bulkSCdata_plot, fig.height=8}
bulkSCdata$ERCC <- grepl("^ERCC-", bulkSCdata$Gene)
ggplot(bulkSCdata) +
  geom_hex(aes(Bulk, Cells), subset(bulkSCdata, !ERCC)) +
  geom_point(aes(Bulk, Cells), subset(bulkSCdata, ERCC),size=0.1,colour="red")+
  geom_abline(slope = 1, intercept = 0, colour = "grey") +
  coord_fixed(ratio = 1) + labs(y="Cell average") + guides(fill = FALSE) +
  facet_grid(Infection + Status ~ Time) +
  theme_minimal() + theme(
    axis.text.x = element_text(size = rel(0.5)),
    axis.text.y = element_text(size = rel(0.5))
  ) + guides(fill = "none")
```

```{r bulkSCdata_pdf, include=FALSE}
ggsave(file.path("03_out/correlation_cells_bulk.pdf"), width = 4, height = 8)
```

The figure above displays generally higher detection levels in bulk samples
for a large proportion of genes;
this demonstrates the higher frequency of dropout
events in single cell samples, and the average measurement of bulk
that rescues detection of rarer transcripts expressed
stochastically at low levels or in few cells.

# Overview of feature detection levels

Having selected single cells that pass all quality control criteria in the
previous sections, let us identify features that are robustly detected across
single cells in the data set.
Indeed, features detected in a small count of cells, or expressed at very
low levels carry limited information for differential expression analyses;
moreover, those features negatively affect the performance of normalisation
algorithms, due to the large number of dropout (*i.e.*, zero count) events
[@67; @Brennecke_2013].

## Counts per million (CPM)

First, let us initalise the matrix of CPM:

```{r cpm}
SingleCellExperiment::cpm(sce.pass) <- edgeR::cpm(counts(sce.pass))
range(colSums(cpm(sce.pass)))
```

**Note:**
The `SingleCellExperiment::cpm` method is an *accessor* to the corresponding slot of
`SingleCellExperiment` objects, while the `edgeR::cpm` method *calculates* normalised
counts per million (CPM) for given a matrix of read counts.
Furthermore, in the absence of _size factors_, CPM are computed using the
total count of read (*i.e.*, library size) for each sample,
without further normalisation of library sizes between samples.

## Detection frequency and mean expression level

For this purpose, let us first calculate updated QC metrics--in the subset
of cells that *passed* quality control in the
[previous section](#03_sample_QC.html)--using once more the
[scater](http://bioconductor.org/packages/scater) `calculateQCMetrics` method:

```{r pass_calculateQCMetrics}
sce.pass <- calculateQCMetrics(
  sce.pass, feature_controls = list(ERCC = ERCCs, MT = MTs)
)
```

We may then examine for each feature the proportion of cells in which it is
detected (non-zero expression level) against its average expression level
across all cells:

```{r plotExprsFreqVsMean}
plotExprsFreqVsMean(
  sce.pass,
  feature_controls = rowData(sce.pass)[,"is_feature_control_ERCC"]
)
```

## Spike-in expectations

Let us compare the expected number of spike-in molecules introduced in each
sample
(calculated in this [earlier section](00b_moleculesERCC.html)) against the
average count detected across all single cells, while indicating the identity
relationship ($y=x$) as a solid black line:

```{r ERCCcorrelation, warning=FALSE, fig.height=6}
mixC$avgCount <- rowMeans(counts(sce.pass)[mixC$ID,])
ggplot(mixC) +
  geom_point(aes(ERCCmolecules, avgCount)) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", colour = "red") +
  labs(
    x = "Estimated ERCC molecules injected",
    y = "Average observed fragment count in cells passed QC") +
  scale_x_log10() + scale_y_log10() + coord_fixed(ratio = 1) + theme_minimal()
```

```{r ERCCcorrelation_pdf, include=FALSE}
ggsave("03_out/correlation_ERCC_expected.pdf", width = 5, height = 5)
```


# Definition of detected features

## Count of genes by mean expression level {.tabset}

Let us examine the distribution of log-mean CPM across all genes,
while indicating an average value of **1** CPM by a vertical blue dashed line:

```{r detectVsMean_hist}
ave.cpm <- rowMeans(cpm(sce.pass))
hist(
  log10(ave.cpm), breaks = 100,
  main = "", xlab = expression(Log[10]~"average CPM", col="grey80")
)
abline(v = log10(1), col = "blue", lwd = 2, lty = 2)
```

## Count of cells with detectable expression

We may compare the above figure with the a smoother representation of the
count of cells in which each feature is detected against the average
log~10~-transformed count of reads assigned to that same feature, while
marking the same cut-off threshold of an average **1** count across all cells:

```{r smoothScatter}
numcells <- nexprs(sce.pass, byrow = TRUE)
is.ercc <- rowData(sce.pass)[,"is_feature_control_ERCC"]
smoothScatter(
  log10(ave.cpm), numcells, xlab=expression(Log[10]~"average CPM"),
  ylab="Number of expressing cells"
)
points(
  log10(ave.cpm[is.ercc]), numcells[is.ercc],
  col = "red", pch = 16, cex = 0.5
)
abline(v = log10(1), col = "blue", lwd = 2, lty = 2)
```

## Distribution of expression levels within groups {.tabset}

### Density

Let us first calculate the average CPM for each gene within each experimental
group:

```{r avgCPMbyGroup}
avgCPMbyGroup <- apply(
  cpm(sce.pass)[!isSpike(sce.pass, "ERCC"),],
  1,
  function(x){tapply(x, sce.pass$Group, mean)})
avgCPMbyGroup <- reshape2::melt(
  avgCPMbyGroup,
  varnames = c("Group", "Feature"),
  value.name = "AvgCPM"
)
avgCPMbyGroup[,c("Time","Infection","Status")] <-
  limma::strsplit2(avgCPMbyGroup$Group, "_")
```

The following figure displays the resulting distribution of
average log~10~(CPM) within each experimental group of single cells
(after removal of zero-count genes):

```{r histogram_avgCPM, echo=FALSE, fig.height=8}
detectedCPM <- data.frame(
  Group = levels(avgCPMbyGroup$Group),
  Zero = with(avgCPMbyGroup, tapply(AvgCPM, Group, function(x){sum(x==0)})),
  LessOne = with(avgCPMbyGroup, tapply(AvgCPM, Group, function(x){
    sum(x<1 & x!= 0)})),
  MoreOne = with(avgCPMbyGroup, tapply(AvgCPM, Group, function(x){sum(x>1)}))
)
detectedCPM <- merge(
  detectedCPM,
  as.data.frame(unique(colData(sce.pass)[,c("Group","Time","Infection","Status")])))
avgCPMbyGroup$Infection <- factor(avgCPMbyGroup$Infection, levels(sce.pass$Infection))
avgCPMbyGroup$Status <- factor(avgCPMbyGroup$Status, levels(sce.pass$Status))
ggplot() +
  geom_histogram(
    aes(log10(AvgCPM), fill = Infection, alpha = Status), avgCPMbyGroup,
    binwidth = 0.1, alpha = 0.7) +
  geom_vline(xintercept = 0, colour = "red", linetype = "dashed") +
  geom_label(aes(label = Zero), detectedCPM, x=0, y=75, size=2, alpha=0.5) +
  geom_label(aes(label = LessOne), detectedCPM, x=-1, y=700, size=2, alpha=0.5) +
  geom_label(aes(label = MoreOne), detectedCPM, x=4, y=700, size=2, alpha=0.5) +
  scale_fill_manual(values = col.infection) +
  facet_grid(Infection + Status ~ Time) + labs(y = "Features") +
  labs(x = "log10(Average CPM)") + theme_bw() +
  theme(legend.position = "bottom")
```

```{r histogram_avgCPM_pdf, include=FALSE}
ggsave("03_out/countFeaturesByAvgLogCPM.pdf", height = 9, width = 6)
```

### Violin

Alternatively, we may also display the distribution of average log~10~(CPM)
within each experimental group of single cells as a violin plot.

Let us first identify the count of detected features in each experimental
group of cells:

```{r detectedCounts}
detectedCounts <- data.frame(
  Group = levels(avgCPMbyGroup$Group), y = 1E-2,
  Count = with(avgCPMbyGroup, tapply(AvgCPM, Group, function(x){sum(x>0)}))
)
detectedCounts$Time <- factor(
  gsub("(.*)_.*_.*", "\\1", detectedCounts$Group), levels(sce.pass$Time)
)
detectedCounts$Infection <- factor(
  gsub(".*_(.*)_.*", "\\1", detectedCounts$Group), levels(sce.pass$Infection)
)
detectedCounts$Status <- factor(
  gsub(".*_.*_(.*)", "\\1", detectedCounts$Group), levels(sce.pass$Status)
)
```

The following figure displays the distribution of average
log~10~(CPM) within each experimental group of single cells, while
indicating for each of them the count of non-zero values in each
experimental groups:

```{r violin_avgCPM, echo=FALSE, warning=FALSE}
ggplot(avgCPMbyGroup, aes(Time, AvgCPM)) +
  geom_violin(
    aes(fill = Infection, alpha = Status),
    draw_quantiles = seq(0.25, 0.75, 0.25)
    ) +
  facet_grid(~ Infection + Status) +
  geom_label(aes(Time, y, label = Count), data = detectedCounts, size = 2) +
  scale_y_log10(limits = c(1E-2, 1E5)) +
  scale_fill_manual(values = col.infection) +
  scale_alpha_discrete(
    breaks = levels(sce.pass$Status),
    range = seq(0.2, 0.8, 0.3)) +
  labs(y = "Average CPM") + guides(alpha = "none") +
  theme_bw() + theme(panel.grid.major.x = element_blank())
```

# Selection of detected features {#selectFeatures}

In the context of the present multifactorial experimental design,
it is expected that many genes display detectable expression levels in
only a subset of experimental conditions (*e.g.*, cells stimulated with
bacteria after sufficient time to allow intracellular signalling).

Consequently, let us define informative endogenous features as those
that display an average expression level above an expression threshold
of **10** counts per million (CPM) in at least **25%** of
at least **1** experimental group of cells
(`r min(table(sce.pass$Group))`-`r max(table(sce.pass$Group))`
cells per group, after the earlier [exclusion of outliers](#03_sample_QC.html)).

In contrast, let us define informative spike-in features
as those with an average count across all single cells greater or equal
to a threhsold of **1** count.
Note that the detailed count of cells per group was reported
[here](03_sample_QC.html#cellSelection).

First, let us obtain---for each feature---the proportion of cells within each
experimental group of cells that display an expression level above $20$ CPM:

```{r propOver10byGroup}
propOver10byGroup <- apply(
  cpm(sce.pass)[!isSpike(sce.pass, "ERCC"),],
  1,
  function(x){
  tapply(x, sce.pass$Group, function(x){sum(x > 10) / length(x)})
})
propOver10byGroup <- reshape2::melt(
  propOver10byGroup, varnames = c("Group", "Feature"), value.name = "Proportion")
```

Let us now define whether each feature is detected in at least **25%**
of each group:

```{r PropOver25pct}
propOver10byGroup$PropOver25pct <- (propOver10byGroup$Proportion > 0.25)
```

Let us then define the count of groups in which each gene is detected
above **10** CPM in over **25%** of the group:

```{r groupsOver25pct}
groupsOver25pct <- as.data.frame(with(
    propOver10byGroup,
    tapply(PropOver25pct, Feature, function(x){sum(x)}))
)
colnames(groupsOver25pct) <- "Groups"
groupsOver25pct$Feature <- rownames(groupsOver25pct)
```

We may now merge this information with the features detected
above **10** CPM in at least **1** group:

```{r detected_CPMandProportion}
detectedCPMandProp <- merge(
  subset(propOver10byGroup, Proportion > 0.25), groupsOver25pct, by = "Feature"
)
detectedCPMandProp <- merge(
  detectedCPMandProp,
  as.data.frame(unique(colData(sce.pass)[,c("Group","Time","Treatment")])),
  by = "Group"
)
```

And visualise the count of feature detected above **10** CPM
of at least **25%** of cells in each group, while indicating
the count of groups in which each feature passed the above detection cut-offs:

```{r detectedCPMandProp_bar, echo=FALSE, fig.height=6}
countPerGroup <- data.frame(with(colData(sce.pass), table(Time, Treatment)))
countPerGroup$Treatment <- with(countPerGroup, factor(
  gsub("_", " ", Treatment),
  gsub("_", " ", levels(Treatment))
))
detectedCPMandProp$Time <- factor(
  gsub("(.*)_.*_.*", "\\1", detectedCPMandProp$Group), levels(sce.pass$Time)
)
detectedCPMandProp$Infection <- factor(
  gsub(".*_(.*)_.*","\\1",detectedCPMandProp$Group), levels(sce.pass$Infection)
)
detectedCPMandProp$Status <- factor(
  gsub(".*_.*_(.*)", "\\1", detectedCPMandProp$Group), levels(sce.pass$Status)
)
detectedCPMandProp$Treatment <- with(detectedCPMandProp, factor(
  gsub("_", " ", Treatment),
  gsub("_", " ", levels(Treatment))
))
ggplot(detectedCPMandProp) +
  geom_bar(aes(Treatment, fill = as.factor(Groups))) +
  geom_label(
    aes(Treatment, 0, label = Freq), countPerGroup, size = 2
  ) +
  facet_grid(~ Time) +
  theme_bw() +
  labs(y = "Features", x = "Treatment") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_y_continuous(limits = c(0, 6E3)) +
  guides(fill = guide_legend(title = "25% group\n>10 CPM"))
```

```{r detectedCPMandProp_pdf, include=FALSE}
ggsave("03_out/countFeatures_10CPM_25percent_byGroup.pdf", width = 6, height = 6)
```

Let us now retain for further analysis:

* endogenous features detected above **10** CPM in at least **25%** of
  at least **1** experimental group of cells
* ERCC spike-in features detected at an average level greater than **1** count
  across all cells:

```{r detectedFeatures}
detectedFeatures <- names(which(tapply(
  propOver10byGroup$Proportion, propOver10byGroup$Feature,
  function(x){sum(x > 0.25) > 0})))
```

```{r keep}
keepEndogenous <- (rownames(sce.pass) %in% detectedFeatures) &
  (!isSpike(sce.pass, "ERCC"))
keepSpike <- (rowMeans(counts(sce.pass)) > 1) & (isSpike(sce.pass))
table(keepEndogenous); table(keepSpike)
```

Finally, we may extract detected features
(endogenous and ERCC spike-in molecules) into a new `SCESet`:

```{r sce.filtered}
sce.filtered <- sce.pass[keepEndogenous | keepSpike,]
dim(sce.filtered)
```

# References

<!-- As HTML pages are built independently, objects need to be saved to disk
and imported as needed by the other R mardown files -->

```{r save_sce.pass, include=FALSE}
saveRDS(sce.filtered, "rds/sce.filtered.rds")
```
