---
title: "Monocle 2 analysis"
author: "Kevin Rue-Albrecht"
date: "14/07/2018"
output: pdf_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
require(monocle)
require(cowplot)
require(circlize)
require(RColorBrewer)
require(cba)
require(ComplexHeatmap)
sce.endo <- readRDS("rds/sce.endo_clusters.rds")
outdir <- "19_out"; dir.create(outdir, showWarnings = FALSE)
```

```{r}
col.time <- brewer.pal(9, "Set3")[c(2,7:8)]
col.infection <- brewer.pal(12, "Paired")[c(9,4,2)]
col.status <- brewer.pal(12, "Paired")[c(9,1,7)]
names(col.time) <- levels(sce.endo$Time)[1:3]
names(col.infection) <- levels(sce.endo$Infection)
names(col.status) <- levels(sce.endo$Status)
```

# Prepare the data set

Prepare the full data set for Monocle:

```{r}
fd <- rowData(sce.endo)
fd$gene_short_name <- fd$gene_name
# Create a CellDataSet from the relative expression levels
HSMM <- newCellDataSet(
    as(assay(sce.endo, "counts"), "sparseMatrix"),
    phenoData = AnnotatedDataFrame(as.data.frame(colData(sce.endo))),
    featureData = AnnotatedDataFrame(as.data.frame(fd)),
    lowerDetectionLimit=0.5,
    expressionFamily=negbinomial.size()
)
HSMM <- detectGenes(HSMM, min_expr = 5)
HSMM <- HSMM[fData(HSMM)$num_cells_expressed > 10, ]
HSMM <- estimateSizeFactors(HSMM)
HSMM <- estimateDispersions(HSMM)
```

# Trajectory step 1: choose genes that define a cell's progress

## Ordering based on genes that differ between 2h and 6h in challenged cells

Prepare a data subset for Monocle:

```{r}
cells_for_contrast <- colnames(sce.endo)[
    sce.endo$Infection %in% c("STM-LT2", "STM-D23580") &
        sce.endo$Time %in% c("2h", "6h")
]
fd <- rowData(sce.endo)
fd$gene_short_name <- fd$gene_name
# Create a CellDataSet from the relative expression levels
HSMM_challenged <- newCellDataSet(
    as(assay(sce.endo[, cells_for_contrast], "counts"), "sparseMatrix"),
    phenoData = AnnotatedDataFrame(as.data.frame(
        droplevels(colData(sce.endo[, cells_for_contrast]))
        )),
    featureData = AnnotatedDataFrame(as.data.frame(fd)),
    lowerDetectionLimit=0.5,
    expressionFamily=negbinomial.size()
)
HSMM_challenged <- detectGenes(HSMM_challenged, min_expr = 5)
HSMM_challenged <- HSMM_challenged[fData(HSMM_challenged)$num_cells_expressed > 10, ]
HSMM_challenged <- estimateSizeFactors(HSMM_challenged)
HSMM_challenged <- estimateDispersions(HSMM_challenged)
```

Find all genes that are differentially expressed between the start and the end of the infection time course in challenged cells:

```{r}
diff_test_res <- differentialGeneTest(HSMM_challenged, fullModelFormulaStr = "~Time")
write.csv(diff_test_res, file.path(outdir, "differentialGeneTest_time.csv"))
ordering_genes <- row.names (subset(diff_test_res, qval < 0.001)) # more stringent
length(ordering_genes)
```

Once we have a list of gene ids to be used for ordering, we need to set them in the _main_ HSMM object, because the next several functions will depend on them.

```{r}
HSMM <- setOrderingFilter(HSMM, ordering_genes)
plot_ordering_genes(HSMM)
```

# Trajectory step 2: reduce data dimensionality

```{r}
HSMM <- reduceDimension(HSMM, max_components = 2, method = 'DDRTree')
```

# Trajectory step 3: order cells along the trajectory

Now that the space is reduced, it's time to order the cells using the orderCells function as shown below.

```{r}
HSMM <- orderCells(HSMM)
```

Once the cells are ordered, we can visualize the trajectory in the reduced dimensional space.

```{r}
p1 <- plot_cell_trajectory(HSMM, color_by = "State")
p2 <- plot_cell_trajectory(HSMM, color_by = "Time")
p3 <- plot_cell_trajectory(HSMM, color_by = "Infection")
p4 <- plot_cell_trajectory(HSMM, color_by = "Status")
gg <- plot_grid(p1, p2, p3, p4)
ggsave(file.path(outdir, "trajectory_state_colData.pdf"), gg, width = 10, height = 8)
```

# Identify genes that vary over pseudotime

```{r}
diff_test_res <- differentialGeneTest(HSMM, fullModelFormulaStr = "~sm.ns(Pseudotime)")
table(diff_test_res$qval < 0.001)
write.csv(diff_test_res, file.path(outdir, "differentialGeneTest_pseudotime.csv"))
```

# Heat map of gene expression over pseudotime

First part, ordering cells and computing distances between genes:

```{r}
orderedCells <- colnames(HSMM)[order(HSMM$Pseudotime)]
deGenes <- subset(diff_test_res, qval < 0.001, "gene_id", drop=TRUE)
# assayNames(sce.endo)
heatmapMatrix <- as.matrix(assay(sce.endo[deGenes, orderedCells], "logcounts"))
heatmapMatrix <- t(scale(t(heatmapMatrix)))
# Gene clusters
# K-means scree plot ----
k_values <- seq(1,15,1)
itermax <- 1E3

KmeanOut <- file.path(
  outdir,
  "kmean",
  paste("iter", itermax, sep = "_")
)
dir.create(KmeanOut, recursive = TRUE)

wss <- c()
for (test_k in k_values){
    message("k: ", test_k)
    k.genes <- kmeans(heatmapMatrix, centers = test_k, iter.max = itermax)
    wss <- c(wss, sum(k.genes$withinss))
}
names(wss) <- k_values
pdf(
  file.path(KmeanOut, sprintf(
    "rowscaled_kmeans_scree.pdf"
  )),
  width = 10, height = 5
)
plot(names(wss), wss, type="b", xlab="Number of Clusters",ylab="Within groups sum of squares")
dev.off()
# Clustering
distGenes <- dist(heatmapMatrix)
clustGenes <- hclust(distGenes)
```

Decide the count of gene clusters

```{r}
k_mean <- 3
geneClusters <- cutree(clustGenes, k = k_mean)
# Color maps
ht_scale <- colorRamp2(
    breaks = seq(-2, 2, length.out = 9),
    colors = rev(brewer.pal(9, "RdBu"))
)
pseudotime_scale <- colorRamp2(
    breaks = seq(0, max(HSMM$Pseudotime), length.out = 5),
    colors = brewer.pal(5, "BuGn")
)
ha <- HeatmapAnnotation(
    # p1 = anno_points(pData(HSMM)[orderedCells, "Pseudotime"], axis = FALSE),
    # annotation_height = c(rep(2, 4), 10),
    pData(HSMM)[orderedCells, c("Pseudotime", "Time", "Infection", "Status")],
    col = list(
        Pseudotime = pseudotime_scale,
        Time = col.time,
        Infection = col.infection,
        Status = col.status
    )
)
# Heat map
hm <- Heatmap(
    matrix = heatmapMatrix, col=ht_scale,
    top_annotation = ha,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    show_row_names = FALSE,
    cluster_rows = TRUE,
    split = geneClusters,
    top_annotation_height = unit(4, "cm"),
    gap = unit(0.25, "cm"),
    row_names_gp = gpar(fontsize=8))
pdfName <- sprintf("heatmap_pseudotime_DEpseudotime_001_k%i_rowscaled.pdf", k_mean)
pdf(file.path(outdir, pdfName), height = 15, width = 30)
draw(hm)
dev.off()
```

Gene ontology of gene clusters

```{r}
geneGroups <- split(names(geneClusters), geneClusters)
lengths(geneGroups)
```

```{r}
for (groupname in names(geneGroups)) {
    geneNames <- rowData(sce.endo)[geneGroups[[groupname]], "gene_name"]
    csvName <- sprintf("GeneNames_001_k%i_group%s.csv", k_mean, groupname)
    write.csv(data.frame(gene_name = geneNames), file.path(outdir, csvName))
}
```

```{r}
require(org.Hs.eg.db)
columns(org.Hs.eg.db)
keytypes(org.Hs.eg.db)
```

```{r}
require(goseq)
geneLengths <- width(sce.endo)
names(geneLengths) <- rownames(sce.endo)
for (groupName in names(geneGroups)) {
    genes <- (rownames(sce.endo) %in% geneGroups[[groupName]])
    names(genes) <- rownames(sce.endo)
    pwf <- nullp(genes, bias.data = geneLengths[names(genes)])
    pvals <- goseq(pwf, 'hg38', 'ensGene')
    pvals$BH <- p.adjust(pvals$over_represented_pvalue, "BH")
    pvals$bonferroni <- p.adjust(pvals$over_represented_pvalue, "bonferroni")
    GOtoAnnotate <- subset(pvals, bonferroni < 0.05, "category", drop=TRUE)
    message("bonferroni < 0.05: ", length(GOtoAnnotate))
    # Add gene names
    rownames(pvals) <- pvals$category
    pvals$Genes <- ""
    message("To do:", nrow(GOtoAnnotate))
    for (tmp_go in GOtoAnnotate) {
        message("Processing: ", match(tmp_go, pvals$category))
        message("pvalue: ", pvals[tmp_go, "over_represented_pvalue"])
        message("numDEInCat: ", pvals[tmp_go, "numDEInCat"])
        GO1geneInfo <- select(x = org.Hs.eg.db, keys = tmp_go, columns = c("SYMBOL", "ENSEMBL"), keytype = "GOALL")
        GO1geneInfoDE <- subset(GO1geneInfo, ENSEMBL %in% geneGroups[[groupName]])
        tmpSymbolsDE <- unique(GO1geneInfoDE$SYMBOL)
        message("Found in org.Hs.eg.db: ", length(tmpSymbolsDE))
        pvals[tmp_go, "Genes"] <- paste(tmpSymbolsDE, collapse = ",")
    }
    #
    csvName <- sprintf("GO_001_k%i_group%s_count%i.csv", k_mean, groupName, sum(genes))
    write.csv(pvals, file.path(outdir, csvName))
}
```
