---
title: Feature selection
author:
  - name: Anna Aulicino
    email: anna.aulicino@ndm.ox.ac.uk
    affiliation: MRC,Human Immunology Unit, Weatherall Institute of Molecular
        Medicine, NIHR Biomedical research centre, University of Oxford,
        John Radcliffe Hospital, Oxford, UK.
  - name: KÃ©vin Rue-Albrecht
    email: kevinrue67@gmail.com
    affiliation: Department of Medicine, Imperial College London,
        Hammersmith Campus, Du Cane Road, London, W12 0NN, UK
date: "`r doc_date()`"
package: "`r pkg_ver('BiocStyle')`"
abstract: >
  Quality control of endogenous features in single cells that passed quality
  control.
vignette: >
  %\VignetteIndexEntry{Quality Control}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::pdf_document2:
    toc_newpage: true
bibliography:
  QC.bib
---

# Prior work

Prior to this document, quantitation data for all 384 samples
(including single cells, blanks and bulks) was obtained using
[Kallisto](https://pachterlab.github.io/kallisto/)
as follows:

```
kallisto quant \
    --index=$kallistoIdx \
    --output-dir=$kallistoDir/$cell \
    --bootstrap-samples=100 \
    $fastqFolder/${cell}_1.fastq.gz $fastqFolder/${cell}_2.fastq.gz
```

Subsequently:

* The quantitation data was imported into the _R_ environment.
* Phenotype data was attached imported and attached to the `SCESet`
* Transcript-level expression data was aggregated at the gene level
* Gene identifiers were made more intuitive by prepending gene symbol where
  possible
* Outliers were identified and removed among single cells,
  using blank and bulk samples as reference for certain metrics.
* The remaining single cells were serialised as a `SCESet` and written to file.

# Settings

Packages required for this vignette:

```{r packages, message=FALSE}
library(scater)
library(ensembldb)
library(EnsDb.Hsapiens.v79)
library(scran)
library(reshape2)
library(ggrepel)
library(dplyr)
library(BiocParallel)
```

Let us set the working directory to execute the following code chunks:

```{r workdir}
workdir <- ifelse(interactive(), ".", "..")
```

Let us set a few paths relevant to the analysis:

```{r paths}
folder.rds <- file.path(workdir, "rds")
folder.expdata <- file.path(workdir, "expData")
```

# Import preprocessed data

Here, we simply import the preprocessed data:

```{r DC_all}
DC_sc <- readRDS(file.path(folder.rds, "goodCells.rds"))
```

# Overwrite QC metrics

In the previous vignette, QC metrics were calculated using _all_ samples
in the data set.
Exclusion of outliers, blanks, and bulks has not updated certain metrics
depending on the number of cells; for instance, `n_cells_exprs`
(the number of cells for which the expression level of the feature is above
the detection limit).
Let us call again the `calculateQCMetrics` method on the subsetted data set:

```{r calculateQCMetrics}
MT_ids <- genes(
  EnsDb.Hsapiens.v79, "gene_id", SeqnameFilter("MT", "=")
  )$gene_id
DC_sc <- calculateQCMetrics(
  DC_sc,
  feature_controls = list(
    ERCC = grepl("ERCC-[[:digit:]]", featureNames(DC_sc)),
    MT = fData(DC_sc)$feature_id %in% MT_ids),
  nmads = 3
)
```

# Gene expression and cell detection level

## Number of features detected and cell counts

### Across all cells 

Let us first visualise the count of detectable features and the number of cells
in which they are detected.
Let us also store in an additional column of the `phenoData` slot the count of
_single cells_ in which each feature was detected:

```{r countSCdetected}
fData(DC_sc)$n_sc_exprs <- rowSums(
    exprs(DC_sc[,!DC_sc$is_cell_control]) > 0
)
```

We may then visualise the number of features detected in each number of cells:

```{r featuresDetectedSC, echo=FALSE, out.width="100%"}
tmp <- data.frame(
    N_cells = 0:sum(!DC_sc$is_cell_control)
)
tmp$N_features <- sapply(tmp$N_cells, function(x){
    sum(fData(DC_sc)$n_sc_exprs >= x)
})
ggplot(tmp, aes(x = N_cells, y = N_features)) +
  geom_point(size = 0.5) +
  scale_y_continuous(limits = c(0, nrow(DC_sc)), labels = scales::comma) +
  labs(
      x = "Detection cut-off (cells)",
      y = "Features detected"
  ) +
  geom_text_repel(
    aes(label = N_features),
    dplyr::filter(tmp, N_cells %in% c(50, 100, 150, 200, 250, 300, 350)),
    min.segment.length = unit(0, 'lines'),
    point.padding = unit(0.5, 'lines')
  )
rm(tmp)
```

### Within experimental groups

Let us examine the number of detected features as a function of the proportion
of _non-outlier_ single cells in which they were detected, within each
experimental group.
First, let us subset the data set to only single cells that passed all
the outlier detection filters detailed in the previous sections.
In addition, let us create two subsets:

* one that contains only ERCC spike-in control features
* one that contains only endogenous gene features

```{r subsetDetectionExpGroup}
sc_DC.subset.ERCC <- DC_sc[fData(DC_sc)$is_feature_control,]
sc_DC.subset.genes <- DC_sc[!fData(DC_sc)$is_feature_control,]
```

Before going further let us examine how many cells are left
in each experimental group:

```{r tablePhenoExpGroup}
print(tt <- table(pData(sc_DC.subset.genes)[,c("Infection","Status","Time")]))
```

Now, for each experimental group `Infection:Status:Time`, let us obtain and
visualise the count of features detected against the count of single cells
they were detected in:

```{r genesCountExpGroup, include=FALSE}
# List of experimental groups
df <- distinct(pData(sc_DC.subset.genes), Infection, Status, Time)
# Data points for each experimental group
dfl <- bplapply(seq_len(nrow(df)), function(r){
    infection <- df[r,"Infection"]
    status <- df[r,"Status"]
    time <- df[r,"Time"]
    cells <- tt[infection, status, time]
    n_cells <- seq_len(cells)
    n_features <- sapply(n_cells, function(x){
        sum(
            rowSums(
                exprs(sc_DC.subset.genes[
                    ,sc_DC.subset.genes$Infection == infection &
                        sc_DC.subset.genes$Status == status &
                        sc_DC.subset.genes$Time == time
                ]) > 0
            ) >= x
        )
    })
    return(data.frame(
        Infection = infection,
        Status = status,
        Time = time,
        Cells = n_cells,
        Features = n_features
    ))
})
ddfl <- do.call(rbind, dfl)
rm(dfl)
```

```{r plotExpGroupGenes, echo=FALSE}
p.sc_detection_genes <- ggplot(ddfl) +
    scale_y_continuous(
        limits = c(0, nrow(sc_DC.subset.genes)),
        labels = scales::comma
    ) +
    labs(
        x = "Detection cut-off (cells)",
        y = "Features detected"
    )
p.sc_detection_genes +
    geom_point(aes(x=Cells, y=Features, colour=Status), size = 0.5) +
    facet_grid(Time ~ Infection) +
    ggtitle("Endogenous feature")
rm(ddfl)
```

And the same approach applied only to ERCC spike-ins:

```{r countExpGroupERCC, include=FALSE}
# Data points for each experimental group
dfl <- bplapply(seq_len(nrow(df)), function(r){
    infection <- df[r,"Infection"]
    status <- df[r,"Status"]
    time <- df[r,"Time"]
    cells <- tt[infection, status, time]
    n_cells <- seq_len(cells)
    n_features <- sapply(n_cells, function(x){
        sum(
            rowSums(
                exprs(sc_DC.subset.ERCC[
                    ,sc_DC.subset.ERCC$Infection == infection &
                        sc_DC.subset.ERCC$Status == status &
                        sc_DC.subset.ERCC$Time == time
                ]) > 0
            ) >= x
        )
    })
    return(data.frame(
        Infection = infection,
        Status = status,
        Time = time,
        Cells = n_cells,
        Features = n_features
    ))
})
ddfl <- do.call(rbind, dfl)
rm(dfl, df)
```

```{r plotExpGroupERCC, echo=FALSE}
p.sc_detection_ERCC <- ggplot(ddfl) +
    geom_point(aes(x=Cells, y=Features, colour=Status), size = 0.5) +
    scale_y_continuous(
        limits = c(0, nrow(sc_DC.subset.ERCC)),
        labels = scales::comma
    ) +
    labs(
        x = "Detection cut-off (cells)",
        y = "Features detected"
    )
p.sc_detection_ERCC +
    geom_point(aes(x=Cells, y=Features, colour=Status), size = 0.5) +
    facet_grid(Time ~ Infection) +
    ggtitle("ERCC spike-ins")
rm(ddfl)
```

See [Appendix 3](#Appendix3) for an overview of the nine panels overlaid.


## Average expression and cell counts

In order to select expressed features appropriate for downstream analyses,
let us visualise, for each measured feature (including ERCC spike-in features),
the count of cells expressing each feature against the log~10~-transformed
average estimated counts.

First, let us calculate the average estimate counts for each gene:

```{r avgCounts}
ave.counts <- rowMeans(counts(DC_sc))
```

We may then identify features showing an average estimated count above `1`:

```{r avgCountsGe1}
keep <- ave.counts >= 1
table(keep)
```

Let us visualise the distribution of average estimated counts,
and mark the cut-off value of `1` used to remove features associated
with excessive dropout events:

```{r histAvgCount}
hist(
  log10(ave.counts), breaks=100, main="", col="grey80",
  xlab=expression(Log[10]~"average count")
)
abline(v=log10(1), col="blue", lwd=2, lty=2)
```

In this plot, the cut-off line crosses the peak, rather than the more flat
region on the left of the peak, as suggested in [@67].
However, the fact that this data set includes more cells allows the
average estimated counts to reach lower values
(_e.g._ a single estimated count in a single cell, divided by `400` cells
produced a value of `0.0025`, while in `100` cells it produces a minimal value
of `0.01`).

It is therefore important to represent the average estimated count value
against the number of cells in which each feature is detected,
and mark the cutoff value of `1` as above:

```{r smoothScatter}
numcells <- nexprs(DC_sc, byrow=TRUE)
setSpike(DC_sc) <- "ERCC"
smoothScatter(
  log10(ave.counts), numcells,
  xlab = expression(Log[10]~"average count"),
  ylab = "Number of expressing cells"
)
points(
  log10(ave.counts[isSpike(DC_sc)]),
  numcells[isSpike(DC_sc)], col = "red", pch = 16, cex = 0.5)
abline(v = 0, col = "orange")
rm(numcells)
```

# Filter detected features

```{r filterFeatures}
DC_detected <- DC_sc[keep,]
```

# Normalisation

This section applies the normalisation procedure as presented in [@67]
for the haematopoietic stem cells (HSCs) data set.

## Cell-specific biases

Let us compute size factors using the `r Biocpkg("scran")` `computeSumFactors`
method:

```{r computeSumFactors}
DC_detected <- computeSumFactors(DC_detected, sizes=c(20, 40, 60, 80))
summary(sizeFactors(DC_detected))
```

The range of size factors is similar to that of the HSCs data set.
This may indicate that the outlier detection procedure applied in the previous
vignette has successfully removed extreme outliers, leaving an homogeneous
data set (as homogeneous as can be an experimental design including three
time points, uninfected cells, and cells exposed to or infected by two
different bacterial strains).

```{r plotsizeFactors}
plot(
  sizeFactors(DC_detected), DC_detected$total_counts/1e6, log="xy",
  ylab="Library size (millions)", xlab="Size factor")
cor.test(
  sizeFactors(DC_detected), DC_detected$total_counts/1e6,
  alternative = "greater")
```

In contrast to [@67], the size factors computed for this data set do not
show an important scatter around the trend of correlation.
This suggests that the systematic differences between cells are _not_ primarily
driven by differences in capture efficiency or sequencing depth.
Instead, this observation hints to the existence of differential expression
between single cells, which is expected from the experimental design.

## Separate size factors for spike-in transcripts

Size factors computed from the counts for endogenous genes are usually not
appropriate for normalising the counts for spike-in transcripts.
To ensure normalisation is performed correctly, we compute a separate set of
size factors for the spike-in set.

```{r computeSpikeFactors}
DC_detected <- computeSpikeFactors(DC_detected, type="ERCC", general.use=FALSE)
```

## Applying the size factors to normalize gene expression

As explained in [@67], the (estimated) count data are used to compute
normalised log-expression values for use in downstream analyses.
Each value is defined as the log-ratio of each count to the size factor for the
corresponding cell, after adding a prior count of `1` to avoid undefined values
at zero counts.
Division by the size factor ensures that any cell-specific biases are removed.

Spike-in-specific size factors are automatically applied to normalize the
spike-in transcripts separately from the endogenous genes.

```{r normalize}
DC_normalised <- normalize(DC_detected)
```

# Save the normalised data set to file

```{r saveRDS}
saveRDS(DC_normalised, file.path(folder.rds, "normalised.rds"))
```

# Visualise normalised data

First of all, let us subset the data set, excluding ERCC spike-in features:

```{r excludeERCC}
endo_sc <- DC_normalised[!fData(DC_normalised)$is_feature_control_ERCC,]
```

## PCA - All samples

### All features

First, let us visualise the two principal components of highest variance
identified considering all the features in the 

```{r prcomp, fig.height=10}
dc_pca.samples <- prcomp(t(assayData(endo_sc)[["norm_exprs"]]))
pca.samples <- cbind(dc_pca.samples$x, pData(endo_sc))
multiplot(
  ggplot(pca.samples, aes(x = PC1, y = PC2, shape = Time, colour = Status)) +
    geom_point(), # + geom_density_2d(aes(linetype = Time))
  ggplot(pca.samples, aes(x = PC1, y = PC2, shape = Time, colour = Infection)) +
    geom_point() #  + geom_density_2d(aes(linetype = Time))
)
```

Let us also repeat the same figure, adding 2D density lines to emphasise the
clustering of cells by `Time` and `Status`:

```{r density2D, fig.height=10}
multiplot(
  ggplot(pca.samples, aes(x = PC1, y = PC2, shape = Time, colour = Status)) +
    geom_point() + geom_density_2d(aes(linetype = Time)),
  ggplot(pca.samples, aes(x = PC1, y = PC2, shape = Time, colour = Infection)) +
    geom_point() + geom_density_2d(aes(linetype = Time))
)
```

### Most variable features

Let us use the `300` most variable features
to obtain an overview of the normalised data set using a separate
principal component analysis:

```{r plotPCASCESet, fig.height=10}
multiplot(
  plotPCASCESet(endo_sc, 300, 3, colour_by = "Status", shape_by = "Time"),
  plotPCASCESet(endo_sc, 300, 3, colour_by = "Infection", shape_by = "Time")
)
```

Several important observations:

* `uninfected` samples show a striking homogeneity, relative to the expression
  profile of stimulated cells increasingly diverging at `4h` and `6h` after
  stimulation.
* `PC3` differentiates to some extent `infected` and `exposed` cells at `6h`
  after stimulation.
* `Time` remains the experimental factor most correlated with `PC1`
  (R^2^: 0.63; see [Appendix 1A](#AppendixPcsTime)),
  similarly to principal component analysis prior to outlier detection
  and removal in a precedent vignette.

Let us immediately find the principal components most associated with each
experimental factor in the study (figures shown in [Appendix 1](#Appendix1)):

```{r findImportantPCs}
pcs.time <- findImportantPCs(endo_sc, "Time", ntop = 300)
pcs.status <- findImportantPCs(endo_sc, "Status", ntop = 300)
pcs.infection <- findImportantPCs(endo_sc, "Infection", ntop = 300)
```

\newpage

## PCA - By time point

The overwhelming effect of `Time` on expression profiles may hide more subtle
differences that exist between cells of distinct `Status` or `Infection`.
Such differences may be revealed by principal component analyses performed
within each time point:

```{r plotPCASCESet2h, fig.height=10}
multiplot(
  plotPCASCESet(
    endo_sc[,endo_sc$Time == "2h"], 300, 3,
    colour_by = "Status", shape_by = "Infection"
  ),
  plotPCASCESet(
    endo_sc[,endo_sc$Time == "2h"], 300, 3,
    colour_by = "Infection", shape_by = "Status"
  )
)
```

\newpage

```{r plotPCASCESet4h, fig.height=10}
multiplot(
  plotPCASCESet(
    endo_sc[,endo_sc$Time == "4h"], 300, 3,
    colour_by = "Status", shape_by = "Infection"
  ),
  plotPCASCESet(
    endo_sc[,endo_sc$Time == "4h"], 300, 3,
    colour_by = "Infection", shape_by = "Status"
  )
)
```

\newpage

```{r plotPCASCESet6h, fig.height=10}
multiplot(
  plotPCASCESet(
    endo_sc[,endo_sc$Time == "6h"], 300, 3,
    colour_by = "Status", shape_by = "Infection"
  ),
  plotPCASCESet(
    endo_sc[,endo_sc$Time == "6h"], 300, 3,
    colour_by = "Infection", shape_by = "Status"
  )
)
```

While `PC1` clearly separates `uninfected` cells from stimulated ones,
discriminates to some extent cells infected by `D23580` from those infected by
`LT2`, as reinforced  by the 

```{r importantPCs6h, fig.height=10}
findImportantPCs(endo_sc[,endo_sc$Time == "6h"], "Infection")
```

# Appendix {#Appendix}

## Principal components and experimental covariates {#Appendix1}

### Time {#AppendixPcsTime}

```{r pcsTime, out.width='80%', fig.width = 8, fig.height = 12}
pcs.time
```

### Status {#AppendixPcsStatus}

```{r pcsStatus, out.width='80%', fig.width = 8, fig.height = 12}
pcs.status
```

### Infection {#AppendixPcsInfection}

```{r pcsInfection, out.width='80%', fig.width = 8, fig.height = 12}
pcs.infection
```

# Session info

Here is the output of `sessionInfo()` on the system on which this
document was compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```


[R]: http://r-project.org
[RStudio]: http://www.rstudio.com/
