---
title: Preprocessing
author:
  - name: Anna Aulicino
    email: anna.aulicino@ndm.ox.ac.uk
    affiliation: MRC,Human Immunology Unit, Weatherall Institute of Molecular
        Medicine, NIHR Biomedical research centre, University of Oxford,
        John Radcliffe Hospital, Oxford, UK.
  - name: KÃ©vin Rue-Albrecht
    email: kevinrue67@gmail.com
    affiliation: Department of Medicine, Imperial College London,
        Hammersmith Campus, Du Cane Road, London, W12 0NN, UK
date: "`r doc_date()`"
package: "`r pkg_ver('BiocStyle')`"
abstract: >
  Preprocessing of _Kallisto_ quantitation data, including aggregation
  of transcript-level expression data into gene-level estimates, and
  annotation of gene identifiers with gene symbol annotation.
vignette: >
  %\VignetteIndexEntry{Quality control}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::pdf_document2
---

# Pre-processed data

Prior to this document, quantitation data for all 384 samples
(including single cells, blanks and bulks) was obtained using _Kallisto_ 
as follows:

```
kallisto quant \
    --index=$kallistoIdx \
    --output-dir=$kallistoDir/$cell \
    --bootstrap-samples=100 \
    $fastqFolder/${cell}_1.fastq.gz $fastqFolder/${cell}_2.fastq.gz
```

# Settings

Packages required for this vignette:

```{r packages, message=FALSE}
library(scater)
library(EnsDb.Hsapiens.v79)
```

```{r workdir}
workdir <- ifelse(interactive(), ".", "..")
```

Let us set the working directory to execute the following code chunks:

```{r workdir}
workdir <- ifelse(interactive(), ".", "..")
```

Let us set a few paths relevant to the analysis:

```{r paths}
folder.kallisto <- file.path(workdir, "boot")
folder.expData <- file.path(workdir, "expData")
folder.rds <- file.path(workdir, "rds")
```

Create folders as necessary:

```{r}
if (!dir.exists(folder.rds)){
  dir.create(folder.rds)
}
```

# Import data

## Transcript quantitation

Let us import the quantitation data, using the trimmed folder name
as a unique identifier for each sample:

```{r kallistoDirs, include=FALSE}
kallisto.dirs <- list.files(folder.kallisto)
length(kallisto.dirs)
```

```{r readKallistoResults, message=FALSE}
sc_DC <- readKallistoResults(
    samples = gsub("WTCHG_[[:digit:]]*_", "", kallisto.dirs),
    directories = file.path(folder.kallisto, kallisto.dirs)
)
```

We can examine the resulting `SCESet`, highlighting how the sum of `TPM` values
totals one million in all samples:

```{r tpmTx}
dim(sc_DC)
summary(colSums(tpm(sc_DC)))
ggplot(data.frame(sum = colSums(tpm(sc_DC)))) +
  geom_point(aes(x = seq_along(sum), y=sum)) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  labs(x = "Sample", title = "Sum of TPM across transcripts")
```

## Phenotypes

Then, let us attach experimental covariates to the `SCESet`, in the `phenoData`
slot:

```{r expData}
varLabels(sc_DC)
samplesData <- read.csv(file.path(folder.expData, "samples.csv"))
pData(sc_DC) <- cbind(samplesData, pData(sc_DC))
```

# Collapse transcript to genes

## Trim transcript identifiers

The index file supplied to _Kallisto_ was generated from the file
`Homo_sapiens.GRCh38.cdna.all.fa`, which supplied sequence information
annotated to versioned Ensembl transcript identifiers of the form
`ENST00000448914.1`.
Before gene annotations may be retrieved from the Ensembl Biomart, the
version identifier `.x` of each transcript must be trimmed:

```{r trimTxId}
fData(sc_DC)$feature_id <- gsub("\\.[0-9]+$", "", fData(sc_DC)$feature_id)
```

## Use `EnsDb` annotation packages

Let us annotate each transcript identifier with the corresponding
gene identifier and gene name using the `EnsDb.Hsapiens.v79` package[^1]:

[^1]: GRCh38 genome build.

```{r mapIds, results='asis'}
fData(sc_DC)$ensembl_gene_id <- mapIds(
    EnsDb.Hsapiens.v79,
    gsub("\\.[[:digit:]]+", "", fData(sc_DC)$feature_id),
    "GENEID", "TXID"
)
fData(sc_DC)$ensembl_gene_id <- ifelse(
  is.na(fData(sc_DC)$ensembl_gene_id),
  fData(sc_DC)$feature_id,
  fData(sc_DC)$ensembl_gene_id
)
pander::pandoc.table(head(fData(sc_DC)))
```

Once the `SCESet` contains gene annotations, transcript-level expression
estimates produced by _Kallisto_ can be collapsed to gene-level estimates
using the `summariseExprsAcrossFeatures` method:

```{r summariseExprsAcrossFeatures}
sc_DC_gene <- summariseExprsAcrossFeatures(
  sc_DC, summarise_by = "ensembl_gene_id"
)
names(fData(sc_DC_gene))
colnames(fData(sc_DC_gene)) <-
    gsub("exprs_collapsed_to", "feature_id", names(fData(sc_DC_gene)))
```

```{r tpmGene}
dim(sc_DC_gene)
summary(colSums(tpm(sc_DC_gene)))
ggplot(data.frame(sum = colSums(tpm(sc_DC_gene)))) +
  geom_point(aes(x = seq_along(sum), y=sum)) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  labs(x = "Sample", title = "Sum of TPM across summarised genes")
```

# Add gene symbols to featureNames to make them more intuitive

```{r mapGeneSymbols}
fData(sc_DC_gene)$feature_id <- as.character(fData(sc_DC_gene)$feature_id)
fData(sc_DC_gene)$gene_name <- mapIds(
  EnsDb.Hsapiens.v79, fData(sc_DC_gene)$feature_id, "GENENAME", "GENEID"
)
```

Now, we may rename features by prepending their gene unique gene identifier
with their official gene symbol, when available:

```{r updateFeatureNames}
featureNames(sc_DC_gene) <- ifelse(
    is.na(fData(sc_DC_gene)$gene_name),
    as.character(fData(sc_DC_gene)$feature_id),
    paste(
        fData(sc_DC_gene)$gene_name,
        fData(sc_DC_gene)$feature_id,
        sep = "_"
    )
)
head(fData(sc_DC_gene), n = 3)
```

# Save the updated SCESet to file

```{r saveRDS}
saveRDS(sc_DC_gene, file.path(folder.rds, "sc_DC_EnsDb.rds"))
```

# Session info

Here is the output of `sessionInfo()` on the system on which this
document was compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```


[R]: http://r-project.org
[RStudio]: http://www.rstudio.com/
