---
title: "Import of RNA-Sequencing read counts and associated metadata"
---

<!-- Enable caching -->
```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(scater)
library(edgeR)
library(rtracklayer)
library(DT)
```

# Phenotype information

Let us import experimental information:

```{r importPheno}
pheno <- xlsx::read.xlsx("expdata/samples.xlsx", sheetName = "pheno")
DT::datatable(
  pheno,
  options = list(pageLength = 10, searching = TRUE), filter = "top",
  rownames = FALSE
)
```

To facilitate later use, let us retype some fields, reorder some factor levels,
and add an additional field `Sample` field that contains a shortened unique
identifier derived from the count file name for each sample:

```{r editPheno}
pheno$File <- as.character(pheno$File)
pheno$Infection <- factor(pheno$Infection, c("Mock","D23580","LT2","Blank"))
pheno$Status <-
  factor(pheno$Status, c("uninfected","exposed","infected","Blank","BULK"))
pheno$Lane <- as.factor(pheno$Lane)
pheno$Sample <- gsub("WTCHG_[[:digit:]]+_", "", pheno$File) # character
pheno$sampleAnnotated <-
  with(pheno, paste(Time, Infection, Status, Sample, sep = "_")) # character
pheno$Treatment <-
  droplevels(interaction(pheno$Infection, pheno$Status, sep = "_")) # factor
pheno$Group <-
  droplevels(interaction(pheno$Time, pheno$Treatment, sep = "_")) # factor
```

# Counts {#Counts}

To import count data that is stored in a separate file for each sample,
let us use the [edgeR](http://bioconductor.org/packages/edgeR)
`readDGE` method and the above `data.frame` of phenotype information:

```{r readDGE}
RG <- readDGE(
  pheno$File,
  "counts",
  c(1, 3),
  paste(pheno$Time, pheno$Infection, pheno$Status, sep = "_"),
  pheno$Sample
)
dim(RG)
```

# Gene annotations

To ensure coherence between gene annotations and gene expression data,
let us also import the `gene` records from the same GTF file that was supplied
to *featureCounts* as `GRanges` of the
[GenomicRanges](http://bioconductor.org/packages/GenomicRanges) package:

```{r import.gff}
gtfData <- import.gff(
  "expdata/genes_with_ERCC.gtf", feature.type = "gene")
names(as.data.frame(gtfData))
```

**Note:** the above code does *not* import records that relate to ERCC
spike-in features, as these are annotated with **exon** *type* (GTF column 2)
from the **ERCC** *soure* (GTF column 3).

Let us match the order of the features in the GTF file to the order of the
features imported from the count files:

```{r gtfOrder}
orderGtf <- match(rownames(RG$counts), gtfData$gene_id)
```

# `SCESet`

Let us assemble the phenotype information and count data above into
[scater](http://bioconductor.org/packages/scater) `SCESet` objects:

* one `SCESet` that contains all samples together
* one for the blank samples
* one for the bulk samples
* one for the single cells

**Note:**
This distinction will be important later, to compute and compared QC metrics
within each sets of samples.

First, let us set the `rownames` of the phenotypes information `data.frame` to
the unique sample identifier for consistency with the sample identifier
associated with the count data imported [above](#Counts):

```{r renamePheno}
rownames(pheno) <- pheno$Sample
```

Similarly, we let us also prepare an `AnnotatedDataFrame` with appropriate
`rownames` to describe the feature meta-information:

```{r}
fdata <- AnnotatedDataFrame(as.data.frame(mcols(gtfData)[orderGtf,]))
rownames(fdata) <- rownames(RG$counts)
```

We may then create the four data sets:

```{r sce.all}
pd.all <- new("AnnotatedDataFrame", data = pheno)
sce.all <- newSCESet(countData = RG$counts, phenoData = pd.all, featureData = fdata)
dim(sce.all)
```
```{r sce.blank}
idx.blank <- pheno$Status == "Blank"
pd.blank <- new("AnnotatedDataFrame", data = droplevels(pheno[idx.blank,]))
sce.blank <- newSCESet(
  countData = RG$counts[,idx.blank], phenoData = pd.blank, featureData = fdata)
dim(sce.blank)
```
```{r sce.bulk}
idx.bulk<- pheno$Status == "BULK"
pd.bulk <- new("AnnotatedDataFrame", data = droplevels(pheno[idx.bulk,]))
sce.bulk <- newSCESet(
  countData = RG$counts[,idx.bulk], phenoData = pd.bulk, featureData = fdata)
dim(sce.bulk)
```
```{r sce.sc}
idx.sc <- pheno$Status %in% c("exposed","infected","uninfected")
pd.sc <- new("AnnotatedDataFrame", data = droplevels(pheno[idx.sc,]))
sce.sc <- newSCESet(
  countData = RG$counts[,idx.sc], phenoData = pd.sc, featureData = fdata)
dim(sce.sc)
```

<!-- As HTML pages are built independently, objects need to be saved to disk
and imported as needed by the following R mardown files -->

```{r saveRDS, include=FALSE}
saveRDS(sce.all, "rds/sce.all.rds")
saveRDS(sce.blank, "rds/sce.blank.rds")
saveRDS(sce.bulk, "rds/sce.bulk.rds")
saveRDS(sce.sc, "rds/sce.sc.rds")
saveRDS(gtfData, "rds/gtfData.rds")
```

```{r loadKallisto, include=FALSE}
kallisto_all <- readRDS("../../01-kallisto/rds/all_EnsDb.rds")
```

```{r overlapKallistoAligned, include=FALSE}
fData(kallisto_all)[,"feature_id"]
features.overlap <-
  intersect(fData(kallisto_all)[,"feature_id"], featureNames(sce.all))
kallisto.overlap <- kallisto_all[
  match(features.overlap, fData(kallisto_all)$feature_id),
]
sce.overlap <- sce.all[features.overlap,]
rm(kallisto_all)
```

```{r erccIds, include=FALSE}
ercc <- grepl("^ERCC-", features.overlap)
table(ercc)
```


```{r prepareBulkPlot, include=FALSE}
comp.bulks <- data.frame(
  Time = character(),
  Infection = character(),
  Kallisto = integer(),
  HF = integer(),
  ERCC = logical()
)
for (time in c("2h", "4h", "6h")){
  for (infection in c("Mock", "D23580", "LT2")){
    message(sprintf("%s - %s", time, infection))
    kallisto <- counts(kallisto.overlap[
      ,
      kallisto.overlap$Time == time &
        kallisto.overlap$Infection == infection &
        kallisto.overlap$Status == "BULK"
    ])[,1]
    print(length(kallisto))
    hf = counts(sce.overlap)[
      ,
      sce.overlap$Time == time &
        sce.overlap$Infection == infection &
        sce.overlap$Status == "BULK"
    ]
    print(length(hf))
    new.df <- data.frame(
      Time = time,
      Infection = infection,
      Kallisto = kallisto,
      HF = hf,
      ERCC = ercc
    )
    comp.bulks <- rbind(comp.bulks, new.df)
  }
}
rm(kallisto, hf, time, infection, new.df)
```

```{r bulkPlot, include=FALSE}
ggplot(comp.bulks, aes(log10(Kallisto + 1), log10(HF + 1))) +
  geom_hex() +
  facet_grid(Infection ~ Time) +
  geom_point(
    aes(log10(Kallisto + 1), log10(HF + 1)),
    subset(comp.bulks, ERCC),
    colour = "red", size = 0.3) +
  labs(y = "HISAT2 + featureCounts") +
  scale_fill_continuous(name = "Density")
ggsave("NGS-SIG/Kallisto-HISAT2_bulk.pdf")
```

```{r prepareSCplot, include=FALSE}
set.seed(1985)
comp.sc <- data.frame(
  Time = character(),
  Infection = character(),
  Status = character(),
  Kallisto = numeric(),
  THF = numeric(),
  ERCC = logical()
)
cells.df <- data.frame(
  Cell = character(),
  Time = character(),
  Infection  = character()
)
for (time in c("2h", "4h", "6h")){
   for (infection in c("Mock", "D23580", "LT2")){
    sname <- as.character(sample(subset(
      pData(kallisto.overlap),
      Time == time & Infection == infection & Status != 'BULK',
      select = sample,
      drop = TRUE), 1))
    kallisto <- counts(kallisto.overlap[,sname])[,1]
    print(length(kallisto))
    hf = counts(sce.overlap)[,sname]
    print(length(hf))
    new.df <- data.frame(
      Time = time,
      Infection = infection,
      Status = subset(
        pData(kallisto.overlap), sample == sname, select = Status, drop = TRUE
      ),
      Kallisto = kallisto,
      HF = hf,
      ERCC = ercc
    )
    comp.sc <- rbind(comp.sc, new.df)
    newCell <- data.frame(
      Cell = sprintf("Cell %s", sname),
      Time = time,
      Infection = infection
    )
    cells.df <- rbind(cells.df, newCell)
  }
}
```

```{r SCplot, include=FALSE}
cells.df$Y = 5
cells.df$X = 0
ggplot(comp.sc, aes(log10(Kallisto + 1), log10(HF + 1))) +
  geom_hex(aes(colour = Status)) +
  facet_grid(Infection ~ Time) +
  geom_point(
    aes(log10(Kallisto + 1), log10(HF + 1)),
    subset(comp.sc, ERCC),
    colour = "red", size = 0.3) +
  ggrepel::geom_label_repel(
    aes(X, Y, label = Cell), cells.df,
    size = 2, alpha = 0.5) +
  scale_fill_continuous(name = "Density")
ggsave("NGS-SIG/Kallisto-HISAT2_SC9.pdf")
```
