---
title: "Import of RNA-Sequencing read counts and associated metadata"
---

<!-- Enable caching -->
```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(scater)
library(edgeR)
library(rtracklayer)
stopifnot(requireNamespace("DT"))
stopifnot(requireNamespace("xlsx"))
```

# Phenotype information {#phenotypes .tabset}

Phenotype information is central to all aspects of the analysis.
In the broad sense, it encompasses sample-level information beyond the
experimental design (*e.g.*, files of read counts, quality control metrics)
to coordinate statistical analyses and data visualisation
throughout downstream analyses.
Defined in the earliest steps of an analytical workflow,
it ensures traceability of sample data from files of raw data to
every output and result.

## Processing

Let us import experimental information:

```{r importPheno}
pheno <- xlsx::read.xlsx("expdata/samples.xlsx", sheetName = "pheno")
colnames(pheno)
```

To facilitate later use, let us retype some fields, reorder some factor levels,
and add an additional field `Sample` field that contains a shortened unique
identifier derived from the count file name for each sample:

```{r editPheno}
pheno$File <- as.character(pheno$File)
pheno$Infection <-
  factor(pheno$Infection, c("Mock","STM-D23580","STM-LT2","Blank"))
pheno$Status <-
  factor(pheno$Status, c("Uninfected","Exposed","Violet +","Blank","Bulk"))
pheno$Lane <- as.factor(pheno$Lane)
pheno$Sample <- gsub("WTCHG_[[:digit:]]+_", "", pheno$File) # character
pheno$sampleAnnotated <-
  with(pheno, paste(Time, Infection, Status, Sample, sep = "_")) # character
pheno$Treatment <-
  droplevels(interaction(pheno$Infection, pheno$Status, sep = "_")) # factor
pheno$Group <-
  droplevels(interaction(pheno$Time, pheno$Treatment, sep = "_")) # factor
```

## Table

```{r pheno_DT, echo=FALSE, results='asis'}
DT::datatable(
  pheno,
  options = list(pageLength = 10, searching = TRUE), filter = "top",
  rownames = FALSE
)
```

# QC metrics {#QC .tabset}

Importantly, [MultiQC](http://multiqc.info) was used to collate metrics
for each sample at each step of preprocessing---from raw reads to assigned
read counts---; those QC metrics provide key information to identify
technical issues and sample outliers.

## Processing

Let us import the QC metrics produced by [MultiQC](http://multiqc.info):

```{r importMultiQC}
multiqc <- read.delim("MultiQC/final_pipeline_data/multiqc_general_stats.txt")
colnames(multiqc)
```

Let us shorten the sample identifier used by [MultiQC](http://multiqc.info)
to a minimal unique identifier
(as done [above](#phenotypes) for phenotype information):

```{r gsub_Sample}
multiqc$Sample <- gsub("WTCHG_[[:digit:]]+_", "", multiqc$Sample)
rownames(multiqc) <- multiqc$Sample
```

Let us append those QC metrics to the phenotype information
(that will [later](#SCESet) be attached to the read counts in a `SCESet`):

```{r addPhenoMultiQC}
pheno <- merge(pheno, multiqc, by = "Sample", sort = FALSE)
```

## Table

```{r multiqc_DT, echo=FALSE}
DT::datatable(
  multiqc,
  options = list(pageLength = 10, searching = TRUE), filter = "top",
  rownames = FALSE
)
```

# Counts {#Counts}

To import count data that is stored in a separate file for each sample,
let us use the [edgeR](http://bioconductor.org/packages/edgeR)
`readDGE` method and the phenotype information prepared [above](#phenotypes):

```{r readDGE}
RG <- readDGE(
  pheno$File,
  "counts",
  c(1, 3),
  paste(pheno$Time, pheno$Infection, pheno$Status, sep = "_"),
  pheno$Sample
)
dim(RG)
```

# Gene annotations {.tabset}

To ensure coherence between gene annotations and gene expression data,
let us also import the records for genes and ERCC spike-ins molecules
from the same GTF file that was supplied to *featureCounts*,
as `GRanges` of the
[GenomicRanges](http://bioconductor.org/packages/GenomicRanges) package;
in the next section, those annotations will be attached to the corresponding
read counts in a [scater](http://bioconductor.org/packages/scater) `SCESet`
object.

First we import all the records from the composite GFF file that includes
both endogenous gene features and ERCC spike-in molecules:

```{r import.gff}
gtfData <- import.gff("expdata/genes_with_ERCC.gtf")
colnames(as.data.frame(gtfData))
```

We may then extract records associated with endogenous gene features:

```{r gtfGenes}
gtf.genes <- subset(gtfData, type == "gene"); length(gtf.genes)
```

Similarly, records associated with ERCC spike-in molecules:

```{r gtf.ERCC}
gtf.ERCC <- subset(gtfData, source == "ERCC"); length(gtf.ERCC)
```

Finally, let us combine the two types of feature annotations,
and reorder them to ensure that they match the order of features imported
[above](#Counts):

```{r gtf.final}
gtf.final <- c(gtf.genes, gtf.ERCC)
stopifnot(all(rownames(RG) %in% gtf.final$gene_id))
gtf.order <- match(rownames(RG), gtf.final$gene_id)
gtf.final <- gtf.final[gtf.order]
```

# SCESet {#SCESet .tabset}

Let us assemble the count data, phenotype information, and feature information
prepared above into [scater](http://bioconductor.org/packages/scater)
`SCESet` objects:

* one `SCESet` that contains all samples together
* one for the blank samples
* one for the bulk samples
* one for the single cells

**Note:**
This distinction will be important later, to compute and compared QC metrics
within each sets of samples.

First, let us set the `rownames` of the phenotypes information `data.frame` to
the unique sample identifier for consistency with the sample identifier
associated with the count data imported [above](#Counts):

```{r renamePheno}
rownames(pheno) <- pheno$Sample
```

Similarly, we let us also prepare an `AnnotatedDataFrame` with appropriate
`rownames` to describe the feature meta-information:

```{r fdata}
fdata <- AnnotatedDataFrame(as.data.frame(gtfData))
rownames(fdata) <- fdata$gene_id
```

We may then create the four data sets:

## All samples

```{r sce.all}
pd.all <- AnnotatedDataFrame(data = pheno)
sce.all <- newSCESet(countData = RG$counts, phenoData = pd.all, featureData = fdata)
dim(sce.all)
```

## Blanks

```{r sce.blank}
idx.blank <- pheno$Status == "Blank"
pd.blank <- AnnotatedDataFrame(droplevels(pheno[idx.blank,]))
sce.blank <- newSCESet(
  countData = RG$counts[,idx.blank], phenoData = pd.blank, featureData = fdata)
dim(sce.blank)
```

## Bulks

```{r sce.bulk}
idx.bulk<- pheno$Status == "Bulk"
pd.bulk <- AnnotatedDataFrame(droplevels(pheno[idx.bulk,]))
sce.bulk <- newSCESet(
  countData = RG$counts[,idx.bulk], phenoData = pd.bulk, featureData = fdata)
dim(sce.bulk)
```

## Single cells

```{r sce.sc}
idx.sc <- pheno$Status %in% c("Exposed", "Violet +", "Uninfected")
pd.sc <- AnnotatedDataFrame(droplevels(pheno[idx.sc,]))
sce.sc <- newSCESet(
  countData = RG$counts[,idx.sc], phenoData = pd.sc, featureData = fdata)
dim(sce.sc)
```

<!-- As HTML pages are built independently, objects need to be saved to disk
and imported as needed by the following R mardown files -->

```{r saveRDS, include=FALSE}
saveRDS(sce.all, "rds/sce.all.rds")
saveRDS(sce.blank, "rds/sce.blank.rds")
saveRDS(sce.bulk, "rds/sce.bulk.rds")
saveRDS(sce.sc, "rds/sce.sc.rds")
saveRDS(gtfData, "rds/gtfData.rds")
```
