---
title: "Sample-level quality control"
---

```{r checkPkgs, child="_checkLibraries.Rmd", include=FALSE}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(Biobase)
library(DT)
library(ggplot2)
library(scater)
library(rtracklayer)
library(ggrepel)
library(RColorBrewer)
library(ComplexHeatmap)
library(dplyr)
sce.all <- readRDS("rds/sce.all.rds")
idx.sc <- sce.all$Status %in% c("exposed","infected","uninfected")
sce.blank <- readRDS("rds/sce.blank.rds")
sce.sc <- readRDS("rds/sce.sc.rds")
gtfData <- readRDS("rds/gtfData.rds")
```

# _MultiQC_
## Import QC metrics

Importantly, [MultiQC](http://multiqc.info) was used to collate metrics
for each sample at each step of preprocessing---from raw reads to assignment
counts---; those QC metrics already provide an idea of technical issues
in certain samples.

Let us import the QC metrics produced by [MultiQC](http://multiqc.info):

```{r importMultiQC}
multiqc <- read.delim("MultiQC/final_pipeline_data/multiqc_general_stats.txt")
names(multiqc)
```

Let us shorten the sample identifier used by [MultiQC](http://multiqc.info)
to a minimal unique identifier (identically to the phenotype information in
[Data imports](02_import.html)):

```{r gsub_Sample}
multiqc$Sample <- gsub("WTCHG_[[:digit:]]+_", "", multiqc$Sample)
```

Let us annotated those QC metrics with the phenotype information:

```{r addPhenoMultiQC}
multiqc <- merge(pData(sce.all), multiqc, by = "Sample", sort = FALSE)
```

Let us confirm that the samples are still ordered identically to the `sce.all`
object:

```{r checkOrder}
all(multiqc$Sample == rownames(pData(sce.all)))
```

## Assigned read pairs / Library size (*featureCounts*) {#featureCounts}

An excessively small library size may indicate technical issues during
library preparation or low quality samples.
Moreover, library size excessively different from the
majority of samples may introduce bias due to under- or over-sampling of
RNA fragments.

Let us identify as outliers single cells
for which the number of assigned read pairs (*i.e.* library size)
is further than `3` MADs (median absolute deviations) from the median
of *single cell* samples:

```{r multiQC_outliers_assigned}
metric <- log10(multiqc$featureCounts_Assigned[idx.sc])
multiqc.assigned.M <- median(metric)
multiqc.assigned.MAD <- mad(metric)
multiqc.assigned.outliers <-
  abs(metric - multiqc.assigned.M) > 3 * multiqc.assigned.MAD
table(multiqc.assigned.outliers)
```

For the record, the lower cut-off value is:

```{r assigned_lowerCutOff}
10^(multiqc.assigned.M - 3 * multiqc.assigned.MAD)
```

Let us visualise those outliers, along with blank and bulk samples
for reference:

```{r multiQC_ggplot_assigned}
ggplot(multiqc, aes(Sample, featureCounts_Assigned)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  ggrepel::geom_text_repel(
    aes(label = Sample), subset(multiqc[idx.sc,], multiqc.assigned.outliers),
    alpha = 0.4
  ) +
  geom_hline(
    yintercept = 10^(multiqc.assigned.M + 3*c(1,-1)*multiqc.assigned.MAD),
    colour = "red"
  ) +
  geom_hline(yintercept = 10^(multiqc.assigned.M)) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical"
  )
```

## GC content (*FastQC*)

GC content is species-specific; it can be estimated from the annotated transcriptome or the sequenced data. Therefore, GC content excessively
different from the majority of samples can indicate issues during
library preparation or contamination.

Let us identify as outliers samples for which GC content is
further than 3 MADs (median absolute deviations) from the median
of *single cell* samples:

```{r multiQC_outliers_GC}
metric <- multiqc$FastQC_percent_gc[idx.sc]
multiqc.GC.M <- median(metric)
multiqc.GC.MAD <- mad(metric)
multiqc.GC.outliers <-
  abs(metric - multiqc.GC.M) > 3 * multiqc.GC.MAD
table(multiqc.GC.outliers)
```

Let us visualise the outlier, along with blank and bulk samples
for reference:

```{r multiQC_ggplot_GC}
ggplot(multiqc, aes(Sample, FastQC_percent_gc)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  ggrepel::geom_text_repel(
    aes(label = Sample), subset(multiqc[idx.sc,], multiqc.GC.outliers),
    alpha = 0.4
  ) +
  geom_hline(
    yintercept = multiqc.GC.M + 3*c(1,-1)*multiqc.GC.MAD,
    colour = "red"
  ) +
  geom_hline(yintercept = multiqc.GC.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical"
  )
```

## Alignment rate (*HISAT2* / *bowtie2*)

Technical issues introduced during library preparation may also result
in excessively *low* alignment rate.
Nevertheless, libraries with slightly lower alignment rates may still yield
a library size comparable to the majority of samples.
For that reason, let us apply a slightly more relaxed cut-off of `5` MADs
to identify as outliers *single cells* with excessively *low* alignment rate:

```{r multiQC_outliers_align}
metric <- multiqc$Bowtie.2_overall_alignment_rate[idx.sc]
multiqc.aligned.M <- median(metric)
multiqc.aligned.MAD <- mad(metric)
multiqc.aligned.outliers <-
  metric < multiqc.aligned.M - 5 * multiqc.aligned.MAD
table(multiqc.aligned.outliers)
```

Let us visualise the outlier, along with blank and bulk samples
for reference:

```{r multiQC_ggplot_align}
ggplot(multiqc, aes(Sample, Bowtie.2_overall_alignment_rate)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  ggrepel::geom_text_repel(
    aes(label = Sample), subset(multiqc[idx.sc,], multiqc.aligned.outliers),
    alpha = 0.4
  ) +
  geom_hline(
    yintercept = multiqc.aligned.M + 5*c(1,-1)*multiqc.aligned.MAD,
    colour = "red",
    linetype = 2:1
  ) +
  geom_hline(yintercept = multiqc.aligned.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical"
  ) +
  ylab("HISAT.2_overall_alignment_rate")
```

## Percentage assigned (*featureCounts*)

After alignment, unusually *low* assignment rate of reads to annotated genomic
features may be indicative of technical issue; for instance during
poly-A selection.

Let us identify as outliers *single cells* with excessively *low*
assignment rate:

```{r multiQC_outliers_assign}
metric <- multiqc$featureCounts_percent_assigned[idx.sc]
multiqc.assignedPerc.M <- median(metric)
multiqc.assignedPerc.MAD <- mad(metric)
multiqc.assignedPerc.outliers <-
  metric < multiqc.assignedPerc.M - 3 * multiqc.assignedPerc.MAD
table(multiqc.assigned.outliers)
```

Let us visualise the outlier, along with blank and bulk samples
for reference:

```{r multiQC_ggplot_assign}
ggplot(multiqc, aes(Sample, featureCounts_percent_assigned)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  ggrepel::geom_text_repel(
    aes(label = Sample), subset(multiqc[idx.sc,], multiqc.assignedPerc.outliers),
    alpha = 0.4
  ) +
  geom_hline(
    yintercept = multiqc.assignedPerc.M + 3*c(1,-1)*multiqc.assignedPerc.MAD,
    colour = "red",
    linetype = 2:1
  ) +
  geom_hline(yintercept = multiqc.assignedPerc.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical"
  )
```

## *MultiQC* summary

```{r multiQC_summary}
outlier.vars <- ls(pattern = "outliers")
colSums(sapply(outlier.vars, function(x){get(x)}))
table(rowSums(sapply(outlier.vars, function(x){get(x)})))
```

# Blanks
## Highest expressed features

Blank samples are designed to detected technical issues during library
preparation; in particular contamination, as only ERCC spike-in features are
expected in those samples.

First, let us obtain the identifier of ERCC spike-in features:

```{r ERCC_ids}
ERCCs <- grep("^ERCC-[[:digit:]]+$", featureNames(sce.blank), value = TRUE)
```

The [scater](http://bioconductor.org/packages/scater) `calculateQCMetrics`
method may then be used to calculate the relevant QC metrics:

```{r blank_QC}
sce.blank <- calculateQCMetrics(sce.blank, list(ERCC = ERCCs))
```

In the `SCESet` object augmnented by QC metrics, let us visualise the
most abundant features which displays the expected over-representation of
ERCC spike-in features:

```{r blank_highExprs}
plotQC(sce.blank, type = "highest-expression")
```

# *scater*
## Calculation of QC metrics

The following sections require the estimation of several QC metrics that
leverage control features and control samples, as means to identify
outliers among single cells.

Let us extract from those annotations the identifier of mitochondrial gene
features; those will be used in a [later section](#MTcontent):

```{r MT_ids}
MTs <- mcols(gtfData)[seqnames(gtfData) == "MT", "gene_id"]
```

Let us apply the [scater](http://bioconductor.org/packages/scater)
`calculateQCMetrics` to calculate the relevant QC metrics in the *single cell*
data set:

```{r sc_QC}
sce.sc <- calculateQCMetrics(
  sce.sc,
  feature_controls = list(ERCC = ERCCs, MT = MTs)
)
```

For reference purpose only, let us also calculate
the relevant QC metrics in the complete data set:

```{r all_QC}
sce.all <- calculateQCMetrics(
  sce.all,
  feature_controls = list(ERCC = ERCCs, MT = MTs),
  cell_controls = list(
    Bulk = which(sce.all$Status == "BULK"),
    Blank = which(sce.all$Status == "Blank")
  )
)
```

## Library size

The *library size* refers to the total count of reads assigned to 
annotated features in each individua sample.
As an internal validation, let us verify that identifying outliers
using the counts imported into the `SCESet` produce the same results as
the [earlier](#featureCounts) use of [MultiQC](http://multiqc.info) reports
of read pairs assignment by
[featureCounts](http://bioinf.wehi.edu.au/featureCounts).

First let us identify as outliers single cells
for which the library size
is further than `3` MADs (median absolute deviations) from the median
of *single cell* samples:

```{r scater_outliers_totalCounts}
metric <- sce.sc$log10_total_counts
scater.totalCounts.M <- median(metric)
scater.totalCounts.MAD <- mad(metric)
scater.totalCounts.outliers <-
  abs(metric - scater.totalCounts.M) > 3 * scater.totalCounts.MAD
table(scater.totalCounts.outliers)
```

In particular, the lower cut-off value is:

```{r totalCounts_lowerCutOff}
10^(scater.totalCounts.M - 3 * scater.totalCounts.MAD)
```

Let us visualise those outliers, along with blank and bulk samples
for reference:

```{r scater_ggplot_totalCounts}
ggplot(pData(sce.all), aes(Sample, log10_total_counts)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  geom_text_repel(
    aes(label = Sample), subset(pData(sce.sc), scater.totalCounts.outliers),
    alpha = 0.4
  ) +
  geom_hline(
    yintercept = scater.totalCounts.M + 3*c(1,-1)*scater.totalCounts.MAD,
    colour = "red"
  ) +
  geom_hline(yintercept = scater.totalCounts.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical"
  )
```

## Library complexity {#libraryComplexity}

*Library complexity* refers to the proportion of assigned reads accounted for
by the *N* most abundant features within each individual sample, with values
for *N* usually ranging from tens to hundreds.
It is expected to observe a lower library complexity in single cells relative
to bulk samples, due to higher dropout events in single cells, and the union of
multiple phenotypes within bulk samples.
In contrast, blank samples are expected to have extremely low library
complexity, as only `92` ERCC spike-ins sequences are expected in those
libraries, in contrast to thousands of genomic features in biological samples.

In the `SCESet` object augmnented by QC metrics, let us visualise the
library complexity profile against experimental phenotype information:

```{r all_libraryComplexity}
plot(
  sce.all, block1 = "Status", block2 = "Infection",
  colour_by = "Time", nfeatures = 300, exprs_values = "counts"
)
```

First, from a QC perspective:

* Blanks samples show a very low library complexity, as expected.
* Bulk samples show a generally high library complexity, as expected.
* Single cells generally show a complexity profile comparable to that of bulk
  samples.
* A small subset of single cells (*e.g.*, `exposed:D23580:6h`,`exposed:LT2:4h`, 
  `infected:D23580:4h`, `infected:D23580:6h`, `infected:LT2:6h`) show
  suspiciously low library complexity.
  Those single cells must therefore be treated as outliers, and be removed from
  the data set prior to feature selection, normalisation, and subsequent
  analyses.

Now, from a biological perspective, an interesting observation may be made:

* Library complexity seems to decrease with `Time` in `exposed` and `infected`
  samples in cells exposed to or infected by bacteria, in contrast to
  control uninfected cells.
  This *may* indicate a polarisation of single cells toward
  specialised phenotype(s) after infection or exposure.
  + `Mock` (*i.e.*, `uninfected`) single cells tend to show a consistent,
    relatively high library complexity
  + Both `LT2` and `D23580` infections show a gradual decrease of library
    complexity over time. This observation *might* be more pronounced
    for `D23580`.

## Most abundant features {.tabset}

Although abundantly expressed features are an expected aspect of cell biology
(*e.g.*, myosin in muscle cells, clusters of differentiation in immune cells)
which may contribute to the identification and classification of cell types, overly abundant feature in RNA-Seq libraries
may indicate issues during library preparation; for instance, cDNA
amplification bias.
Moreover, when the most abundant features account for an excessive proportion
of the library, expression levels for the majority of genes may not be
estimated with sufficient accuracy.

Let us identify as outliers single cells
for which the `50`, `100`, and `200` most abundant features account for an
excessively *high* proportion of the library size:

```{r scater_outliers_top50}
metric <- sce.sc$pct_counts_top_50_features
scater.top50.M <- median(metric)
scater.top50.MAD <- mad(metric)
scater.top50.outliers <-
  metric > scater.top50.M + 3 * scater.top50.MAD
table(scater.top50.outliers)
```

```{r scater_outliers_top100}
metric <- sce.sc$pct_counts_top_100_features
scater.top100.M <- median(metric)
scater.top100.MAD <- mad(metric)
scater.top100.outliers <-
  metric > scater.top100.M + 3 * scater.top100.MAD
table(scater.top100.outliers)
```

```{r scater_outliers_top200}
metric <- sce.sc$pct_counts_top_200_features
scater.top200.M <- median(metric)
scater.top200.MAD <- mad(metric)
scater.top200.outliers <-
  metric > scater.top200.M + 3 * scater.top200.MAD
table(scater.top200.outliers)
```

Let us visualise those outliers, along with blank and bulk samples
for reference:

### 50

```{r scater_ggplot_top50}
ggplot(pData(sce.all), aes(Sample, pct_counts_top_50_features)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  geom_text_repel(
    aes(label = Sample), subset(pData(sce.sc), scater.top50.outliers),
    alpha = 0.4) +
  geom_hline(
    yintercept = scater.top50.M + 3 * c(-1,1) * scater.top50.MAD,
    colour = "red", linetype = 2:1) +
  geom_hline(yintercept = scater.top50.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical")
```

### 100

```{r scater_ggplot_top100}
ggplot(pData(sce.all), aes(Sample, pct_counts_top_100_features)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  geom_text_repel(
    aes(label = Sample), subset(pData(sce.sc), scater.top100.outliers),
    alpha = 0.4) +
  geom_hline(
    yintercept = scater.top100.M + 3 * c(-1,1) * scater.top100.MAD,
    colour = "red", linetype = 2:1) +
  geom_hline(yintercept = scater.top100.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical")
```

### 200

```{r scater_ggplot_top200}
ggplot(pData(sce.all), aes(Sample, pct_counts_top_200_features)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  geom_text_repel(
    aes(label = Sample), subset(pData(sce.sc), scater.top200.outliers),
    alpha = 0.4) +
  geom_hline(
    yintercept = scater.top200.M + 3 * c(-1,1) * scater.top200.MAD,
    colour = "red", linetype = 2:1) +
  geom_hline(yintercept = scater.top200.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical")
```

In this last figure (*i.e.*, `200` most abundant features),
note how two `infected:D23580` samples appear near the cut-off threshold.
This observation can be related to the overall library complexity figure
shown [earlier](#libraryComplexity).

## Count of detected features

A notion closely related to library complexity is the count of features
detected within each individual sample. Although the definition of "detection
level" may be debated, we will accept here any feature with non-zero count.

Abnormal counts of detected features may be indicative of technical issue
during RNA capture and library preparation; for instance,
excessively *low* counts of detected features may indicate RNA degradation,
while excessively *high* counts of detected features may indicate capture of
multiple cells.

Due to the considerable amount of variation that results from the relaxed
definition of detection level above, let us apply a somewhat stringent cut-off
of `2` MADs to identify as outliers *single cells* with excessively *low*
counts of detected features:

```{r scater_outliers_features}
metric <- sce.sc$total_features
scater.features.M <- median(metric)
scater.features.MAD <- mad(metric)
scater.features.outliers <-
  metric < scater.features.M - 2 * scater.features.MAD
table(scater.features.outliers)
```

Let us visualise those outliers, along with blank and bulk samples
for reference:

```{r scater_ggplot_features}
ggplot(pData(sce.all), aes(Sample, total_features)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  geom_text_repel(
    aes(label = Sample), subset(pData(sce.sc), scater.features.outliers),
    alpha = 0.4
  ) +
  geom_hline(
    yintercept = scater.features.M + 2 * c(-1,1) * scater.features.MAD,
    colour = "red", linetype = 1:2
  ) +
  geom_hline(yintercept = scater.features.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical"
  )
```

## ERCC content {#ERCCoutliers}

An identical quantity of ERCC spike-in features is added to each sample
prior to library preparation. As a result, technical variation and issues
during library preparation may be estimated from the proportion of reads
assigned to those spike-in features; an excessive proportion of reads
assigned to ERCC spike-in features may indicate eukaryotic RNA degradation,
while excessive variation may indicate cDNA amplication bias.

Due to the limited amount of variation that results from the very low levels
of ERCC spike-in features generally detected,
let us apply a somewhat relaxed cut-off of `5` MADs
to identify as outliers *single cells* with excessively *high* ERCC content:

```{r scater_outliers_ercc}
metric <- sce.sc$pct_counts_feature_controls_ERCC
scater.ERCC.M <- median(metric)
scater.ERCC.MAD <- mad(metric)
scater.ERCC.outliers <- metric > scater.ERCC.M + 5 * scater.ERCC.MAD
table(scater.ERCC.outliers)
```

Let us visualise those outliers, along with blank and bulk samples
for reference:

```{r scater_ggplot_ERCC}
ggplot(pData(sce.all), aes(Sample, pct_counts_feature_controls_ERCC)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  geom_text_repel(
    aes(label = Sample), subset(pData(sce.sc), scater.ERCC.outliers),
    alpha = 0.6
  ) +
  geom_hline(
    yintercept = scater.ERCC.M + 5 * c(-1,1) * scater.ERCC.MAD,
    colour = "red", linetype = 2:1
  ) +
  geom_hline(yintercept = scater.ERCC.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical"
  )
```

## Mitochondrial content {#MTcontent}

Similarly to the ERCC spike-in features, mitochondrial RNA may be used
as an endogenous source of control features.
Albeit mitochondrial gene expression levels may be altered by physiological
processes, the limited count of genes encoded on the mitochondrial chromosome
implies a similarly limited proportion of reads assigned to those gene
features.
Consequently, abnormal proportions of reads assigned to mitochondrial gene
features may be indicative of technical issue during RNA capture and
library preparation, or abnormal count of mitochondria within the single cells.

Let us identify as outliers single cells for which the proportion of reads
assigned to mitochondrial gene features is is further than `3` MADs
from the median of *single cell* samples:

```{r scater_outliers_MT}
metric <- sce.sc$pct_counts_feature_controls_MT
scater.MT.M <- median(metric)
scater.MT.MAD <- mad(metric)
scater.MT.outliers <-
  metric > scater.MT.M + 5 * scater.MT.MAD
table(scater.MT.outliers)
```

Let us visualise those outliers, along with blank and bulk samples
for reference:

```{r scater_ggplot_MT}
ggplot(pData(sce.all), aes(Sample, pct_counts_feature_controls_MT)) +
  geom_point(aes(colour = Status, shape = Infection)) +
  geom_text_repel(
    aes(label = Sample), subset(pData(sce.sc), scater.MT.outliers),
    alpha = 0.4
  ) +
  geom_hline(
    yintercept = scater.MT.M + 5 * c(-1,1) * scater.MT.MAD,
    colour = "red", linetype = 2:1
  ) +
  geom_hline(yintercept = scater.MT.M) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical"
  )
```

# Global summary of QC metrics
## Collection of QC filters

First, let us collect results of the ``r length(ls(pattern="\\.outliers$"))``
QC filters described in the previous sections into a summary `data.frame`:

```{r QC_collectOutliers}
outlier.vars <- ls(pattern = "\\.outliers$")
outliers.df <- data.frame(lapply(
  outlier.vars,
  function(x){ifelse(get(x), "Fail", "Pass")}
))
colnames(outliers.df) <- gsub("\\.outliers$", "", outlier.vars)
rownames(outliers.df) <- sampleNames(sce.sc)
```

Let us also define an order of sample that groups single cells by
experimental phenotypes:

```{r sampleSupervisedOrder}
sample.order.supervised <- order(sce.sc$Time, sce.sc$Status, sce.sc$Infection)
```

## Counts of outliers across filters

First, let us examine the count of *single cells* that fail each of the QC
metrics described in the previous sections:

```{r QC_countByMetric}
colSums(outliers.df == "Fail")
```

Secondly, we may also examine the count of *single cells*, according to the
number of QC filters that they have failed:

```{r QC_countByFailed}
table(rowSums(outliers.df == "Fail"))
```

Let us define the single cells that did not fail any of the QC filter:

```{r QC_finalOutliers}
idx.pass <- rowSums(outliers.df == "Fail") == 0
table(idx.pass)
```

## Heat map summary view

Let us visually summarise *single cells* that pass each filter as a heat map.

First, let us define colours used to indicate phenotype information:

```{r palette9}
col9 <- brewer.pal(10, "Set3")[c(1:8, 10)]
```

We may then display the QC metrics failed by each single cell as red cells
in an heat map:

```{r QC_heatmap, fig.height=7}
hm_row <- rowAnnotation(
  df = pData(sce.sc)[sample.order.supervised,c("Time", "Status", "Infection")],
  col = list(
    Time = c("2h" = col9[1], "4h" = col9[2], "6h" = col9[3]),
    Status=c("uninfected"=col9[4],"exposed"=col9[5],"infected"=col9[6]),
    Infection = c("Mock" = col9[7], "LT2" = col9[8], "D23580" = col9[9])
  )
)
hm <- Heatmap(
  as.matrix(outliers.df)[sample.order.supervised,],
  name = "Filter",
  col = c("Pass" = "lightblue", "Fail" = "red"),
  cluster_rows = FALSE, show_row_names = FALSE
)
draw(hm + hm_row)
```

## Phenotype information of single cell outliers

We may summarise experimental phenotype information for the outliers samples:

```{r QC_summary}
summary(pData(sce.sc)[
  !idx.pass,c("Infection","Status","Time","Lane","Plate","Well")
])
```

Or alternatively display the full list of outliers in a table:

```{r QC_tableOutliers}
datatable(
  arrange(
    cbind(
      pData(sce.sc)[
        !idx.pass,c("Infection","Status","Time","Lane","Plate","Well")
      ],
      Filters = rowSums(outliers.df == "Fail")[!idx.pass]
    ),
    desc(Filters)
  ),
  options = list(pageLength = 10, searching = TRUE), filter = "top",
  rownames = FALSE
)
```

# Single cells selection

Let us select the single cells that passed all QC metrics into a new `SCESet`:

```{r subset_goodCells}
sce.pass <- sce.sc[,idx.pass]
```

Let us finally examine the phenotype information of the selected single cells:

```{r goodCells_summary}
summary(pData(sce.pass)[
  !idx.pass,c("Infection","Status","Time","Lane","Plate","Well")
])
```


<!-- As HTML pages are built independently, objects need to be saved to disk
and imported as needed by the other R mardown files -->

```{r save_sce.pass, include=FALSE}
saveRDS(sce.pass, "rds/sce.pass.rds")
```
