---
title: "Sample-level quality control"
bibliography:
  bibtex.bib
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(cache = TRUE)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(RColorBrewer)
library(scater)
library(SummarizedExperiment)
library(ComplexHeatmap)
stopifnot(requireNamespace("DT"))
sce.all <- readRDS("rds/sce.all.rds")
sce.blank <- readRDS("rds/sce.blank.rds")
sce.sc <- readRDS("rds/sce.sc.rds")
```

# Plotting themes

First of all, let us define:

* theme elements used throughout various figures in the following sections:

```{r sampleQCtheme}
sampleQCtheme <- theme_minimal() + theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    panel.grid.major.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical")
```

* a set of colours used to indicate phenotype information consistently
  in the upcoming sections:

```{r col.set3}
col.set3 <- brewer.pal(12, "Set3")
col.status<-col.set3[c(3,5:6,7,4)];names(col.status)<-levels(sce.all$Status)
col.infection<-col.set3[c(3,8,10)];names(col.infection)<-levels(sce.all$Infection)[1:3]
col.time<-col.set3[c(2,7:8)];names(col.time)<-levels(sce.all$Time)[1:3]
```

* let us also initialise a `data.frame` that will be used to store outliers
  identified in each of the quality control filters defined below:

```{r outliers}
outliers <- data.frame(row.names = sampleNames(sce.sc))
```

# Blanks {#blanks}

Blank samples are designed to detect major technical issues during library
preparation; in particular contamination, as only ERCC spike-in features are
expected in those samples.

First, let us obtain the identifier of ERCC spike-in features:

```{r ERCC_ids}
ERCCs <- grep("^ERCC-[[:digit:]]+$", featureNames(sce.blank), value = TRUE)
```

The [scater](http://bioconductor.org/packages/scater) `calculateQCMetrics`
method may then be used to calculate the relevant QC metrics:

```{r blank_QC}
sce.blank <- calculateQCMetrics(sce.blank, list(ERCC = ERCCs))
setSpike(sce.blank) <- "ERCC"
```

In the `SCESet` object augmnented by QC metrics, let us visualise the
most abundant features which displays the expected over-representation of
ERCC spike-in features:

```{r blank_highExprs, fig.height=5}
plotQC(sce.blank, type = "highest-expression")
```

# FastQC / MultiQC

## GC content (*FastQC*)

GC content is species-specific; it can be estimated from the annotated transcriptome or the sequenced data. Therefore, GC content excessively
different from the majority of samples can indicate issues during
library preparation or contamination.

Let us identify as outliers single cells that display a GC content
excessively different from the median of *single cell* samples,
here beyond **3** MADs:

```{r GC}
metric <- sce.sc$FastQC_percent_gc
GC.M <- median(metric)
GC.MAD <- mad(metric)
outliers$GC <- (abs(metric - GC.M) > 3 * GC.MAD)
table(outliers$GC)
```

Let us visualise outliers, along with blank and bulk samples for reference:

```{r plot.GC, echo=FALSE, fig.height=4}
plot.GC <- ggplot(pData(sce.all), aes(Sample, FastQC_percent_gc)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$GC,], alpha = 0.6,
    min.segment.length = unit(0, "lines")) +
  geom_hline(yintercept = GC.M + 3 * c(1,-1) * GC.MAD, colour="red",alpha=0.5) +
  geom_hline(yintercept = GC.M, alpha=0.5) +
  labs(y = "% GC (FastQC)", x = "Samples") +
  scale_y_continuous(limits = c(0,100)) +
  scale_colour_manual(values = col.status) +
  sampleQCtheme
plot.GC
ggsave(
  "03_out/GC.pdf",
  plot.GC + guides(colour = "none", shape = "none"), width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.0f%%", GC.M)`
* cut-offs
    - upper bound: `r sprintf("%.0f%%", GC.M+GC.MAD*3)`
    - lower bound: `r sprintf("%.0f%%", GC.M-GC.MAD*3)`

## Alignment rate (*HISAT2*)

Technical issues introduced during library preparation may result
in an excessively *low* alignment rate.
Nevertheless, libraries with slightly lower alignment rates may still yield
a library size comparable to the majority of samples.

For those reasons reasons,
let us identify as outliers single cells with an alignment rate
excessively *lower* than the median of *single cell* samples,
here beyond **3** MADs:

```{r AlignedPercent}
metric <- sce.sc$Bowtie.2_overall_alignment_rate
align.M <- median(metric)
align.MAD <- mad(metric)
outliers$AlignedPercent <- (metric < align.M - 3 * align.MAD)
table(outliers$AlignedPercent)
```

Let us visualise outliers, along with blank and bulk samples for reference:

```{r plot.aligned, echo=FALSE, fig.height=4}
plot.aligned <-
  ggplot(pData(sce.all), aes(Sample, Bowtie.2_overall_alignment_rate)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$AlignedPercent,], alpha = 0.6,
    min.segment.length = unit(0, "lines")) +
  scale_y_continuous(limits = c(0,100)) +
  scale_colour_manual(values = col.status) +
  geom_hline(yintercept = align.M, alpha = 0.5) +
  geom_hline(
    yintercept = align.M + 3*c(1,-1)*align.MAD, colour="red", linetype=2:1,
    alpha = 0.5) +
  labs(y = "Overall alignment rate (HISAT2)", x = "Samples") + sampleQCtheme
plot.aligned
ggsave(
  "03_out/alignment_rate.pdf",
  plot.aligned + guides(colour = "none", shape = "none"), width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.0f%%", align.M)`
* cut-off
    - lower bound: `r sprintf("%.0f%%", align.M-align.MAD*3)`

```{r posterAligned, include=FALSE}
posterAligned <-
  ggplot(pData(sce.all), aes(Sample, Bowtie.2_overall_alignment_rate)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$Aligned,], alpha = 0.4,
    min.segment.length = unit(0, "lines")) +
  scale_color_manual(values = col.status) +
  geom_hline(
    yintercept = align.M + 3*c(1,-1)*align.MAD, colour="red", linetype=2:1) +
  scale_y_continuous(limits = c(0,100)) +
  geom_hline(yintercept = align.M) + ylab("Overall alignment rate (HISAT2)") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
ggsave("poster/alignedQC.pdf", posterAligned, width = 9, height = 6)
```

## Assigned reads / Library size (*featureCounts*) {#assigned}

An excessively small count of aligned reads assigned to 
annotated genomics features (*i.e.*, library size) may indicate
technical issues during library preparation or low quality samples.
Moreover, library size excessively different from the
majority of samples may introduce
bias due to under- or over-sampling of RNA fragments, and negatively affect
the performance of between-sample normalisation algorithms.

Let us identify as outliers single cells
with a count of assigned read pairs (*i.e.* library size)
excessively different from the median of *single cell* samples,
here beyond **3** MADs:

```{r AssignedCount}
metric <- log10(sce.sc$featureCounts_Assigned)
cAssigned.M <- median(metric)
cAssigned.MAD <- mad(metric)
outliers$AssignedCount <- (abs(metric - cAssigned.M) > 3 * cAssigned.MAD)
table(outliers$AssignedCount)
```

Let us visualise outliers, along with blank and bulk samples, for reference:

```{r plot.cAssigned, echo=FALSE, fig.height=4}
plot.cAssigned <-
  ggplot(pData(sce.all), aes(Sample, featureCounts_Assigned)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$AssignedCount,], alpha = 0.6,
    min.segment.length = unit(0, "lines")
  ) + geom_hline(
    yintercept = 10^(cAssigned.M + 3*c(1,-1)*cAssigned.MAD),
    colour = "red", alpha = 0.5) +
  scale_color_manual(values = col.status) +
  geom_hline(yintercept = 10^(cAssigned.M), alpha = 0.5) +
  labs(y = "Count assigned (featureCounts)", x = "Samples") + sampleQCtheme
plot.cAssigned
ggsave(
  "03_out/assigned_count.pdf",
  plot.cAssigned + guides(colour = "none", shape = "none"), width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.1fe+06", 10^cAssigned.M/1E6)`
* cut-offs
    - lower bound:
      `r sprintf("%.1fe+06", 10^(cAssigned.M-cAssigned.MAD*3)/1E6)`
    - upper bound:
      `r sprintf("%.1fe+06", 10^(cAssigned.M+cAssigned.MAD*3)/1E6)`

## Percentage assigned (*featureCounts*)

Similarly, unusually *low* assignment rate of reads to annotated genomic
features may indicate technical issue; for instance during
poly-A selection.

Let us identify as outliers single cells
with a rate of assigned reads
excessively *lower* than the median of *single cell* samples,
here beyond **3** MADs:

```{r AssignedPercent}
metric <- sce.sc$featureCounts_percent_assigned
pAssigned.M <- median(metric)
pAssigned.MAD <- mad(metric)
outliers$AssignedPercent <- (metric < pAssigned.M - 3 * pAssigned.MAD)
table(outliers$AssignedPercent)
```

Let us visualise outliers, along with blank and bulk samples, for reference:

```{r plot.pAssigned, echo=FALSE, fig.height=4}
plot.pAssigned <-
  ggplot(pData(sce.all), aes(Sample, featureCounts_percent_assigned)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$AssignedPercent,], alpha = 0.6,
    min.segment.length = unit(0, "lines")
  ) + scale_y_continuous(limits = c(0,100)) +
  geom_hline(
    yintercept = (pAssigned.M + 3*c(1,-1)*pAssigned.MAD), colour = "red",
    linetype = 2:1, alpha = 0.5) +
  geom_hline(yintercept = pAssigned.M, alpha = 0.5) +
  scale_colour_manual(values = col.status) +
  labs(y = "% assigned (featureCounts)", x = "Samples") + sampleQCtheme
plot.pAssigned
ggsave(
  "03_out/assigned_rate.pdf",
  plot.pAssigned + guides(colour = "none", shape = "none"), width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.0f%%", pAssigned.M)`
* cut-off
    - lower bound:
      `r sprintf("%.0f%%", pAssigned.M-pAssigned.MAD*3)`

## Summary

Let us display the count of cells identified as outliers for each of the
filters so far:

```{r fastqc.fr}
fastqc.fr <- FilterRules(colnames(outliers))
fastqc.fr
```

```{r fastqc.fr_table, results='asis', echo=FALSE}
pander::pandoc.table(
  data.frame(
    Fail = summary(evalSeparately(fastqc.fr, outliers)),
    Pass = summary(!evalSeparately(fastqc.fr, outliers))
))
```

# *scater*

## Calculation of QC metrics {#calculateQCMetrics}

The following sections require the estimation of several QC metrics that
leverage control features and control samples,
as means to identify outliers among single cells.

Let us extract from the `featureData` slot of the `SCESet` object,
the identifier of mitochondrial gene features;
those will be used for quality control in an [upcoming section](#MTcontent):

```{r MT_ids}
MTs <- subset(fData(sce.sc), seqnames == "MT", "gene_id", drop = TRUE)
length(MTs)
```

Let us apply the [scater](http://bioconductor.org/packages/scater)
`calculateQCMetrics` to calculate the relevant QC metrics in the *single cell*
data set, and define the set of spike-in control features:

```{r sc_QC}
sce.sc <- calculateQCMetrics(
  sce.sc,
  feature_controls = list(ERCC = ERCCs, MT = MTs)
)
setSpike(sce.sc) <- "ERCC"
```

For reference purpose in upcoming figures, let us also calculate
the relevant QC metrics in the complete data set:

```{r all_QC}
sce.all <- calculateQCMetrics(
  sce.all,
  feature_controls = list(ERCC = ERCCs, MT = MTs),
  cell_controls = list(
    Bulk = which(sce.all$Status == "BULK"),
    Blank = which(sce.all$Status == "Blank")
  )
)
setSpike(sce.all) <- "ERCC"
```

## Library size

The library size refers to the total count of reads assigned to 
annotated features in each individual sample.

Here, let us simply show that the counts imported into the `SCESet`
are identical to those detected by [MultiQC](http://multiqc.info) in
an earlier section [above](#assigned):

```{r scater_outliers_totalCounts}
with(
  pData(sce.sc),
  all(log10_total_counts == log10(featureCounts_Assigned))
)
```

As a result, this metric is identical to the *featureCounts_Assigned*
metric described [above](#assigned); and thereby would
unnecessarily identify the same outliers.

## Library complexity {#libraryComplexity}

*Library complexity* refers to the proportion of assigned reads accounted for
by the $N$ most abundant features within each individual sample, with values
for $N$ usually ranging from tens to hundreds.
It is expected to observe a lower library complexity in single cells relative
to bulk samples, due to higher dropout events in single cells, and the union of
multiple phenotypes within bulk samples.
In contrast, blank samples are expected to have extremely low library
complexity, as only **92** ERCC spike-ins sequences are expected in those
libraries, in contrast to thousands of genomic features in biological samples.

In the `SCESet` object augmnented by QC metrics, let us visualise the
library complexity profile against experimental phenotype information:

```{r all_libraryComplexity}
plot(
  sce.all, block1 = "Status", block2 = "Infection",
  colour_by = "Time", nfeatures = 300, exprs_values = "counts"
)
```

First, from a QC perspective:

* Blanks samples show a very low library complexity, as expected.
* Bulk samples show a generally high library complexity, as expected.
* Single cells generally show a complexity profile comparable to that of bulk
  samples, indicative of healthy dendritic cells and unbiased sampling of
  diverse transcripts.
* A small subset of single cells (*e.g.*, `exposed:D23580:6h`,`exposed:LT2:4h`, 
  `infected:D23580:4h`, `infected:D23580:6h`, `infected:LT2:6h`) show
  suspiciously low library complexity; suggestive of technical issues during
  library prepration (*e.g.*, amplification bias, dropout events).
  Those single cells must therefore be treated as outliers, and be removed from
  the data set prior to feature selection, normalisation, and subsequent
  analyses.

Now, from a biological perspective, interesting observations may be made:

* Library complexity seems to decrease with `Time` in `exposed` and `infected`
  samples in cells exposed to or infected by bacteria, in contrast to
  control uninfected cells.
  This may indicate a polarisation of single cells toward
  specialised phenotype(s) after infection or exposure
  (*e.g.*, cytokine production, antigen presentation).
  + `Mock` (*i.e.*, `uninfected`) single cells tend to show a consistent,
    relatively high library complexity
  + Both `LT2` and `D23580` infections show a gradual decrease of library
    complexity over time. This observation *might* be more pronounced
    for `D23580`.

## Most abundant features

Although abundantly expressed features are an expected aspect of cell biology
(*e.g.*, myosin in muscle cells, clusters of differentiation in immune cells)
which may contribute to the identification and classification of cell types,
overly abundant features in RNA-Seq libraries
may indicate issues during library preparation;
in particular, cDNA amplification bias.
Moreover, when a small number of highly abundant features
account for an excessive proportion of sequenced reads,
expression levels for the vast majority of other genes may not be
estimated with sufficient accuracy.

### Identification {.tabset}

Let us identify as outliers single cells
in which the `50`, `100`, and `200` most abundant features account for a
proportion of library size
excessively *higher* than the median of *single cell* samples,
here beyond **3** MADs:

#### Top 50

```{r scater_outliers_top50}
metric <- sce.sc$pct_counts_top_50_features
top50.M <- median(metric)
top50.MAD <- mad(metric)
outliers$top50 <- (metric > top50.M + 3 * top50.MAD)
table(outliers$top50)
```

#### Top 100

```{r scater_outliers_top100}
metric <- sce.sc$pct_counts_top_100_features
top100.M <- median(metric)
top100.MAD <- mad(metric)
outliers$top100 <- (metric > top100.M + 3 * top100.MAD)
table(outliers$top100)
```

#### Top 200

```{r scater_outliers_top200}
metric <- sce.sc$pct_counts_top_200_features
top200.M <- median(metric)
top200.MAD <- mad(metric)
outliers$top200 <- (metric > top200.M + 3 * top200.MAD)
table(outliers$top200)
```

### Visualisation {.tabset}

Let us visualise outliers, along with blank and bulk samples, for reference:

#### 50

```{r plot.top50, echo=FALSE, fig.height=4}
plot.top50 <-
  ggplot(pData(sce.all), aes(Sample, pct_counts_top_50_features)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$top50,], alpha=0.6,
    min.segment.length = unit(0, "lines")) +
  scale_y_continuous(limits = c(0,100)) +
  scale_color_manual(values = col.status) +
  geom_hline(
    yintercept = top50.M + 3*c(-1,1)*top50.MAD, colour="red", linetype=2:1,
    alpha = 0.5) + labs(y = "% counts (top 50 features)", x = "Samples") +
  geom_hline(yintercept = top50.M, alpha = 0.5) + sampleQCtheme
plot.top50
ggsave(
  "03_out/complexity_50.pdf",
  plot.top50 + guides(colour = "none", shape = "none"), width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.0f%%", top50.M)`
* cut-off
    - upper bound: `r sprintf("%.0f%%", top50.M+top50.MAD*3)`

#### 100

```{r plot.top100, echo=FALSE, fig.height=4}
plot.top100 <-
  ggplot(pData(sce.all), aes(Sample, pct_counts_top_100_features)) +
  geom_point(aes(colour = Status, shape = Infection), alpha=0.75) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$top100,], alpha=0.6,
    min.segment.length = unit(0, "lines")) +
  scale_y_continuous(limits = c(0,100)) +
  scale_colour_manual(values = col.status) +
  geom_hline(
    yintercept = top100.M + 3*c(-1,1)*top100.MAD,
    colour = "red", linetype = 2:1, alpha=0.4) +
  labs(y = "% counts (top 100 features)", x = "Samples") +
  geom_hline(yintercept = top100.M, alpha=0.4) + sampleQCtheme
plot.top100
ggsave(
  "03_out/complexity_100.pdf",
  plot.top100 + guides(colour = "none", shape = "none"), width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.0f%%", top100.M)`
* cut-off
    - upper bound: `r sprintf("%.0f%%", top100.M+top100.MAD*3)`

**Note**:
Two `infected:D23580` samples appear near (below) the cut-off
threshold.
This observation can be related to the overall library complexity figure
shown [earlier](#libraryComplexity).

#### 200

```{r scater_ggplot_top200, echo=FALSE, fig.height=4}
plot.top200 <-
  ggplot(pData(sce.all), aes(Sample, pct_counts_top_200_features)) +
  geom_point(aes(colour = Status, shape = Infection), alpha=0.75) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$top200,], alpha=0.6,
    min.segment.length = unit(0, "lines")) +
  scale_y_continuous(limits = c(0,100)) +
  scale_colour_manual(values = col.status) +
  geom_hline(
    yintercept = top200.M + 3*c(-1,1)*top200.MAD,
    colour = "red", linetype = 2:1, alpha=0.4) +
  labs(y = "% counts (top 100 features)", x = "Samples") +
  geom_hline(yintercept = top200.M, alpha=0.4) + sampleQCtheme
plot.top200
ggsave(
  "03_out/complexity_200.pdf",
  plot.top200 + guides(colour = "none", shape = "none"), width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.0f%%", top200.M)`
* cut-off
    - upper bound: `r sprintf("%.0f%%", top100.M+top200.MAD*3)`

**Note**:
Two `infected:D23580` samples appear near (below) the cut-off threshold.
This observation can be related to the overall library complexity figure
shown [earlier](#libraryComplexity).

## Count of detected features

A notion closely related to library complexity is the count of features
detected within each individual sample. Although the definition of "detection
level" may be debated, we will consider here any feature with non-zero count.

Abnormal counts of detected features may be indicative of technical issue
during RNA capture and library preparation; for instance,
excessively *low* counts of detected features may indicate RNA degradation,
while excessively *high* counts of detected features may indicate capture of
multiple cells.

Owing to the considerable amount of variation that results from the relaxed
definition of detection level above, let us apply a somewhat stringent cut-off
of **2** MADs to identify as outliers single cells
with a count of detected features
excessively *lower* than the median of *single cell* samples,
here beyond **2** MADs::

```{r scater_outliers_features}
metric <- sce.sc$total_features
features.M <- median(metric)
features.MAD <- mad(metric)
outliers$FeaturesCount <- (metric < features.M - 2 * features.MAD)
table(outliers$FeaturesCount)
```

Let us visualise outliers, along with blank and bulk samples, for reference:

```{r plot.features, echo=FALSE, fig.height=4}
plot.features <-
  ggplot(pData(sce.all), aes(Sample, total_features)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_hline(yintercept = features.M, alpha = 0.5) +
  geom_hline(
    yintercept = features.M + 2 * c(-1,1) * features.MAD, colour = "red",
    linetype = 1:2, alpha = 0.5) +
  scale_colour_manual(values = col.status) +
  scale_y_continuous(limits=c(0,max(sce.all$total_features))) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$FeaturesCount,], alpha = 0.6,
    min.segment.length = unit(0, "lines")) +
  labs(y = "Count of detected features", x = "Samples") + sampleQCtheme
plot.features
ggsave(
  "03_out/features_count.pdf",
  plot.features + guides(colour = "none", shape = "none"), width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r features.M`
* cut-offs
    - upper bound:
      `r sprintf("%.1f", features.M+2*features.MAD)`
    - lower bound:
      `r sprintf("%.1f", features.M-2*features.MAD)`

## ERCC content {#ERCCoutliers}

An identical--yet very small--quantity of ERCC spike-in features is added
to each sample prior to library preparation.
As a result, technical variation and issues during library preparation may be
estimated from the proportion of reads assigned to those spike-in features;
an excessive proportion of reads assigned to ERCC spike-in features may
indicate eukaryotic RNA degradation,
while excessive variation may indicate cDNA amplication bias.

Due to the limited amount of variation that results from the generally
low levels of ERCC spike-in features detected,
let us apply a somewhat relaxed cut-off of **5** MADs
to identify as outliers single cells an ERCC content excessively *higher* than
the median of *single cell* samples:

```{r scater_outliers_ercc}
metric <- sce.sc$pct_counts_feature_controls_ERCC
ERCC.M <- median(metric)
ERCC.MAD <- mad(metric)
outliers$ERCC <- (abs(metric - ERCC.M) > 5 * ERCC.MAD)
table(outliers$ERCC)
```

Let us visualise those outliers, along with blank and bulk samples
for reference:

```{r scater_ggplot_ERCC, echo=FALSE, fig.height=4, warning=FALSE}
plot.ERCC <-
  ggplot(pData(sce.all), aes(Sample, pct_counts_feature_controls_ERCC)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_hline(yintercept = ERCC.M, alpha = 0.5) +
  geom_hline(
    yintercept = ERCC.M + 5*c(-1,1)*ERCC.MAD, colour="red", linetype=2:1,
    alpha = 0.5, min.segment.length = unit(0, "lines")) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$ERCC,], alpha = 0.6,
    min.segment.length = unit(0, "lines")) +
  scale_y_continuous(limits = c(0, 100)) +
  scale_colour_manual(values = col.status) +
  labs(y = "ERCC content (% counts)") +
  sampleQCtheme
plot.ERCC
ggsave(
  "03_out/ERCC.pdf", plot.ERCC + guides(colour = "none", shape = "none"),
  width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.0f%%", ERCC.M)`
* cut-offs
    - upper bound:
      `r sprintf("%.0f%%", ERCC.M+ERCC.MAD*5)`
    - lower bound:
      `r sprintf("%.0f%%", ERCC.M-ERCC.MAD*5)`

## Mitochondrial content {#MTcontent}

Similarly to ERCC spike-in features, mitochondrial RNA may be used
as an endogenous source of control features;
the limited count of genes encoded on the mitochondrial chromosome
relative to the human chromosomes implies that,
even at their highest levels of expression,
mitochondrial genes are expected to account for a *low* proportion
of the total read count in each sample.
Consequently, abnormally *high* proportions of reads assigned to mitochondrial
gene features may be indicative of technical issue during RNA capture and
library preparation, or abnormal count of mitochondria within the single cells.

Conversely, an unusually *low* proportion of reads assigned to mitochondrial
gene features may also indicate issues related to RNA capture and
library preparation.

Let us identify as outliers single cells for which the proportion of reads
assigned to mitochondrial gene features is excessively different
from the median of *single cell* samples:

```{r scater_outliers_MT}
metric <- sce.sc$pct_counts_feature_controls_MT
MT.M <- median(metric)
MT.MAD <- mad(metric)
outliers$MT <- (metric > MT.M + 3 * MT.MAD)
table(outliers$MT)
```

Let us visualise those outliers, along with blank and bulk samples
for reference:

```{r scater_ggplot_MT, echo=FALSE, fig.height=4, warning=FALSE}
plot.MT <-
  ggplot(pData(sce.all), aes(Sample, pct_counts_feature_controls_MT)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_hline(yintercept = MT.M, alpha = 0.4) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$MT,], alpha = 0.6,
    min.segment.length = unit(0, "lines")) +
  scale_colour_manual(values = col.status) +
  scale_y_continuous(limits = c(0, 100)) +
  geom_hline(
    yintercept = MT.M + 3 * c(-1,1) * MT.MAD, colour = "red",
    linetype = 2:1, alpha = 0.5
  ) + labs(y = "MT content (% counts)", x = "Samples") + sampleQCtheme
plot.MT
ggsave(
  "03_out/MT.pdf", plot.MT + guides(colour = "none", shape = "none"),
  width = 6, height = 4
)
```

For reference:

* single cells
    - median: `r sprintf("%.0f%%", MT.M)`
* cut-offs
    - upper bound:
      `r sprintf("%.0f%%", MT.M+MT.MAD*5)`
    - lower bound:
      `r sprintf("%.0f%%", MT.M-MT.MAD*5)`

## Summary

Let us display the count of cells identified as outliers for each of the
filters in this section:

```{r scater.fr}
scater.fr <- FilterRules(setdiff(colnames(outliers), names(fastqc.fr)))
scater.fr
```

```{r scater.fr_table, results='asis', echo=FALSE}
pander::pandoc.table(
  data.frame(
    Fail = summary(evalSeparately(scater.fr, outliers)),
    Pass = summary(!evalSeparately(scater.fr, outliers))
))
```

# Global summary of QC metrics

## By filter

Let us now collect the results of *all* the QC metrics obtained in the previous
sections, and display the count of cells :

```{r all.fr}
all.fr <- FilterRules(colnames(outliers))
all.fr
```

```{r all.fr_table, results='asis', echo=FALSE}
pander::pandoc.table(data.frame(
    Fail = summary(evalSeparately(all.fr, outliers)),
    Pass = summary(!evalSeparately(all.fr, outliers))
))
```

In the table above, none of the cells fails *all* the filters;
this demonstrates the importance of each individual filter, despite the
apparent relatedness and overlap between many of the metrics.

As a result, let us identify the cells that passed *all* of the above filters:

```{r idx.pass}
idx.pass <- (rowSums(evalSeparately(all.fr, outliers)) == 0)
```

## By cell

Secondly, we may also examine the count of single cell samples,
according to the number of QC filters that they have *failed*:

```{r failedByCell, results='asis', echo=FALSE}
failedByCell <- as.data.frame(table(rowSums(evalSeparately(all.fr, outliers))))
colnames(failedByCell) <- c("Filters", "Cells")
pander::pandoc.table(failedByCell)
```

## By phenotype {.tabset}

We may summarise experimental phenotype information for the outliers samples,
or alternatively, the samples that *successfully* passed all filters:

### Outliers

```{r summary_failed}
summary(pData(sce.sc)[
  !idx.pass,
  c("Infection", "Status", "Time", "Lane", "Plate")
])
```

### Passed

```{r summary_passed}
summary(pData(sce.sc)[
  idx.pass,
  c("Infection", "Status", "Time", "Lane", "Plate")
])
```

## Heat map summary view

Let us define an order of sample that groups single cells by
experimental phenotypes:

```{r sampleSupervisedOrder}
cells.order <- order(sce.sc$Time, sce.sc$Status, sce.sc$Infection)
```

We may then display the QC metrics failed by each single cell as red cells
in an heat map:

```{r QC_heatmap, fig.height=7, echo=FALSE}
hm_row <- rowAnnotation(
  df = pData(sce.sc)[cells.order, c("Time", "Status", "Infection")],
  col = list(
    Time = col.time,
    Status = col.status,
    Infection = col.infection
  )
)
hm <- Heatmap(
  as(evalSeparately(all.fr, outliers)[cells.order,], "matrix") + 0,
  name = "Filter",
  col = c("0" = "lightblue", "1" = "red"),
  cluster_rows = FALSE, show_row_names = FALSE, cluster_columns = FALSE,
  split = sce.sc$Time[cells.order]
)
draw(hm + hm_row)
```

```{r QC_heatmap_pdf, include=FALSE}
pdf("03_out/Heatmap_QC.pdf", height = 6, width = 4)
draw(hm + hm_row)
dev.off()
```


## Interactive tables {.tabset}

We may also display the full list of outliers samples as a table,
or alternatively, the samples that *successfully* passed all filters:

### Outliers

```{r table_outliers, echo=FALSE}
DT::datatable(
  dplyr::arrange(
    cbind(
      pData(sce.sc)[
        !idx.pass,
        c("Sample", "Infection", "Status", "Time", "Lane", "Plate")
      ],
      Filters = rowSums(evalSeparately(all.fr, outliers))[!idx.pass]
    ),
    desc(Filters)
  ),
  options = list(pageLength = 10, searching = TRUE), filter = "top",
  rownames = FALSE
)
```

### Passed

```{r table_pass, echo=FALSE}
DT::datatable(
  pData(sce.sc)[
    idx.pass,
    c("Sample", "Infection", "Status", "Time", "Lane", "Plate")
  ],
  options = list(pageLength = 10, searching = TRUE), filter = "top",
  rownames = FALSE
)
```

# Single cells selection {#cellSelection}

Let us finally retain the single cells that passed all QC metrics into a new
`SCESet`:

```{r sce.pass}
sce.pass <- sce.sc[,idx.pass]
dim(sce.pass)
```

For each experimental groups, we may display the count of cells that passed the
above QC are as follow, as well as the initial count of cells, for reference :

```{r summary_table, echo=FALSE}
DT::datatable(
  data.frame(
    Initial = as.numeric(table(sce.sc$Group)),
    Passed = as.numeric(table(sce.pass$Group)),
    Failed = as.numeric(table(sce.sc$Group))-as.numeric(table(sce.pass$Group)),
    Time = gsub("(.*)_.*_.*", "\\1", levels(sce.pass$Group)),
    Infection = gsub(".*_(.*)_.*", "\\1", levels(sce.pass$Group)),
    Status = gsub(".*_.*_(.*)", "\\1", levels(sce.pass$Group))
  ),
  options = list(pageLength = 10, searching = TRUE), filter = "top",
  rownames = FALSE
)
```

<!-- As HTML pages are built independently, objects need to be saved to disk
and imported as needed by the other R mardown files -->

```{r save_sce.pass, include=FALSE}
saveRDS(sce.pass, "rds/sce.pass.rds")
```

```{r posterAlignedDensity, include=FALSE}
# Distribution of library size per group _after filtering out outliers!_
posterDensity <-
  ggplot(subset(pData(sce.all),Sample%in%colnames(sce.pass)&Status!="BULK")) +
  geom_density(aes(Bowtie.2_overall_alignment_rate, colour = Group)) +
  scale_x_continuous(name = "Overall alignment rate (HISAT2)", limits = c(0,100)) + theme_minimal()
ggsave("poster/densityLibrary.pdf", posterDensity, width=8, height = 6)
```

[^1]: Note that shared theme settings are defined in the
[Themes](00a_themes.html) panel.

```{r ggplot_ERCC_label_legend, include=FALSE, fig.height=4, warning=FALSE}
plot.ERCC <-
  ggplot(pData(sce.all), aes(Sample, pct_counts_feature_controls_ERCC)) +
  geom_point(aes(colour = Status, shape = Infection), alpha = 0.75) +
  geom_hline(yintercept = ERCC.M, alpha = 0.5) +
  geom_hline(
    yintercept = ERCC.M + 5*c(-1,1)*ERCC.MAD, colour="red", linetype=2:1,
    alpha = 0.5) +
  geom_text_repel(
    aes(label = Time), pData(sce.sc)[outliers$ERCC,], alpha = 0.6,
    min.segment.length = unit(0, "lines")) +
  scale_y_continuous(limits = c(0, 100)) +
  scale_color_manual(values = col.status) +
  labs(y = "ERCC content (% counts)") + sampleQCtheme
plot.ERCC
ggsave("03_out/ERCC_legend.pdf", plot.ERCC, width = 6, height = 4)
```
