---
title: "Differential expression between experimental groups using *scde*"
bibliography:
  bibtex.bib
---

<!-- As HTML pages are built independently, large objects need to be
reimported from disk, and smaller ones quickly recomputed. -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(scater)
library(scran)
library(scde)
library(ensembldb)
library(dplyr)
library(goseq)
sce.norm <- readRDS("rds/sce.norm.tSNE.rds")
EnsDb.Hsapiens.v79 <- EnsDb.Hsapiens.v79::EnsDb.Hsapiens.v79
scde.res <- readRDS("rds/scde.res_v8.rds")
o.ifm.2h <- readRDS("rds/o.ifm.2h_v8.rds")
o.ifm.4h <- readRDS("rds/o.ifm.4h_v8.rds")
o.ifm.6h <- readRDS("rds/o.ifm.6h_v8.rds")
goseq.res <- readRDS("rds/goseq.res_v8.rds")
```

<!-- Cleaner code for v2
Remove ERCC spike-ins
!! Split cells by time
!! Retain genes detected at each time (10 counts, 5 cells, 1 group)
Error model: {cells @ time; genes @ time}
Prior @ time
DE: {group on all cells with NA; no batch correction, P-value < 0.01}
-->

# Prepare data

## Remove spike-in features

The analysis starts with a matrix of read counts,
filtered based on gene and cell requirements.
In this case, let us use the raw numbers of reads mapped to each endogenous
feature, recast as `integer` type:

```{r table_isSpike}
table(isSpike(sce.norm))
```

```{r sce.endo}
sce.endo <- sce.norm[!isSpike(sce.norm),]
```

```{r sce.endo_show, echo=FALSE}
dim(sce.endo)
```

## Subset data by time point {.tabset}

Considering the demonstrated impact on the `Time` factor on gene expression
profiles, and the fact that differential expression will *not* be assessed
between time points, individual time points will be processed separately for
the remainder of this *scde* analysis.

### 2h

```{r sce.2h}
sce.2h <- sce.endo[,sce.endo$Time == '2h']
dim(sce.2h)
```

### 4h

```{r sce.4h}
sce.4h <- sce.endo[,sce.endo$Time == '4h']
dim(sce.4h)
```

### 6h

```{r sce.6h}
sce.6h <- sce.endo[,sce.endo$Time == '6h']
dim(sce.6h)
```

## Filter count matrices {.tabset}

The analysis starts with a matrix of read counts,
filtered based on gene and cell requirements.

In this case, *for each time point*,
let us retain only features detected with at least **10** counts in at least
**5** cells of any of the five experimental groups,
use the raw numbers of reads mapped to each endogenous
feature, recast as `integer` type.

Let us first define a function to apply the detection cutoff:

```{r filterCounts}
filterCounts <- function(m, counts = 10, cells = 10){
  apply(m, 1, function(e){
    return(sum(e >= counts) >= cells)
  })
}
```

### 2h

```{r cleanCounts_2h}
sg.2h <- droplevels(sce.2h$Group)
keep.2h <- filterCounts(counts(sce.2h))
cd.2h <- counts(sce.2h)[keep.2h,]; storage.mode(cd.2h) <- 'integer'
```

```{r cd.2h_show, echo=FALSE}
dim(cd.2h)
```

### 4h

```{r cleanCounts_4h}
sg.4h <- droplevels(sce.4h$Group)
keep.4h <- filterCounts(counts(sce.4h))
cd.4h <- counts(sce.4h)[keep.4h,]; storage.mode(cd.4h) <- 'integer'
```

```{r cd.4h_show, echo=FALSE}
dim(cd.4h)
```

### 6h

```{r cleanCounts_6h}
sg.6h <- droplevels(sce.6h$Group)
keep.6h <- filterCounts(counts(sce.6h))
cd.6h <- counts(sce.6h)[keep.6h,]; storage.mode(cd.6h) <- 'integer'
```

```{r cd.6h_show, echo=FALSE}
dim(cd.6h)
```

# Error models {.tabset}

Here, we fit the error models on which all subsequent calculations
will rely. The fitting process relies on a subset of robust genes that are
detected in multiple cross-cell comparisons. Here we supply the
`groups` argument, so that the error models for each
experimental group of cells are fit independently.
If the groups argument is omitted, the models would be fit using a common set.

## 2h

```{r o.ifm.2h, eval=FALSE}
o.ifm.2h <- scde.error.models(
  counts = cd.2h, groups = sg.2h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.2h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.2h, "rds/o.ifm.2h_v8.rds")
write.csv(o.ifm.2h, "SCDE_v8/o.ifm.2h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.2h}
valid.cells <- o.ifm.2h$corr.a > 0; table(valid.cells)
o.ifm.2h <- o.ifm.2h[valid.cells, ]
dim(o.ifm.2h)
```

## 4h

```{r o.ifm.4h, eval=FALSE}
o.ifm.4h <- scde.error.models(
  counts = cd.4h, groups = sg.4h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.4h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.4h, "rds/o.ifm.4h_v8.rds")
write.csv(o.ifm.4h, "SCDE_v8/o.ifm.4h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.4h}
valid.cells <- o.ifm.4h$corr.a > 0; table(valid.cells)
o.ifm.4h <- o.ifm.4h[valid.cells, ]
dim(o.ifm.4h)
```

## 6h

```{r o.ifm.6h, eval=FALSE}
o.ifm.6h <- scde.error.models(
  counts = cd.6h, groups = sg.6h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.6h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.6h, "rds/o.ifm.6h_v8.rds")
write.csv(o.ifm.6h, "SCDE_v8/o.ifm.6h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.6h}
valid.cells <- o.ifm.6h$corr.a > 0; table(valid.cells)
o.ifm.6h <- o.ifm.6h[valid.cells, ]
dim(o.ifm.6h)
```

# Reorder counts to match error models {.tabset}

The `scde.error.models` produces error models with cells reordered 
by experimental group.
For clarity, let us prepare a `SCESet` and count matrix that match this order:

## 2h

```{r sce.ifm.2h}
sce.ifm.2h <- sce.2h[,rownames(o.ifm.2h)]
cd.ifm.2h <- counts(sce.ifm.2h); storage.mode(cd.ifm.2h) <- "integer"
```

## 4h

```{r sce.ifm.4h}
sce.ifm.4h <- sce.4h[,rownames(o.ifm.4h)]
cd.ifm.4h <- counts(sce.ifm.4h); storage.mode(cd.ifm.4h) <- "integer"
```

## 6h

```{r sce.ifm.6h}
sce.ifm.6h <- sce.6h[,rownames(o.ifm.6h)]
cd.ifm.6h <- counts(sce.ifm.6h); storage.mode(cd.ifm.6h) <- "integer"
```

# Prior distribution for gene expression magnitudes {.tabset}

Finally, we need to define an expression magnitude prior for the genes.
Its main function, however, is to define a grid of expression magnitude values
on which the numerical calculations will be carried out.

## 2h

```{r o.prior.2h}
o.prior.2h <- scde.expression.prior(
  models = o.ifm.2h, counts = cd.ifm.2h, show.plot = TRUE
)
```

## 4h

```{r o.prior.4h}
o.prior.4h <- scde.expression.prior(
  models = o.ifm.4h, counts = cd.ifm.4h, show.plot = TRUE
)
```

## 6h

```{r o.prior.6h}
o.prior.6h <- scde.expression.prior(
  models = o.ifm.6h, counts = cd.ifm.6h, show.plot = TRUE
)
```

# Differential expression

## Setup

Let us first define:

* a list to store the result tables returned by *scde*

```{r scde.res_init, eval=FALSE}
scde.res <- list()
```

* a function used to convert the Z-score computed by *scde* to a empirical
P-value ([reference](https://www.biostars.org/p/17227/)):

```{r convert.z.score}
convert.z.score <- function(x, one.sided = NULL) {
  z <- x$Z
  if(is.null(one.sided)) {
    pval = pnorm(-abs(z));
    pval = 2 * pval
  } else if(one.sided=="-") {
    pval = pnorm(z);
  } else {
    pval = pnorm(-z);
  }
    x <- cbind(
      x,
      p.value = pval
  )
  return(x);
}   
```

* a function used to annotate the tables of results returned by *scde*:

```{r addGENENAME}
addGENENAME <- function(x){
  x <- cbind(
    GENENAME=mapIds(EnsDb.Hsapiens.v79, rownames(x), 'GENENAME', 'GENEID'),
    x
  )
  return(x)
}
```

* a function to order results by decreasing absolute Z-score:

```{r orderResults}
orderResults <- function(x){
  x <- x[with(x, order(abs(Z), decreasing = TRUE)),]
  return(x)
}
```

* a function to visualise *scran*-normalised gene expression for a given
gene:

Let us also define a function that displays normalised (*scran*) expression
data in each group:

```{r normExprsByName}
normExprsById <- function(geneId){
  geneName <- subset(fData(sce.endo),gene_id==geneId,"gene_name",drop=TRUE)
  gdata <- data.frame(
    norm_exprs = norm_exprs(sce.endo)[geneId,],
    pData(sce.endo)[,c("Infection","Status","Time")],
    row.names = sampleNames(sce.endo)
  )
  ggplot(gdata, aes(Infection, norm_exprs)) + 
    geom_violin() + geom_jitter(width = 0.1) +
    facet_grid(Time ~ Status) +
    ggtitle(sprintf("%s - %s", geneId, geneName))
}
```

* functions to visualise *scde* estimate of expression and
  differential expression
  for a given gene (identifier)
  in a given contrast
  within a specific time point:
  
```{r single.scde}
single.scde.2h <- function(gene, groupTarget, groupRef){
  gT <- sprintf("2h_%s", groupTarget); gR <- sprintf("2h_%s", groupRef)
  sg.test <- factor(pData(sce.ifm.2h)[,"Group"], levels = c(gT, gR))
  scde.test.gene.expression.difference(
    gene, o.ifm.2h, cd.ifm.2h, o.prior.2h, sg.test,
    n.cores = 4, verbose = 1
  )
}
single.scde.4h <- function(gene, groupTarget, groupRef){
  gT <- sprintf("4h_%s", groupTarget); gR <- sprintf("4h_%s", groupRef)
  sg.test <- factor(pData(sce.ifm.4h)[,"Group"], levels = c(gT, gR))
  scde.test.gene.expression.difference(
    gene, o.ifm.4h, cd.ifm.4h, o.prior.4h, sg.test,
    n.cores = 4, verbose = 1
  )
}
single.scde.6h <- function(gene, groupTarget, groupRef){
  gT <- sprintf("6h_%s", groupTarget); gR <- sprintf("6h_%s", groupRef)
  sg.test <- factor(pData(sce.ifm.6h)[,"Group"], levels = c(gT, gR))
  scde.test.gene.expression.difference(
    gene, o.ifm.6h, cd.ifm.6h, o.prior.6h, sg.test,
    n.cores = 4, verbose = 1
  )
}
```

* various significance levels:

```{r volcano.sig}
sig.levels <- c(0.05, 0.01)
volcano.sig <- data.frame(
  P = sig.levels,
  level = as.character(sig.levels)
)
```

* a function to visualise *scde* differential expression statistics, and
  return a table with those statistics augmented by a an empirical *P* value
  computed from the *Z* score returned by *scde*:

```{r volcano.mle}
volcano.mle <- function(x, sub = NULL){
  varName <- deparse(substitute(x))
  x <- convert.z.score(x)
  gg <- ggplot(x, aes(mle, -log10(p.value))) +
    geom_point(aes(colour = (cZ != 0))) +
    geom_hline(aes(yintercept=-log10(p.value),linetype=level),volcano.sig) +
    ggtitle(varName, sub)
  print(gg)
  return(x)
}
```

## Contrasts {.tabset}

### List

```{r contrasts.time}
contrasts.2h <- list(
  c("2h_D23580_infected", "2h_Mock_uninfected"), # vs. Mock
  c("2h_LT2_infected", "2h_Mock_uninfected"),
  c("2h_D23580_exposed", "2h_Mock_uninfected"),
  c("2h_LT2_exposed", "2h_Mock_uninfected"),
  c("2h_D23580_infected", "2h_LT2_infected"), # direct
  c("2h_D23580_infected", "2h_D23580_exposed"),
  c("2h_LT2_infected", "2h_LT2_exposed"),
  c("2h_D23580_exposed", "2h_LT2_exposed")
)
contrasts.4h <- list(
  c("4h_D23580_infected", "4h_Mock_uninfected"), # 4h
  c("4h_LT2_infected", "4h_Mock_uninfected"),
  c("4h_D23580_exposed", "4h_Mock_uninfected"),
  c("4h_LT2_exposed", "4h_Mock_uninfected"),
  c("4h_D23580_infected", "4h_LT2_infected"), # 4h
  c("4h_D23580_infected", "4h_D23580_exposed"),
  c("4h_LT2_infected", "4h_LT2_exposed"),
  c("4h_D23580_exposed", "4h_LT2_exposed")
)
contrasts.6h <- list(
  c("6h_D23580_infected", "6h_Mock_uninfected"), # 6h
  c("6h_LT2_infected", "6h_Mock_uninfected"),
  c("6h_D23580_exposed", "6h_Mock_uninfected"),
  c("6h_LT2_exposed", "6h_Mock_uninfected"),
  c("6h_D23580_infected", "6h_LT2_infected"), # 6h
  c("6h_D23580_infected", "6h_D23580_exposed"),
  c("6h_LT2_infected", "6h_LT2_exposed"),
  c("6h_D23580_exposed", "6h_LT2_exposed")
)
```

### 2h

```{r contrasts.2h_run, eval=FALSE}
for (contrastNames in contrasts.2h){
  groupTarget <- contrastNames[1]; groupRef <- contrastNames[2]
  sg.test <- factor(sce.ifm.2h$Group, levels = c(groupTarget, groupRef))
  names(sg.test) <- sampleNames(sce.2h); summary(sg.test)
  contrastName <- sprintf("%s-%s", groupTarget, groupRef); message(contrastName)
  scde.res[[contrastName]] <- scde.expression.difference(
    o.ifm.2h, cd.2h, o.prior.2h, sg.test, n.cores = 4, verbose = 1
  )
  saveRDS(scde.res, "rds/scde.res_v8.rds") # checkpoint
}
```

### 4h

```{r contrasts.4h_run, eval=FALSE}
for (contrastNames in contrasts.4h){
  groupTarget <- contrastNames[1]; groupRef <- contrastNames[2]
  sg.test <- factor(sce.ifm.4h$Group, levels = c(groupTarget, groupRef))
  names(sg.test) <- sampleNames(sce.4h); summary(sg.test)
  contrastName <- sprintf("%s-%s", groupTarget, groupRef); message(contrastName)
  scde.res[[contrastName]] <- scde.expression.difference(
    o.ifm.4h, cd.4h, o.prior.4h, sg.test, n.cores = 4, verbose = 1
  )
  saveRDS(scde.res, "rds/scde.res_v8.rds") # checkpoint
}
```

### 6h

```{r contrasts.6h_run, eval=FALSE}
for (contrastNames in contrasts.6h){
  groupTarget <- contrastNames[1]; groupRef <- contrastNames[2]
  sg.test <- factor(sce.ifm.6h$Group, levels = c(groupTarget, groupRef))
  names(sg.test) <- sampleNames(sce.6h); summary(sg.test)
  contrastName <- sprintf("%s-%s", groupTarget, groupRef); message(contrastName)
  scde.res[[contrastName]] <- scde.expression.difference(
    o.ifm.6h, cd.6h, o.prior.6h, sg.test, n.cores = 4, verbose = 1
  )
  saveRDS(scde.res, "rds/scde.res_v8.rds") # checkpoint
}
```

```{r scde.table_csv, echo=FALSE, eval=FALSE}
for (contrastName in names(scde.res)){
  scde.table <- scde.res[[contrastName]]
  scde.table <- convert.z.score(addGENENAME(orderResults(scde.table)))
  csv.file <- sprintf("SCDE_v8/%s.csv", contrastName)
  write.csv(scde.table, csv.file)
}
```

# Volcano plots {.tabset}

## Uninfected

```{r volcano_uninfected, echo=FALSE, fig.height=9}
v.data <- do.call(
  "rbind",
  lapply(
    grep("uninfected", names(scde.res), value = TRUE),
    function(contrastName){
      timepoint <- gsub("([[:digit:]]h)_.*", "\\1", contrastName)
      contrastTreatment <-
        gsub("[[:digit:]]h_(.*)-[[:digit:]]h_(.*)", "\\1-\\2", contrastName)
      x <- data.frame(
        Gene = rownames(scde.res[[contrastName]]),
        scde.res[[contrastName]][,c("mle", "Z", "cZ")],
        Contrast = gsub("_", " ", contrastTreatment),
        Time = timepoint
      )
      x <- convert.z.score(x)
    }
  )
)
ggplot(v.data) +
  geom_point(aes(mle, -log10(p.value), colour = (cZ != 0)), size = 0.5) +
  facet_grid(Contrast ~ Time) +
  geom_hline(
    aes(yintercept = -log10(P), linetype = level), data = volcano.sig
  )
```

## Direct

```{r volcano_direct, echo=FALSE, fig.height=9}
v.data <- do.call(
  "rbind",
  lapply(
    grep("uninfected", names(scde.res), value = TRUE, invert = TRUE),
    function(contrastName){
      timepoint <- gsub("([[:digit:]]h)_.*", "\\1", contrastName)
      contrastTreatment <-
        gsub("[[:digit:]]h_(.*)-[[:digit:]]h_(.*)", "\\1-\\2", contrastName)
      x <- data.frame(
        Gene = rownames(scde.res[[contrastName]]),
        scde.res[[contrastName]][,c("mle", "Z", "cZ")],
        Contrast = gsub("_", " ", contrastTreatment),
        Time = timepoint
      )
      x <- convert.z.score(x)
    }
  )
)
ggplot(v.data) +
  geom_point(aes(mle, -log10(p.value), colour = (cZ != 0)), size = 0.5) +
  facet_grid(Contrast ~ Time) +
  geom_hline(
    aes(yintercept = -log10(P), linetype = level), data = volcano.sig
  )
```

# Count DE genes at various cut-offs {.tabset}

## Uninfected

```{r countDE_control, echo=FALSE, results='asis'}
v.data <- do.call(
  "rbind",
  lapply(
    grep("uninfected", names(scde.res), value = TRUE),
    function(contrastName){
      x.res <- scde.res[[contrastName]][,c("Z", "cZ")]
      x.res <- convert.z.score(x.res)
      x.table <- data.frame(
        P.01 = sum(x.res$p.value < 0.01),
        P.05 = sum(x.res$p.value < 0.05),
        cZ = sum(x.res$cZ != 0),
        row.names = contrastName
      )
      return(x.table)
    }
  )
)
pander::pandoc.table(v.data)
```

## Direct

```{r countDE_direct, echo=FALSE, results='asis'}
v.data <- do.call(
  "rbind",
  lapply(
    grep("uninfected", names(scde.res), value = TRUE, invert = TRUE),
    function(contrastName){
      x.res <- scde.res[[contrastName]][,c("Z", "cZ")]
      x.res <- convert.z.score(x.res)
      x.table <- data.frame(
        P.01 = sum(x.res$p.value < 0.01),
        P.05 = sum(x.res$p.value < 0.05),
        cZ = sum(x.res$cZ != 0),
        row.names = contrastName
      )
      return(x.table)
    }
  )
)
pander::pandoc.table(v.data)
```


# Gene ontology

Let us use the [goseq](http://bioconductor.org/packages/goseq) package to
identify the most enriched gene ontologies among the various lists of DE genes.
Note that we restrict the results to GO categories associated with at least
**10** genes, for robustness.

Let us first define:

* a list to store the GO enrichment tables returned by *goseq*

```{r goseq.res_init, eval=FALSE}
goseq.res <- list()
```

* the gene length information reported by *featureCounts*:

```{r geneLengths}
geneLengths <-
  read.delim("counts/WTCHG_305264_201201", row.names = 1)[,1, drop = FALSE]
```

## Compute {.tabset}

Let us identify over- and under-represented GO categories in DE genes
identified in each contrast,
using as background the list of genes detected at each time point:

prepare binary vectors to indicate 

### 2h

```{r GO_2h_DE, eval=FALSE}
bg.2h <- featureNames(sce.ifm.2h)
for (contrastName in grep("^2h", names(scde.res), value = TRUE)){
  message(contrastName); scde.table <- scde.res[[contrastName]]
  scde.table <- convert.z.score(scde.table)
  scde.table <- subset(scde.table, p.value < 0.01)
  de.genes <- (bg.2h %in% rownames(scde.table))
  names(de.genes) <- bg.2h; table(de.genes)
  pwf <- nullp(de.genes, bias.data = geneLengths[names(de.genes),])
  go.res <- goseq(pwf, "hg38", "ensGene")
  goseq.res[[contrastName]] <- go.res
  saveRDS(goseq.res, "rds/goseq.res_v8.rds") # checkpoint
}
```

### 4h

```{r GO_4h_DE, eval=FALSE}
bg.4h <- featureNames(sce.ifm.4h)
for (contrastName in grep("^4h", names(scde.res), value = TRUE)){
  message(contrastName); scde.table <- scde.res[[contrastName]]
  scde.table <- convert.z.score(scde.table)
  scde.table <- subset(scde.table, p.value < 0.01)
  de.genes <- (bg.4h %in% rownames(scde.table))
  names(de.genes) <- bg.4h; table(de.genes)
  pwf <- nullp(de.genes, bias.data = geneLengths[names(de.genes),])
  go.res <- goseq(pwf, "hg38", "ensGene")
  goseq.res[[contrastName]] <- go.res
  saveRDS(goseq.res, "rds/goseq.res_v8.rds") # checkpoint
}
```

### 6h

```{r GO_6h_DE, eval=FALSE}
bg.6h <- featureNames(sce.ifm.6h)
for (contrastName in grep("^6h", names(scde.res), value = TRUE)){
  message(contrastName); scde.table <- scde.res[[contrastName]]
  scde.table <- convert.z.score(scde.table)
  scde.table <- subset(scde.table, p.value < 0.01)
  de.genes <- (bg.6h %in% rownames(scde.table))
  names(de.genes) <- bg.6h; table(de.genes)
  pwf <- nullp(de.genes, bias.data = geneLengths[names(de.genes),])
  go.res <- goseq(pwf, "hg38", "ensGene")
  goseq.res[[contrastName]] <- go.res
  saveRDS(goseq.res, "rds/goseq.res_v8.rds") # checkpoint
}
```

```{r goseq.table_csv, echo=FALSE, eval=FALSE}
for (contrastName in names(goseq.res)){
  goseq.table <- goseq.res[[contrastName]]
  write.csv(goseq.table, sprintf("SCDE_v8/GO_%s.csv", contrastName))
}
```

## Tables {.tabset}

### Setup

Let us define a function to display the **50** most significantly
`"over"`- or `"under"`-represented GO categories
annotated to at least **10** genes (in the entire genome):

```{r showTopGO}
showTopGO <- function(x, direction = "over", cutoff = 0.05, min = 10){
  goCols <- c(
  "category","over_represented_pvalue","under_represented_pvalue",
  "numDEInCat","numInCat","ontology","term")
  directionCol <- sprintf("%s_represented_pvalue", direction)
  tmp.go <- x[,goCols]
  tmp.sig <- (tmp.go[,directionCol] < cutoff); tmp.go <- tmp.go[tmp.sig,]
  tmp.min <- (tmp.go[,"numInCat"] >= min); tmp.go <- tmp.go[tmp.min,]
  tmp.go <- tmp.go[order(tmp.go[,directionCol]),]
  tmp.go <- head(tmp.go, 50)
  DT::datatable(
    tmp.go, list(pageLength = 10, searching = TRUE), filter = "top",
    rownames = FALSE
  )
}
```

### Over-represented {.tabset}

#### `r names(goseq.res)[1]`

```{r goseq.over_1, echo=FALSE}
showTopGO(goseq.res[[1]])
```

#### `r names(goseq.res)[2]`

```{r goseq.over_2, echo=FALSE}
showTopGO(goseq.res[[2]])
```

#### `r names(goseq.res)[3]`

```{r goseq.over_3, echo=FALSE}
showTopGO(goseq.res[[3]])
```

#### `r names(goseq.res)[4]`

```{r goseq.over_4, echo=FALSE}
showTopGO(goseq.res[[4]])
```

#### `r names(goseq.res)[5]`

```{r goseq.over_5, echo=FALSE}
showTopGO(goseq.res[[5]])
```

#### `r names(goseq.res)[6]`

```{r goseq.over_6, echo=FALSE}
showTopGO(goseq.res[[6]])
```

#### `r names(goseq.res)[7]`

```{r goseq.over_7, echo=FALSE}
showTopGO(goseq.res[[7]])
```

#### `r names(goseq.res)[8]`

```{r goseq.over_8, echo=FALSE}
showTopGO(goseq.res[[8]])
```

#### `r names(goseq.res)[9]`

```{r goseq.over_9, echo=FALSE}
showTopGO(goseq.res[[9]])
```

#### `r names(goseq.res)[10]`

```{r goseq.over_10, echo=FALSE}
showTopGO(goseq.res[[10]])
```

#### `r names(goseq.res)[11]`

```{r goseq.over_11, echo=FALSE}
showTopGO(goseq.res[[11]])
```

#### `r names(goseq.res)[12]`

```{r goseq.over_12, echo=FALSE}
showTopGO(goseq.res[[12]])
```

#### `r names(goseq.res)[13]`

```{r goseq.over_13, echo=FALSE}
showTopGO(goseq.res[[13]])
```

#### `r names(goseq.res)[14]`

```{r goseq.over_14, echo=FALSE}
showTopGO(goseq.res[[14]])
```

#### `r names(goseq.res)[15]`

```{r goseq.over_15, echo=FALSE}
showTopGO(goseq.res[[15]])
```

#### `r names(goseq.res)[16]`

```{r goseq.over_16, echo=FALSE}
showTopGO(goseq.res[[16]])
```

#### `r names(goseq.res)[17]`

```{r goseq.over_17, echo=FALSE}
showTopGO(goseq.res[[17]])
```

#### `r names(goseq.res)[18]`

```{r goseq.over_18, echo=FALSE}
showTopGO(goseq.res[[18]])
```

#### `r names(goseq.res)[19]`

```{r goseq.over_19, echo=FALSE}
showTopGO(goseq.res[[19]])
```

#### `r names(goseq.res)[20]`

```{r goseq.over_20, echo=FALSE}
showTopGO(goseq.res[[20]])
```

#### `r names(goseq.res)[21]`

```{r goseq.over_21, echo=FALSE}
showTopGO(goseq.res[[21]])
```

#### `r names(goseq.res)[22]`

```{r goseq.over_22, echo=FALSE}
showTopGO(goseq.res[[22]])
```

#### `r names(goseq.res)[23]`

```{r goseq.over_23, echo=FALSE}
showTopGO(goseq.res[[23]])
```

#### `r names(goseq.res)[24]`

```{r goseq.over_24, echo=FALSE}
showTopGO(goseq.res[[24]])
```


### Under-represented {.tabset}

#### `r names(goseq.res)[1]`

```{r goseq.under_1, echo=FALSE}
showTopGO(goseq.res[[1]], "under")
```

#### `r names(goseq.res)[2]`

```{r goseq.under_2, echo=FALSE}
showTopGO(goseq.res[[2]], "under")
```

#### `r names(goseq.res)[3]`

```{r goseq.under_3, echo=FALSE}
showTopGO(goseq.res[[3]], "under")
```

#### `r names(goseq.res)[4]`

```{r goseq.under_4, echo=FALSE}
showTopGO(goseq.res[[4]], "under")
```

#### `r names(goseq.res)[5]`

```{r goseq.under_5, echo=FALSE}
showTopGO(goseq.res[[5]], "under")
```

#### `r names(goseq.res)[6]`

```{r goseq.under_6, echo=FALSE}
showTopGO(goseq.res[[6]], "under")
```

#### `r names(goseq.res)[7]`

```{r goseq.under_7, echo=FALSE}
showTopGO(goseq.res[[7]], "under")
```

#### `r names(goseq.res)[8]`

```{r goseq.under_8, echo=FALSE}
showTopGO(goseq.res[[8]], "under")
```

#### `r names(goseq.res)[9]`

```{r goseq.under_9, echo=FALSE}
showTopGO(goseq.res[[9]], "under")
```

#### `r names(goseq.res)[10]`

```{r goseq.under_10, echo=FALSE}
showTopGO(goseq.res[[10]], "under")
```

#### `r names(goseq.res)[11]`

```{r goseq.under_11, echo=FALSE}
showTopGO(goseq.res[[11]], "under")
```

#### `r names(goseq.res)[12]`

```{r goseq.under_12, echo=FALSE}
showTopGO(goseq.res[[12]], "under")
```

#### `r names(goseq.res)[13]`

```{r goseq.under_13, echo=FALSE}
showTopGO(goseq.res[[13]], "under")
```

#### `r names(goseq.res)[14]`

```{r goseq.under_14, echo=FALSE}
showTopGO(goseq.res[[14]], "under")
```

#### `r names(goseq.res)[15]`

```{r goseq.under_15, echo=FALSE}
showTopGO(goseq.res[[15]], "under")
```

#### `r names(goseq.res)[16]`

```{r goseq.under_16, echo=FALSE}
showTopGO(goseq.res[[16]], "under")
```

#### `r names(goseq.res)[17]`

```{r goseq.under_17, echo=FALSE}
showTopGO(goseq.res[[17]], "under")
```

#### `r names(goseq.res)[18]`

```{r goseq.under_18, echo=FALSE}
showTopGO(goseq.res[[18]], "under")
```

#### `r names(goseq.res)[19]`

```{r goseq.under_19, echo=FALSE}
showTopGO(goseq.res[[19]], "under")
```

#### `r names(goseq.res)[20]`

```{r goseq.under_20, echo=FALSE}
showTopGO(goseq.res[[20]], "under")
```

#### `r names(goseq.res)[21]`

```{r goseq.under_21, echo=FALSE}
showTopGO(goseq.res[[21]], "under")
```

#### `r names(goseq.res)[22]`

```{r goseq.under_22, echo=FALSE}
showTopGO(goseq.res[[22]], "under")
```

#### `r names(goseq.res)[23]`

```{r goseq.under_23, echo=FALSE}
showTopGO(goseq.res[[23]], "under")
```

#### `r names(goseq.res)[24]`

```{r goseq.under_24, echo=FALSE}
showTopGO(goseq.res[[24]], "under")
```

# Expression profile of selected genes {.tabset}

Least significant gene record (P ~ 0.01)
in contrast D23580 vs. LT2 infected at 6h:

```{r ex.scde.last.sig}
ex.scde.last.sig <- addGENENAME(tail(subset(orderResults(convert.z.score(
  scde.res[["6h_D23580_infected-6h_LT2_infected"]])), p.value < 0.01), 1))
```

```{r ex.scde.last.sig_table, echo=FALSE, results='asis'}
pander::pandoc.table(ex.scde.last.sig)
```


## `r rownames(ex.scde.last.sig)`

Least significant gene record (P ~ 0.01)
in contrast D23580 vs. LT2 infected at 6h:

### *scran*

```{r normExprsById_ex.scde.last.sig}
normExprsById(rownames(ex.scde.last.sig))
```

### *scde*

```{r single.scde.6h_ex.scde.last.sig}
single.scde.6h(rownames(ex.scde.last.sig), "D23580_infected", "LT2_infected")
```

## TLR2 {.tabset}

Up-regulated in D23580 infected vs. exposed ($P < 10^{-4}$),
while $NS$ in LT2 (infected, 2h):

### *scran*

```{r normExprsById_TLR2}
normExprsById("ENSG00000137462")
```

### *scde* D23580

```{r single.scde.6h_D23580_TLR2}
single.scde.6h("ENSG00000137462", "D23580_infected", "D23580_exposed")
```

### *scde* LT2

```{r single.scde.6h_LT2_TLR2}
single.scde.6h("ENSG00000137462", "LT2_infected", "LT2_exposed")
```

## IL10 {.tabset}

Up-regulated in D23580 infected vs. uninfected at 4h ($P < 5 \times 10^{-4}$),
while $NS$ in LT2 ($P \approx 0.40$):

### *scran*

```{r normExprsById_IL10}
normExprsById("ENSG00000136634")
```

### *scde* D23580

```{r single.scde.6h_D23580_IL10}
single.scde.6h("ENSG00000136634", "D23580_infected", "D23580_exposed")
```

### *scde* LT2

```{r single.scde.6h_LT2_IL10}
single.scde.6h("ENSG00000136634", "LT2_infected", "LT2_exposed")
```


## RAB2B {.tabset}

Up-regulated in D23580 vs. LT2 (infected, 2h):

### *scran*

```{r normExprsById_RAB2B}
normExprsById("ENSG00000129472")
```

### *scde*

```{r single.scde.2h_RAB2B}
single.scde.2h("ENSG00000129472", "D23580_infected", "LT2_infected")
```

## RAB29

Gene of interest: *P*-value ~ 0.025 between D23580 and LT2 infected at 6h:

### *scran*

```{r normExprsById_RAB29}
normExprsById("ENSG00000117280")
```

### *scde*

```{r}
single.scde.6h("ENSG00000117280", "D23580_infected", "LT2_infected")
```

# Comparison of fold-change estimates

## Contrasts {.tabset}

### Setup

Let us define:

* the required columns in the *scde* result tables

```{r resCols}
resCols <- c("mle","Z")
```

* a function used to annotate each gene in the tables of *scde* results
  with the significance of differential expression in the pair of contrasts
  compared:

```{r add.sig.XY}
add.sig.XY <- function(x, cutoff = 0.01){
  sig.levels <- c(NA_character_, "X", "Y", "Both")
  alpha.levels <- c(0.3, 1, 1, 1)
  sig.x <- (x$p.value.x < cutoff); sig.y <- (x$p.value.y < cutoff)
  x$p.01 <- sig.levels[1 + sig.x + 2 * sig.y]
  x$alpha <- alpha.levels[1 + sig.x + 2 * sig.y]
  return(x)
}
```

* a function used to annotate each gene in the tables of *scde* results
  whether it is significantly differentially expressed in opposite direction
  in the pair of contrasts compared:

```{r}
addOpposite <- function(x, cutoff = 0.01){
  x$opposite <- NA_character_
  x.sig.idx <- (x$p.value.x < cutoff & x$p.value.y < cutoff)
  x.opposite <- (x$mle.x * x$mle.y) < 0
  x[x.sig.idx, "opposite"] <- x.opposite[x.sig.idx]
  return(x)
}
```

Let us define:

* the maximal range of expression fold changes observed across contrasts;
  used to scale figures for symmetry of axes
  and comparabison between time points:

```{r mleRange}
mleRange <-
  max(abs(do.call("c", lapply(scde.res, function(x){return(x$mle)}))))*c(-1,1)
```

### List

```{r contrasts.contrasts.time}
contrasts.contrasts.2h <- list(
c("2h_D23580_infected-2h_Mock_uninfected","2h_LT2_infected-2h_Mock_uninfected"),
c("2h_D23580_infected-2h_Mock_uninfected","2h_D23580_exposed-2h_Mock_uninfected"),
c("2h_LT2_infected-2h_Mock_uninfected","2h_LT2_exposed-2h_Mock_uninfected"),
c("2h_D23580_exposed-2h_Mock_uninfected","2h_LT2_exposed-2h_Mock_uninfected")
)
contrasts.contrasts.4h <- list(
c("4h_D23580_infected-4h_Mock_uninfected","4h_LT2_infected-4h_Mock_uninfected"),
c("4h_D23580_infected-4h_Mock_uninfected","4h_D23580_exposed-4h_Mock_uninfected"),
c("4h_LT2_infected-4h_Mock_uninfected","4h_LT2_exposed-4h_Mock_uninfected"),
c("4h_D23580_exposed-4h_Mock_uninfected","4h_LT2_exposed-4h_Mock_uninfected")
)
contrasts.contrasts.6h <- list(
c("6h_D23580_infected-6h_Mock_uninfected","6h_LT2_infected-6h_Mock_uninfected"),
c("6h_D23580_infected-6h_Mock_uninfected","6h_D23580_exposed-6h_Mock_uninfected"),
c("6h_LT2_infected-6h_Mock_uninfected","6h_LT2_exposed-6h_Mock_uninfected"),
c("6h_D23580_exposed-6h_Mock_uninfected","6h_LT2_exposed-6h_Mock_uninfected")
)
```

### 2h

```{r corPlot_2h, eval=TRUE, echo=FALSE, fig.height=7}
v.data <- do.call(
  "rbind",
  lapply(
    contrasts.contrasts.2h,
    function(contrastNames){
      cn1 <- contrastNames[1]; cn2 <- contrastNames[2]
      tn1<-gsub("[[:digit:]]h_(.*)-[[:digit:]]h_(.*)", "\\1-\\2", cn1)
      tn2<-gsub("[[:digit:]]h_(.*)-[[:digit:]]h_(.*)", "\\1-\\2", cn2)
      x1 <- scde.res[[cn1]][,resCols]; x2 <- scde.res[[cn2]][,resCols]
      x1 <- convert.z.score(x1); x2 <- convert.z.score(x2)
      x1$Contrast <- gsub("_", " ", tn1);x2$Contrast <- gsub("_", " ", tn2)
      colnames(x1) <- gsub("$", ".y", colnames(x1))
      colnames(x2) <- gsub("$", ".x", colnames(x2))
      x <- cbind(x1, x2)
      x <- add.sig.XY(x)
      x$Contrast <- sprintf("Y : %s\nX : %s", tn1, tn2)
      return(x)
    }
  )
)
ggplot(v.data) +
  geom_point(aes(mle.x, mle.y, colour = p.01, alpha = alpha), size = 0.5) +
  facet_wrap(~ Contrast) +
  scale_x_continuous(limits=mleRange)+scale_y_continuous(limits=mleRange) +
  ggtitle("2h") + scale_alpha_continuous(guide = FALSE) + labs(x="",y="")
```

```{r compare.2h_csv, echo=FALSE, eval=FALSE}
for (contrastName in unique(v.data$Contrast)){
  v.subset <- subset(addGENENAME(v.data), Contrast == contrastName)
  v.subset$Contrast <- gsub("\n", "-", v.subset$Contrast)
  v.subset <- addOpposite(v.subset)
  c.y <- gsub(" ", "_", unique(v.subset$Contrast.y))
  c.x <- gsub(" ", "_", unique(v.subset$Contrast.x))
  message(c.y); message(c.x)
  csvFile <- sprintf("SCDE_v8/compare.2h_%s-2h_%s.csv", c.y, c.x)
  message(csvFile)
  write.csv(v.subset, csvFile)
}
```

### 4h

```{r corPlot_4h, eval=TRUE, echo=FALSE, fig.height=7}
v.data <- do.call(
  "rbind",
  lapply(
    contrasts.contrasts.4h,
    function(contrastNames){
      cn1 <- contrastNames[1]; cn2 <- contrastNames[2]
      tn1<-gsub("[[:digit:]]h_(.*)-[[:digit:]]h_(.*)", "\\1-\\2", cn1)
      tn2<-gsub("[[:digit:]]h_(.*)-[[:digit:]]h_(.*)", "\\1-\\2", cn2)
      x1 <- scde.res[[cn1]][,resCols]; x2 <- scde.res[[cn2]][,resCols]
      x1 <- convert.z.score(x1); x2 <- convert.z.score(x2)
      x1$Contrast <- gsub("_", " ", tn1);x2$Contrast <- gsub("_", " ", tn2)
      colnames(x1) <- gsub("$", ".y", colnames(x1))
      colnames(x2) <- gsub("$", ".x", colnames(x2))
      x <- cbind(x1, x2)
      x <- add.sig.XY(x)
      x$Contrast <- sprintf("Y : %s\nX : %s", tn1, tn2)
      return(x)
    }
  )
)
ggplot(v.data) +
  geom_point(aes(mle.x, mle.y, colour = p.01, alpha = alpha), size = 0.5) +
  facet_wrap(~ Contrast) +
  scale_x_continuous(limits=mleRange)+scale_y_continuous(limits=mleRange) +
  ggtitle("4h") + scale_alpha_continuous(guide = FALSE) + labs(x="",y="")
```

```{r compare.4h_csv, echo=FALSE, eval=FALSE}
for (contrastName in unique(v.data$Contrast)){
  v.subset <- subset(addGENENAME(v.data), Contrast == contrastName)
  v.subset$Contrast <- gsub("\n", "-", v.subset$Contrast)
  v.subset <- addOpposite(v.subset)
  c.y <- gsub(" ", "_", unique(v.subset$Contrast.y))
  c.x <- gsub(" ", "_", unique(v.subset$Contrast.x))
  message(c.y); message(c.x)
  csvFile <- sprintf("SCDE_v8/compare.4h_%s-2h_%s.csv", c.y, c.x)
  message(csvFile)
  write.csv(v.subset, csvFile)
}
```

### 6h

```{r corPlot_6h, eval=TRUE, echo=FALSE, fig.height=7}
v.data <- do.call(
  "rbind",
  lapply(
    contrasts.contrasts.6h,
    function(contrastNames){
      cn1 <- contrastNames[1]; cn2 <- contrastNames[2]
      tn1<-gsub("[[:digit:]]h_(.*)-[[:digit:]]h_(.*)", "\\1-\\2", cn1)
      tn2<-gsub("[[:digit:]]h_(.*)-[[:digit:]]h_(.*)", "\\1-\\2", cn2)
      x1 <- scde.res[[cn1]][,resCols]; x2 <- scde.res[[cn2]][,resCols]
      x1 <- convert.z.score(x1); x2 <- convert.z.score(x2)
      x1$Contrast <- gsub("_", " ", tn1);x2$Contrast <- gsub("_", " ", tn2)
      colnames(x1) <- gsub("$", ".y", colnames(x1))
      colnames(x2) <- gsub("$", ".x", colnames(x2))
      x <- cbind(x1, x2)
      x <- add.sig.XY(x)
      x$Contrast <- sprintf("Y : %s\nX : %s", tn1, tn2)
      return(x)
    }
  )
)
ggplot(v.data) +
  geom_point(aes(mle.x, mle.y, colour = p.01, alpha = alpha), size = 0.5) +
  facet_wrap(~ Contrast) +
  scale_x_continuous(limits=mleRange)+scale_y_continuous(limits=mleRange) +
  ggtitle("6h") + scale_alpha_continuous(guide = FALSE) + labs(x="",y="")
```

```{r compare.6h_csv, echo=FALSE, eval=FALSE}
for (contrastName in unique(v.data$Contrast)){
  v.subset <- subset(addGENENAME(v.data), Contrast == contrastName)
  v.subset$Contrast <- gsub("\n", "-", v.subset$Contrast)
  v.subset <- addOpposite(v.subset)
  c.y <- gsub(" ", "_", unique(v.subset$Contrast.y))
  c.x <- gsub(" ", "_", unique(v.subset$Contrast.x))
  message(c.y); message(c.x)
  csvFile <- sprintf("SCDE_v8/compare.6h_%s-2h_%s.csv", c.y, c.x)
  message(csvFile)
  write.csv(v.subset, csvFile)
}
```
