---
title: "Sub-group analyses"
---
  
<!-- As HTML pages are built independently, large objects need to be
reimported from disk, and smaller ones quickly recomputed. -->
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(scater)
library(scran)
library(ggplot2)
library(scde)
EnsDb.Hsapiens.v79 <- EnsDb.Hsapiens.v79::EnsDb.Hsapiens.v79
sce.norm <- readRDS("rds/sce.norm.tSNE.rds")
```

# Stimulated cells at 6h post challenge

Subset to only cells at **6h** and challenged with bacteria 
(*i.e.*, exclude `Mock`):
  
```{r sce.6hStim_cells}
sce.6hStim <- sce.norm[,sce.norm$Time == '6h' & sce.norm$Infection != 'Mock']
dim(sce.6hStim)
```

Subset to only **endogenous** features detected with at least **10** counts
in at least **5** of the cells above:

```{r sce.6hStim_features}
sce.6hStim <- sce.6hStim[
  rowSums(counts(sce.6hStim) >= 10) >= 5 & !isSpike(sce.6hStim),]
dim(sce.6hStim)
```

## Highly variable genes (HVGs) {.tabset}

### Code

Use centred size factors to re-compute normalised expression values:

```{r sce.6hStim_geomMean}
exp(mean(log(sce.6hStim$size_factor)))
sce.6hStim <- normalise(sce.6hStim)
exp(mean(log(sce.6hStim$size_factor)))
```

Identify HVGs,
taking into account technical covariates `Plate` and `Lane`

```{r HVGs.6hStim}
dm <- model.matrix(~Plate+Lane, data = pData(sce.6hStim))
var.fit <- trendVar(
  sce.6hStim, assay = "norm_exprs", trend = "loess", design = dm,
  use.spikes = FALSE, span = 0.1
)
var.out <- decomposeVar(sce.6hStim, var.fit)
var.out <- var.out[order(var.out$p.value),]
HVGs.6hStim <- rownames(var.out)[which(var.out$FDR < 0.05)]
var.out <- cbind(
  GENENAME = mapIds(EnsDb.Hsapiens.v79,rownames(var.out),'GENENAME','GENEID'),
  var.out
)
length(HVGs.6hStim)
```

### Table

Table of `r length(HVGs.6hStim)` significant HVGs (*i.e.*, **FDR < 0.05**):
  
```{r datatable_HVGs.6hStim}
DT::datatable(var.out[HVGs.6hStim,])
```

## Clusters {.tabset}

### Code

Identify clusters using only HVGs identified above,
given a minimal cluster size of `r min(table(sce.6hStim$Group))`:
  
```{r quickCluster.6hStim}
group.6hStim <- droplevels(sce.6hStim$Group)
clusters <- quickCluster(
  sce.6hStim[HVGs.6hStim,],
  min.size = min(table(group.6hStim)))
names(clusters) <- sampleNames(sce.6hStim)
sce.6hStim$quickCluster <- clusters
sce.norm$quickCluster.6hStim <- 0
pData(sce.norm)[names(clusters),'quickCluster.6hStim'] <- clusters
table(clusters)
```

### Table

```{r quickCluster.6hStim_table, results='asis'}
pander::pandoc.table(
  as.data.frame(with(pData(sce.6hStim), table(Treatment, quickCluster)))
)
```

### tSNE

Overlay cluster membership and `Treatment` phenotype onto the tSNE coordinate
originally calculated from the complete data set:
  
```{r plotReducedDim_quickCluster.6hStim}
plotReducedDim(sce.norm,colour_by='quickCluster.6hStim',shape_by='Treatment')
```

## Differential expression {.tabset}

### Code

Reorder and group samples by `Treatment` (*i.e.*, `Infection:Status`)

```{r}
sce.6hStim <- sce.6hStim[,order(sce.6hStim$Treatment)]
```

Further subset count matrix to endogenous features
detected at sufficient expression levels:

```{r cd.6hStim}
cd.6hStim <- clean.counts(counts(sce.6hStim))
storage.mode(cd.6hStim) <- 'integer'
dim(cd.6hStim)
```

Fit error models using [scde](http://bioconductor.org/packages/scde):
  
```{r o.ifm.6hStim, eval=FALSE}
sg.6hStim <- droplevels(sce.6hStim$Treatment)
o.ifm.6hStim <- scde.error.models(cd.6hStim, sg.6hStim, verbose = 1)
```

```{r saveRDS_o.ifm.6hStim, include=FALSE, eval=FALSE}
saveRDS(o.ifm.6hStim, "rds/o.ifm.6hStim.rds")
```

```{r read_o.ifm.6hStim, include=FALSE}
o.ifm.6hStim <- readRDS("rds/o.ifm.6hStim.rds")
```

Remove particularly poor cells (*i.e.*, abnormal fits),
most commonly showing negative `corr.a`:
  
```{r validCells_6hStim}
valid.cells <- o.ifm.6hStim$corr.a > 0
table(valid.cells)
o.ifm.6hStim <- o.ifm.6hStim[valid.cells, ]
```

Estimate prior distribution for gene expression magnitudes:
  
```{r oPrior_6hStim}
o.prior.6hStim <- scde.expression.prior(
  models = o.ifm.6hStim, counts = cd.6hStim, show.plot = TRUE
)
```

### Call

Call differentially expressed features
in cluster **2** relative to cluster **1**
(*i.e.*, a positive fold-change is a higher expression in cluster **2**):
  
```{r scde.res_6hStim}
order.6hStim <- match(rownames(o.ifm.6hStim), colnames(sce.6hStim))
sg.newOrder <- as.factor(sce.6hStim$quickCluster[order.6hStim])
levels(sg.newOrder) <- c(2,1)
batch <- sce.6hStim$Plate[order.6hStim]
scde.res_6hStim <- scde.expression.difference(
  o.ifm.6hStim, cd.6hStim, o.prior.6hStim, groups = sg.newOrder, batch = batch,
  n.cores  =  4, verbose = 1
)
```

```{r}
saveRDS(scde.res_6hStim, "rds/scde.res_6hStim.rds")
```


### Table

Batch-adjusted DE calls:
  
```{r DEtables.6hStim, include=FALSE}
de.raw.6hStim <- scde.res_6hStim$results
de.adjusted.6hStim <- scde.res_6hStim$batch.adjusted
de.batch.6hStim <- scde.res_6hStim$batch.effect
de.raw.6hStim <- de.raw.6hStim[order(abs(de.raw.6hStim$Z), decreasing = TRUE),]
de.adjusted.6hStim <- de.adjusted.6hStim[order(abs(de.adjusted.6hStim$Z), decreasing = TRUE),]
de.batch.6hStim <- de.batch.6hStim[order(abs(de.batch.6hStim$Z), decreasing = TRUE),]
de.raw.6hStim <- cbind(
  GENENAME=mapIds(EnsDb.Hsapiens.v79,rownames(de.raw.6hStim),"GENENAME","GENEID"),
  de.raw.6hStim
)
de.adjusted.6hStim <- cbind(
  GENENAME=mapIds(EnsDb.Hsapiens.v79,rownames(de.adjusted.6hStim),"GENENAME","GENEID"),
  de.adjusted.6hStim
)
de.batch.6hStim <- cbind(
  GENENAME=mapIds(EnsDb.Hsapiens.v79,rownames(de.batch.6hStim),"GENENAME","GENEID"),
  de.batch.6hStim
)
```

```{r DEtables.6hStim_csv}
DT::datatable(head(de.adjusted.6hStim[rev(order(de.adjusted.6hStim$Z)),], 500))
```

```{r, include=FALSE}
write.csv(de.raw.6hStim, 'SCDE/raw_6h_cluster2-clusters1.csv')
write.csv(de.adjusted.6hStim, 'SCDE/adjusted_6h_cluster2-cluster1.csv')
write.csv(de.batch.6hStim, 'SCDE/batch_6h_cluster1-clusters2.csv')
```

Visualise normalised expression data for the top differentially expressed
feature:
  
```{r topDE4hVSuninfected}
geneId <- rownames(de.adjusted.6hStim)[1]
geneName <- de.adjusted.6hStim[1,'GENENAME']
geneExpr <- norm_exprs(sce.6hStim)[geneId,]
sampleAes <- pData(sce.6hStim)[,c("Infection","Status","quickCluster")]
sampleAes[,3] <- as.factor(sampleAes[,3])
ggData <- cbind(norm_exprs = geneExpr, sampleAes)
ggplot(ggData, aes(quickCluster, norm_exprs)) +
  geom_violin(draw_quantiles = c(0.25,0.5,0.75)) + geom_jitter(width = 0.1) +
  facet_grid(Infection ~ Status) +
  ggtitle(sprintf("%s (symbol: %s)", geneId, geneName))
```
