---
title: "Sub-group analyses"
---
  
<!-- As HTML pages are built independently, large objects need to be
reimported from disk, and smaller ones quickly recomputed. -->
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(scater)
library(scran)
library(ggplot2)
library(scde)
EnsDb.Hsapiens.v79 <- EnsDb.Hsapiens.v79::EnsDb.Hsapiens.v79
sce.norm <- readRDS("rds/sce.norm.tSNE.rds")
scde.res <- readRDS("rds/13_scde.clusters.rds")
```

# Prepare data

```{r}
sce.endo <- sce.norm[!isSpike(sce.norm),]
dim(sce.endo)
```

# Identify clusters {.tabset}

## 2h

Subset data to stimulated cells at 2h:

```{r sce.2h.stim}
sce.2h.stim <- sce.endo[,sce.endo$Time == "2h" & !sce.endo$Infection == "Mock"]
dim(sce.2h.stim)
```

Subset data to genes sufficiently detected at 2h
(*e.g.*, 10 counts in at least 10 cells):

```{r sce.2h.detected}
sce.2h.stim <- sce.2h.stim[rowSums(counts(sce.2h.stim) >= 10) >= 10,]
dim(sce.2h.stim)
```

```{r clusters.2h}
sce.2h.stim$Treatment <- droplevels(sce.2h.stim$Treatment)
clusters <- quickCluster(
  sce.2h.stim,
  min.size = min(table(sce.2h.stim$Treatment))
)
sce.2h.stim$quickCluster <- clusters
sce.endo$quickCluster.2h <- 0
pData(sce.endo)[sampleNames(sce.2h.stim),"quickCluster.2h"] <- clusters
```

```{r table.quickCluster.2h}
table(sce.endo$quickCluster.2h)
```

```{r plotReducedDim.2h}
plotReducedDim(sce.endo,colour_by='quickCluster.2h',shape_by='Treatment')
```

Cross-tabulated counts of cells by cluster and experimental treatment:

```{r table.2h, results='asis', echo=FALSE}
pander::pandoc.table(
  with(pData(sce.2h.stim), table(Treatment, quickCluster))
)
```

## 4h

Subset data to stimulated cells at 4h:

```{r sce.4h.stim}
sce.4h.stim <- sce.endo[,sce.endo$Time == "4h" & !sce.endo$Infection == "Mock"]
dim(sce.4h.stim)
```

Subset data to genes sufficiently detected at 4h
(*e.g.*, 10 counts in at least 10 cells):

```{r sce.4h.detected}
sce.4h.stim <- sce.4h.stim[rowSums(counts(sce.4h.stim) >= 10) >= 10,]
dim(sce.4h.stim)
```

```{r clusters.4h}
sce.4h.stim$Treatment <- droplevels(sce.4h.stim$Treatment)
clusters <- quickCluster(
  sce.4h.stim,
  min.size = min(table(sce.4h.stim$Treatment))
)
sce.4h.stim$quickCluster <- clusters
sce.endo$quickCluster.4h <- 0
pData(sce.endo)[sampleNames(sce.4h.stim),"quickCluster.4h"] <- clusters
```

```{r table.quickCluster.4h}
table(sce.endo$quickCluster.4h)
```

```{r plotReducedDim.4h}
plotReducedDim(sce.endo,colour_by='quickCluster.4h',shape_by='Treatment')
```

Cross-tabulated counts of cells by cluster and experimental treatment:

```{r table.4h, results='asis', echo=FALSE}
pander::pandoc.table(
  with(pData(sce.4h.stim), table(Treatment, quickCluster))
)
```

```{r, echo=FALSE}
clusterTable <- with(pData(sce.4h.stim), table(Treatment, quickCluster))
clusterTable <- as.data.frame(clusterTable)
ggplot(clusterTable, aes(Treatment, Freq, fill=quickCluster)) + geom_col() +
  facet_grid(~quickCluster) + ggtitle("4h clusters") +
  theme_light() +
  theme(
    axis.text.x = element_text(angle=45, vjust=1, hjust=1)
    )
```

## 6h

Subset data to stimulated cells at 6h:

```{r sce.6h.stim}
sce.6h.stim <- sce.endo[,sce.endo$Time == "6h" & !sce.endo$Infection == "Mock"]
dim(sce.6h.stim)
```

Subset data to genes sufficiently detected at 6h
(*e.g.*, 10 counts in at least 10 cells):

```{r sce.6h.detected}
sce.6h.stim <- sce.6h.stim[rowSums(counts(sce.6h.stim) >= 10) >= 10,]
dim(sce.6h.stim)
```

```{r clusters_6h}
sce.6h.stim$Treatment <- droplevels(sce.6h.stim$Treatment)
clusters <- quickCluster(
  sce.6h.stim,
  min.size = min(table(sce.6h.stim$Treatment))
)
sce.6h.stim$quickCluster <- clusters
sce.endo$quickCluster.6h <- 0
pData(sce.endo)[sampleNames(sce.6h.stim),"quickCluster.6h"] <- clusters
```

```{r table_quickCluster.6h}
table(sce.endo$quickCluster.6h)
```

```{r plotReducedDim_6h}
plotReducedDim(sce.endo,colour_by='quickCluster.6h',shape_by='Treatment')
```

Cross-tabulated counts of cells by cluster and experimental treatment:

```{r table.6h, results='asis', echo=FALSE}
pander::pandoc.table(
  with(pData(sce.6h.stim), table(Treatment, quickCluster))
)
```

```{r, echo=FALSE}
clusterTable <- with(pData(sce.6h.stim), table(Treatment, quickCluster))
clusterTable <- as.data.frame(clusterTable)
ggplot(clusterTable, aes(Treatment, Freq, fill=quickCluster)) + geom_col() +
  facet_grid(~quickCluster) + ggtitle("6h clusters") +
  theme_light() +
  theme(
    axis.text.x = element_text(angle=45, vjust=1, hjust=1)
    )
```

# Differential expression {.tabset}

## Setup

Load previously computed error models,
using experimental group information:

```{r o.ifm}
o.ifm <- readRDS("rds/o.ifm.rds")
```

Reorder count matrix to match the precomputed error models:

```{r o.ifm.order}
o.ifm.order <- match(rownames(o.ifm), sampleNames(sce.norm))
sce.ifm <- sce.endo[,o.ifm.order]
cd.ifm <- counts(sce.ifm)[!isSpike(sce.ifm),]; storage.mode(cd.ifm) <- "integer"
```

```{r o.ifm.order_stopifnot, include=FALSE}
stopifnot(rownames(o.ifm) == sampleNames(sce.ifm))
stopifnot(rownames(o.ifm) == colnames(cd.ifm))
```

Estimate prior distribution for gene expression magnitudes:

```{r oPrior}
o.prior <- scde.expression.prior(
  models = o.ifm, counts = cd.ifm, length.out = 400, show.plot = TRUE
)
```

Let us also define:

* a list to store the result tables returned by *scde*

```{r scde.res_init}
scde.res <- list()
```

* a function used to convert the Z-score computed by *scde* to a empirical
P-value ([reference](https://www.biostars.org/p/17227/)):

```{r convert.z.score}
convert.z.score <- function(x, one.sided = NULL) {
  z <- x$Z
  if(is.null(one.sided)) {
    pval = pnorm(-abs(z));
    pval = 2 * pval
  } else if(one.sided=="-") {
    pval = pnorm(z);
  } else {
    pval = pnorm(-z);
  }
    x <- cbind(
      x,
      p.value = pval
  )
  return(x);
}   
```

* a function used to annotate the tables of results returned by *scde*:

```{r addGENENAME}
addGENENAME <- function(x){
  x <- cbind(
    GENENAME=mapIds(EnsDb.Hsapiens.v79, rownames(x), 'GENENAME', 'GENEID'),
    x
  )
  return(x)
}
```

* a function to order results by decreasing absolute
corrected Z-score and subsequently absolute Z-score:

```{r orderResults}
orderResults <- function(x){
  x <- x[with(x, order(abs(cZ), abs(Z), decreasing = TRUE)),]
  return(x)
}
```

* a function to visualise *scran*-normalised gene expression for a given
gene:

Let us also define a function that displays normalised (*scran*) expression
data in each group:

```{r normExprsByName}
normExprsById <- function(geneId){
  geneName <- subset(fData(sce.norm), gene_id == geneId, "gene_name", drop = TRUE)
  gdata <- data.frame(
    norm_exprs = norm_exprs(sce.norm)[geneId,],
    pData(sce.norm)[,c("Infection","Status","Time")],
    row.names = sampleNames(sce.norm)
  )
  ggplot(gdata, aes(Infection, norm_exprs)) + 
    geom_violin() + geom_jitter(width = 0.1) +
    facet_grid(Time ~ Status) +
    ggtitle(sprintf("%s - %s", geneId, geneName))
}
```


## 2h (clusters 1 vs {2 and 3})

Collapse clusters 2 and 3:

```{r 2h_1vs23}
groupTarget <- "1"; groupRef <- c("2","3")
sg.test <- factor(sce.endo$quickCluster.2h,levels=c(groupTarget,groupRef))
groupRef <- paste0(groupRef, collapse = "")
levels(sg.test) <- c(groupTarget, rep(groupRef, nlevels(sg.test) - 1))
names(sg.test) <- sampleNames(sce.ifm); summary(sg.test)
contrastName <- sprintf("2h_cluster%s-cluster%s", groupTarget, groupRef)
```

```{r scde_2h_1vs23, eval=FALSE}
scde.res[[contrastName]] <- scde.expression.difference(
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
topAdjusted <-
  head(addGENENAME(orderResults(scde.res[[contrastName]]$batch.adjusted)), 1)
topAdjusted
```

```{r, eval=FALSE, echo=FALSE}
scde.test.gene.expression.difference(
  gene = rownames(topAdjusted), 
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
normExprsById(rownames(topAdjusted))
```

## 2h (clusters 2 vs {1 and 3})

Collapse clusters 1 and 3:

```{r 2h_2vs13}
groupTarget <- "2"; groupRef <- c("1","3")
sg.test <- factor(sce.endo$quickCluster.2h,levels=c(groupTarget,groupRef))
groupRef <- paste0(groupRef, collapse = "")
levels(sg.test) <- c(groupTarget, rep(groupRef, nlevels(sg.test) - 1))
names(sg.test) <- sampleNames(sce.ifm); summary(sg.test)
contrastName <- sprintf("2h_cluster%s-cluster%s", groupTarget, groupRef)
```

```{r scde_2h_2vs13, eval=FALSE}
scde.res[[contrastName]] <- scde.expression.difference(
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
topAdjusted <-
  head(addGENENAME(orderResults(scde.res[[contrastName]]$batch.adjusted)), 1)
topAdjusted
```

```{r, eval=FALSE, echo=FALSE}
scde.test.gene.expression.difference(
  gene = rownames(topAdjusted), 
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
normExprsById(rownames(topAdjusted))
```

## 2h (clusters 3 vs {1 and 2})

Collapse clusters 1 and 2:

```{r 2h_3vs12}
groupTarget <- "3"; groupRef <- c("1","2")
sg.test <- factor(sce.endo$quickCluster.2h,levels=c(groupTarget,groupRef))
groupRef <- paste0(groupRef, collapse = "")
levels(sg.test) <- c(groupTarget, rep(groupRef, nlevels(sg.test) - 1))
names(sg.test) <- sampleNames(sce.ifm); summary(sg.test)
contrastName <- sprintf("2h_cluster%s-cluster%s", groupTarget, groupRef)
```

```{r scde_2h, eval=FALSE}
scde.res[[contrastName]] <- scde.expression.difference(
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
topAdjusted <-
  head(addGENENAME(orderResults(scde.res[[contrastName]]$batch.adjusted)), 1)
topAdjusted
```

```{r, eval=FALSE, echo=FALSE}
scde.test.gene.expression.difference(
  gene = rownames(topAdjusted), 
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
normExprsById(rownames(topAdjusted))
```

## 4h (clusters 1 vs 2)

```{r 4h_1vs2}
groupTarget <- "1"; groupRef <- "2"
sg.test <- factor(sce.ifm$quickCluster.4h, levels = c(groupTarget, groupRef))
names(sg.test) <- sampleNames(sce.ifm); summary(sg.test)
contrastName <- sprintf("4h_cluster%s-cluster%s", groupTarget, groupRef)
```

```{r scde_4h, eval=FALSE}
scde.res[[contrastName]] <- scde.expression.difference(
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
topAdjusted <-
  head(addGENENAME(orderResults(scde.res[[contrastName]]$batch.adjusted)), 1)
topAdjusted
```

```{r, eval=FALSE, echo=FALSE}
scde.test.gene.expression.difference(
  gene = rownames(topAdjusted), 
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
normExprsById(rownames(topAdjusted))
```

## 6h (clusters 1 vs 2)

```{r 6h_1vs2}
groupTarget <- "1"; groupRef <- "2"
sg.test <- factor(sce.ifm$quickCluster.6h, levels = c(groupTarget, groupRef))
names(sg.test) <- sampleNames(sce.ifm); summary(sg.test)
contrastName <- sprintf("6h_cluster%s-cluster%s", groupTarget, groupRef)
```

```{r scde_6h, eval=FALSE}
scde.res[[contrastName]] <- scde.expression.difference(
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
topAdjusted <-
  head(addGENENAME(orderResults(scde.res[[contrastName]]$batch.adjusted)), 1)
topAdjusted
```

```{r, eval=FALSE, echo=FALSE}
scde.test.gene.expression.difference(
  gene = rownames(topAdjusted), 
  o.ifm, cd.ifm, o.prior, groups = sg.test, batch = sce.ifm$Plate,
  n.cores = 4, verbose = 1
)
```

```{r, eval=FALSE, echo=FALSE}
normExprsById(rownames(topAdjusted))
```

## Post-processing

<!-- Post processing -->

```{r, eval=FALSE, echo=FALSE}
saveRDS(scde.res, "rds/13_scde.clusters.rds")
```

```{r, eval=FALSE, echo=FALSE}
for (contrastName in names(scde.res)){
  scde.raw <- scde.res[[contrastName]]$results
  scde.raw <- convert.z.score(addGENENAME(orderResults(scde.raw)))
  csv.raw <- sprintf("out_13/raw.%s.csv", contrastName)
  write.csv(scde.raw, csv.raw)
}
for (contrastName in names(scde.res)){
  scde.adjusted <- scde.res[[contrastName]]$batch.adjusted
  scde.adjusted <- convert.z.score(addGENENAME(orderResults(scde.adjusted)))
  csv.raw <- sprintf("out_13/adjusted.%s.csv", contrastName)
  write.csv(scde.adjusted, csv.raw)
}
for (contrastName in names(scde.res)){
  scde.batch <- scde.res[[contrastName]]$batch.effect
  scde.batch <- convert.z.score(addGENENAME(orderResults(scde.batch)))
  csv.raw <- sprintf("out_13/batch.%s.csv", contrastName)
  write.csv(scde.batch, csv.raw)
}
```


