---
title: "Differential expression between unsupervised clusters using *scde*"
bibliography:
  bibtex.bib
---

<!-- As HTML pages are built independently, large objects need to be
reimported from disk, and smaller ones quickly recomputed. -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(scater)
library(scran)
library(ggplot2)
library(scde)
library(ensembldb)
library(ComplexHeatmap)
library(dplyr)
library(ggrepel)
library(RColorBrewer)
library(goseq)
sce.norm <- readRDS("rds/sce.norm.tSNE.rds")
o.ifm.2h <- readRDS("rds/o.ifm.2h_v8.rds")
o.ifm.4h <- readRDS("rds/o.ifm.4h_v8.rds")
o.ifm.6h <- readRDS("rds/o.ifm.6h_v8.rds")
scde.res <- readRDS("rds/13_cluster_scde.res_v2.rds")
goseq.res <- readRDS("rds/13_cluster_goseq.res_v2.rds")
scde.groupRes <- readRDS("rds/scde.res_v8.rds")
```

<!-- Cleaner code based on 08_SCDE_v8
Remove ERCC spike-ins
!! Split cells by time
!! Retain genes detected at each time (10 counts, 5 cells, 1 group)
Identify cell clusters @ time
<!-- Error model: {cells @ time; genes @ time} <- same as 08_SCDE
Prior @ time
DE: {group on all cells with NA; no batch correction; P-value < 0.01}
-->

# Prepare data

**Note:**
for this analysis, the data is prepared identically to the
[differential expression between experimental groups using *scde*](08_SCDE_v8.html):

* ERCC spike-in features are removed,
* Cells in the data set is subsetted by time point,
* Genes in the data set are subsetted for those detected at sufficient levels
  within each time point,
* Error models are computed on the subsetted data sets
* Differentially expressed genes are called within the subsetted data sets
  using a cutoff *P*-value less than **0.01**

[Jump](#identifyClusters) to the section where this analysis of
unsupervised clusters diverges from the differential expression
between supervised experimental groups.

## Remove spike-in features

The analysis starts with a matrix of read counts,
filtered based on gene and cell requirements.
In this case, let us use the raw numbers of reads mapped to each endogenous
feature, recast as `integer` type:

```{r table_isSpike}
table(isSpike(sce.norm))
```

```{r sce.endo}
sce.endo <- sce.norm[!isSpike(sce.norm),]
```

```{r sce.endo_show, echo=FALSE}
dim(sce.endo)
```

## Subset data by time point {.tabset}

Considering the demonstrated impact on the `Time` factor on gene expression
profiles, and the fact that differential expression will *not* be assessed
between time points, individual time points will be processed separately for
the remainder of this *scde* analysis.

### 2h

```{r sce.2h}
sce.2h <- sce.endo[,sce.endo$Time == '2h']
dim(sce.2h)
```

### 4h

```{r sce.4h}
sce.4h <- sce.endo[,sce.endo$Time == '4h']
dim(sce.4h)
```

### 6h

```{r sce.6h}
sce.6h <- sce.endo[,sce.endo$Time == '6h']
dim(sce.6h)
```

## Filter count matrices {.tabset}

The analysis starts with a matrix of read counts,
filtered based on gene and cell requirements.

In this case, *for each time point*,
let us retain only features detected with at least **10** counts in at least
**10** cells of any of the five experimental groups,
use the raw numbers of reads mapped to each endogenous
feature, recast as `integer` type.

Let us first define a function to apply the detection cutoff:

```{r filterCounts}
filterCounts <- function(m, counts = 10, cells = 10){
  apply(m, 1, function(e){
    return(sum(e >= counts) >= cells)
  })
}
```

### 2h

```{r cleanCounts_2h}
sg.2h <- droplevels(sce.2h$Group)
keep.2h <- filterCounts(counts(sce.2h))
cd.2h <- counts(sce.2h)[keep.2h,]; storage.mode(cd.2h) <- 'integer'
```

```{r cd.2h_show, echo=FALSE}
dim(cd.2h)
```

### 4h

```{r cleanCounts_4h}
sg.4h <- droplevels(sce.4h$Group)
keep.4h <- filterCounts(counts(sce.4h))
cd.4h <- counts(sce.4h)[keep.4h,]; storage.mode(cd.4h) <- 'integer'
```

```{r cd.4h_show, echo=FALSE}
dim(cd.4h)
```

### 6h

```{r cleanCounts_6h}
sg.6h <- droplevels(sce.6h$Group)
keep.6h <- filterCounts(counts(sce.6h))
cd.6h <- counts(sce.6h)[keep.6h,]; storage.mode(cd.6h) <- 'integer'
```

```{r cd.6h_show, echo=FALSE}
dim(cd.6h)
```

# Error models {.tabset}

Here, we fit the error models on which all subsequent calculations
will rely. The fitting process relies on a subset of robust genes that are
detected in multiple cross-cell comparisons. Here we supply the
`groups` argument, so that the error models for each
experimental group of cells are fit independently.
If the groups argument is omitted, the models would be fit using a common set.

## 2h

```{r o.ifm.2h, eval=FALSE}
o.ifm.2h <- scde.error.models(
  counts = cd.2h, groups = sg.2h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.2h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.2h, "rds/o.ifm.2h_v8.rds")
write.csv(o.ifm.2h, "SCDE_v8/o.ifm.2h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.2h}
valid.cells <- o.ifm.2h$corr.a > 0; table(valid.cells)
o.ifm.2h <- o.ifm.2h[valid.cells, ]
dim(o.ifm.2h)
```

## 4h

```{r o.ifm.4h, eval=FALSE}
o.ifm.4h <- scde.error.models(
  counts = cd.4h, groups = sg.4h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.4h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.4h, "rds/o.ifm.4h_v8.rds")
write.csv(o.ifm.4h, "SCDE_v8/o.ifm.4h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.4h}
valid.cells <- o.ifm.4h$corr.a > 0; table(valid.cells)
o.ifm.4h <- o.ifm.4h[valid.cells, ]
dim(o.ifm.4h)
```

## 6h

```{r o.ifm.6h, eval=FALSE}
o.ifm.6h <- scde.error.models(
  counts = cd.6h, groups = sg.6h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.6h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.6h, "rds/o.ifm.6h_v8.rds")
write.csv(o.ifm.6h, "SCDE_v8/o.ifm.6h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.6h}
valid.cells <- o.ifm.6h$corr.a > 0; table(valid.cells)
o.ifm.6h <- o.ifm.6h[valid.cells, ]
dim(o.ifm.6h)
```

# Reorder counts to match error models {.tabset}

The `scde.error.models` produces error models with cells reordered 
by experimental group.
For clarity, let us prepare a `SCESet` and count matrix that match this order:

## 2h

```{r sce.ifm.2h}
sce.ifm.2h <- sce.2h[,rownames(o.ifm.2h)]
cd.ifm.2h <- counts(sce.ifm.2h); storage.mode(cd.ifm.2h) <- "integer"
```

## 4h

```{r sce.ifm.4h}
sce.ifm.4h <- sce.4h[,rownames(o.ifm.4h)]
cd.ifm.4h <- counts(sce.ifm.4h); storage.mode(cd.ifm.4h) <- "integer"
```

## 6h

```{r sce.ifm.6h}
sce.ifm.6h <- sce.6h[,rownames(o.ifm.6h)]
cd.ifm.6h <- counts(sce.ifm.6h); storage.mode(cd.ifm.6h) <- "integer"
```

# Prior distribution for gene expression magnitudes {.tabset}

Finally, we need to define an expression magnitude prior for the genes.
Its main function, however, is to define a grid of expression magnitude values
on which the numerical calculations will be carried out.

## 2h

```{r o.prior.2h}
o.prior.2h <- scde.expression.prior(
  models = o.ifm.2h, counts = cd.ifm.2h, show.plot = TRUE
)
```

## 4h

```{r o.prior.4h}
o.prior.4h <- scde.expression.prior(
  models = o.ifm.4h, counts = cd.ifm.4h, show.plot = TRUE
)
```

## 6h

```{r o.prior.6h}
o.prior.6h <- scde.expression.prior(
  models = o.ifm.6h, counts = cd.ifm.6h, show.plot = TRUE
)
```

# Identify clusters {#identifyClusters .tabset}

## 2h

```{r clusters.2h}
clusters <- quickCluster(sce.ifm.2h,min.size=min(table(sce.ifm.2h$Treatment)))
sce.ifm.2h$quickCluster <- clusters
qc.2h <- rep(NA, ncol(sce.endo)); names(qc.2h) <- sampleNames(sce.endo)
qc.2h[sampleNames(sce.ifm.2h)] <- as.numeric(levels(clusters)[clusters])
sce.endo$quickCluster.2h <- as.factor(qc.2h)
```

## 4h

```{r clusters.4h}
clusters <- quickCluster(sce.ifm.4h,min.size=min(table(sce.ifm.4h$Treatment)))
sce.ifm.4h$quickCluster <- clusters
qc.4h <- rep(NA, ncol(sce.endo)); names(qc.4h) <- sampleNames(sce.endo)
qc.4h[sampleNames(sce.ifm.4h)] <- as.numeric(levels(clusters)[clusters])
sce.endo$quickCluster.4h <- as.factor(qc.4h)
```

## 6h

```{r clusters.6h}
clusters <- quickCluster(sce.ifm.6h,min.size=min(table(sce.ifm.6h$Treatment)))
sce.ifm.6h$quickCluster <- clusters
qc.6h <- rep(NA, ncol(sce.endo)); names(qc.6h) <- sampleNames(sce.endo)
qc.6h[sampleNames(sce.ifm.6h)] <- as.numeric(levels(clusters)[clusters])
sce.endo$quickCluster.6h <- as.factor(qc.6h)
```

```{r, echo=FALSE, eval=FALSE}
saveRDS(sce.endo, "rds/13_sce.endo_clusters.rds")
```


# tSNE {.tabset}

Let us first define:

* a set of colours used to indicate *cluster membership*,
  *Infection* and *Status* phenotypes consistently in the upcoming sections:

```{r col.sets}
col.set3 <- brewer.pal(12, "Set3")
col.cluster <- col.set3[c(1,12,3,4,9)]; names(col.cluster) <- c(1:3,0,"NA")
col.status<-col.set3[c(3,5:6)];names(col.status)<-levels(sce.norm$Status)
col.infection<-col.set3[c(3,8,10)];names(col.infection)<-levels(sce.norm$Infection)
col.time<-col.set3[c(2,7:8)];names(col.time)<-levels(sce.norm$Time)
```

## 2h

```{r plotReducedDim.2h, echo=FALSE}
tmpGG <- data.frame(
  reducedDimension(sce.endo),
  Cluster = sce.endo$quickCluster.2h,
  Treatment = gsub("_", " ", sce.endo$Treatment)
)
ggplot(tmpGG, aes(X1, X2, shape = Treatment, colour = Cluster)) +
  geom_point() +
  theme_bw() +
  labs(x="Dimension 1",y="Dimension 2",title="2h") +
  scale_color_manual(values = col.cluster, na.value = col.cluster["NA"])
```

```{r reducedDimension_pdf.2h, include=FALSE}
ggsave("13_out/tSNE_clusters_2h.pdf", width = 6, height = 4)
```

Cross-tabulated counts of cells by cluster and experimental treatment:

```{r table.2h, results='asis', echo=FALSE}
pander::pandoc.table(
  with(pData(sce.endo), table(Treatment, quickCluster.2h))
)
```

```{r clusterBarplot.2h, echo=FALSE}
tmpGG <- ggplot(pData(sce.ifm.2h)) +
  stat_count(aes(x = Treatment, fill = quickCluster)) +
  facet_grid(~ quickCluster) +
  scale_fill_manual(values = col.cluster, na.value = col.cluster["NA"]) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,face="bold")
  ) +
  labs(y = "Cells", fill = "Cluster")
ggsave("13_out/cluster_count_2h.pdf", width = 9, height = 4)
```

## 4h

```{r plotReducedDim.4h, echo=FALSE}
tmpGG <- data.frame(
  reducedDimension(sce.endo),
  Cluster = sce.endo$quickCluster.4h,
  Treatment = gsub("_", " ", sce.endo$Treatment),
  label = ifelse(
    sce.endo$Treatment == "LT2_infected" & sce.endo$quickCluster.4h == 2, 2, NA
  )
)
ggplot(tmpGG, aes(X1, X2, shape = Treatment, colour = Cluster)) +
  geom_point() +
  theme_bw() +
  # theme(legend.position = "bottom", legend.box = "vertical") +
  labs(x="Dimension 1",y="Dimension 2",title="4h") +
  scale_color_manual(values = col.cluster, na.value = col.cluster["NA"]) #+
  #geom_point(aes(colour = "red", size = as.numeric(label)), colour = "red", alpha = 0.25) # highlight
```

```{r reducedDimension_pdf.4h, include=FALSE}
ggsave("13_out/tSNE_clusters_4h.pdf", width = 6, height = 4)
```

Cross-tabulated counts of cells by cluster and experimental treatment:

```{r table.4h, results='asis', echo=FALSE}
pander::pandoc.table(
  with(pData(sce.endo), table(Treatment, quickCluster.4h))
)
```

```{r clusterBarplot.4h, echo=FALSE}
tmpGG <- ggplot(pData(sce.ifm.4h)) +
  stat_count(aes(x = Treatment, fill = quickCluster)) +
  facet_grid(~ quickCluster) +
  scale_fill_manual(values = col.cluster, na.value = col.cluster["NA"]) +
  theme_minimal() + theme(
    axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,face="bold")
  ) + labs(y = "Cells", fill = "Cluster")
ggsave("13_out/cluster_count_4h.pdf", width = 9, height = 4)
```

## 6h

```{r plotReducedDim.6h, echo=FALSE}
tmpGG <- data.frame(
  reducedDimension(sce.endo),
  Cluster = sce.endo$quickCluster.6h,
  Treatment = gsub("_", " ", sce.endo$Treatment)
)
ggplot(tmpGG, aes(X1, X2, shape = Treatment, colour = Cluster)) +
  geom_point() + theme_bw() + labs(x="Dimension 1",y="Dimension 2",title="6h") +
  scale_color_manual(values = col.cluster, na.value = col.cluster["NA"])
```

```{r reducedDimension_pdf.6h, include=FALSE}
ggsave("13_out/tSNE_clusters_6h.pdf", width = 6, height = 4)
```

Cross-tabulated counts of cells by cluster and experimental treatment:

```{r table.6h, results='asis', echo=FALSE}
pander::pandoc.table(
  with(pData(sce.endo), table(Treatment, quickCluster.6h))
)
```

```{r clusterBarplot.6h, echo=FALSE}
tmpGG <- ggplot(pData(sce.ifm.6h)) +
  stat_count(aes(x = Treatment, fill = quickCluster)) +
  facet_grid(~ quickCluster) +
  scale_fill_manual(values = col.cluster, na.value = col.cluster["NA"]) +
  theme_minimal() + theme(
    axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,face="bold")
  ) + labs(y = "Cells", fill = "Cluster")
ggsave("13_out/cluster_count_6h.pdf", width = 9, height = 4)
```

# Differential expression {#DE}

## Setup

Let us first define:

* a list to store the result tables returned by *scde*

```{r scde.res_init, eval=FALSE}
scde.res <- list()
```

* a function used to convert the Z-score computed by *scde* to a empirical
P-value ([reference](https://www.biostars.org/p/17227/)):

```{r convert.z.score}
convert.z.score <- function(x, one.sided = NULL) {
  z <- x$Z
  if(is.null(one.sided)) {
    pval = pnorm(-abs(z));
    pval = 2 * pval
  } else if(one.sided=="-") {
    pval = pnorm(z);
  } else {
    pval = pnorm(-z);
  }
    x <- cbind(
      x,
      p.value = pval
  )
  return(x);
}   
```

* a function used to annotate the tables of results returned by *scde*:

```{r addGENENAME}
addGENENAME <- function(x){
  x <- cbind(
    GENENAME = with(fData(sce.endo), gene_name[match(rownames(x), gene_id)]),
    x
  )
  return(x)
}
```

* a function to order results by decreasing absolute Z-score:

```{r orderResults}
orderResults <- function(x){
  x <- x[with(x, order(abs(Z), decreasing = TRUE)),]
  return(x)
}
```

* a function to visualise *scran*-normalised gene expression for a given
gene:

Let us also define a function that displays normalised (*scran*) expression
data in each group:

```{r normExprsByName}
normExprsById <- function(geneId){
  geneName <- subset(fData(sce.endo),gene_id==geneId,"gene_name",drop=TRUE)
  gdata <- data.frame(
    norm_exprs = norm_exprs(sce.endo)[geneId,],
    pData(sce.endo)[,c("Infection","Status","Time")],
    row.names = sampleNames(sce.endo)
  )
  ggplot(gdata, aes(Infection, norm_exprs)) + 
    geom_violin() + geom_jitter(width = 0.1) +
    facet_grid(Time ~ Status) +
    ggtitle(sprintf("%s - %s", geneId, geneName))
}
```

* functions to visualise *scde* estimate of expression and
  differential expression
  for a given gene (identifier)
  in a given contrast of clusters
  within a specific time point:
  
```{r single.scde}
single.scde.4h <- function(gene, groupTarget, groupRef){
  sg.test <- factor(sce.ifm.4h$quickCluster, levels = c(groupTarget, groupRef))
  scde.test.gene.expression.difference(
    gene, o.ifm.4h, cd.ifm.4h, o.prior.4h, sg.test,
    n.cores = 4, verbose = 1
  )
}
single.scde.6h <- function(gene, groupTarget, groupRef){
  sg.test <- factor(sce.ifm.6h$quickCluster, levels = c(groupTarget, groupRef))
  scde.test.gene.expression.difference(
    gene, o.ifm.6h, cd.ifm.6h, o.prior.6h, sg.test,
    n.cores = 4, verbose = 1
  )
}
```

* various significance levels:

```{r volcano.sig}
sig.levels <- c(0.05, 0.01)
volcano.sig <- data.frame(
  P = sig.levels,
  level = as.character(sig.levels)
)
```

* a function to visualise *scde* differential expression statistics, and
  return a table with those statistics augmented by a an empirical *P* value
  computed from the *Z* score returned by *scde*:

```{r volcano.mle}
volcano.mle <- function(x, sub = NULL){
  varName <- deparse(substitute(x))
  x <- convert.z.score(x)
  gg <- ggplot(x, aes(mle, -log10(p.value))) +
    geom_point(aes(colour = (cZ != 0))) +
    geom_hline(aes(yintercept=-log10(p.value),linetype=level),volcano.sig) +
    ggtitle(varName, sub)
  print(gg)
  return(x)
}
```

## Contrasts {.tabset}

### 4h (2 vs 1)

```{r contrasts.4h_run, eval=FALSE}
groupTarget <- 2; groupRef <- 1
sg.test <- factor(sce.ifm.4h$quickCluster, levels = c(groupTarget, groupRef))
names(sg.test) <- sampleNames(sce.4h); summary(sg.test)
contrastName <- sprintf("4h_cluster%s-cluster%s", groupTarget, groupRef); message(contrastName)
scde.res[[contrastName]] <- scde.expression.difference(
    o.ifm.4h, cd.4h, o.prior.4h, sg.test, n.cores = 4, verbose = 1
  )
```

### 6h (2 vs 1)

```{r contrasts.6h_run, eval=FALSE}
groupTarget <- 2; groupRef <- 1
sg.test <- factor(sce.ifm.6h$quickCluster, levels = c(groupTarget, groupRef))
names(sg.test) <- sampleNames(sce.6h); summary(sg.test)
contrastName <- sprintf("6h_cluster%s-cluster%s", groupTarget, groupRef); message(contrastName)
scde.res[[contrastName]] <- scde.expression.difference(
    o.ifm.6h, cd.6h, o.prior.6h, sg.test, n.cores = 4, verbose = 1
  )
```

```{r scde.table_csv, echo=FALSE, eval=FALSE}
saveRDS(scde.res, "rds/13_cluster_scde.res_v2.rds")
for (contrastName in names(scde.res)){
  scde.table <- scde.res[[contrastName]]
  scde.table <- convert.z.score(addGENENAME(orderResults(scde.table)))
  csv.file <- sprintf("SCDE_cluster_v2/%s.csv", contrastName)
  write.csv(scde.table, csv.file)
}
```

# Volcano plots {.tabset}

Let us first identify the extreme values of maximum likelihood estimate of
fold-change to scale subsequent plot for comparability:

```{r mleRange}
mleRange <-
  max(abs(do.call("c", lapply(scde.res, function(x){return(x$mle)}))))*c(-1,1)
```

## 4h (2 vs 1)

```{r volcano_4h_2v1, echo=FALSE}
v.data <- scde.res[["4h_cluster2-cluster1"]][,c("mle", "Z", "cZ")]
v.data <- convert.z.score(v.data)
ggplot(v.data) +
  geom_point(aes(mle, -log10(p.value), colour = (cZ != 0)), size = 0.5) +
  geom_hline(
    aes(yintercept = -log10(P), linetype = level), data = volcano.sig
  ) + scale_x_continuous(limits = mleRange) + theme_bw() +
  ggtitle("4h: cluster 2 - cluster 1")
```

```{r fig1_volcano_4h_2v1, echo=FALSE, eval=FALSE}
geneData <- subset(
  mutate(v.data,gene_name = fData(sce.ifm.4h)[rownames(v.data),"gene_name"]),
  gene_name %in% c(
    "NFKB1","IRF8","STAT5A","CEBPA","CD1A","CD1C","CD1E","CLECL1", # 8
    "IL1B","IL1A","EBI3","CD40","CD83", # 5
    "CTSS","CTSD","CTSL","GRN","APOE","APOC1","ACOT2", # 7
    "S100A9","STAB1","LRP1","PSAP") # 4
)
ggplot(v.data) +
  geom_point(aes(Z, -log10(p.value), colour = p.value < 0.01), size = 0.5) +
  geom_hline(yintercept = -log10(0.01), linetype = "dashed", alpha = 0.25) +
  geom_text_repel(
    aes(Z, -log10(p.value), label = gene_name), geneData,
    size = 2, min.segment.length = unit(0, "mm"), alpha = 0.75,
    nudge_x = ifelse(geneData$Z > 0, 1, -1), nudge_y = 0, segment.size = 0.2,
    fontface = "bold"
  ) +
  scale_x_continuous(limits = mleRange) + theme_bw() +
  ggtitle("4h: cluster 2 - cluster 1")
ggsave("13_out/4h_volcano_Vogel.pdf", height = 5, width = 6.5)
```

```{r fig2_volcano_4h_2v1, echo=FALSE, eval=FALSE}
ggplot(v.data) +
  geom_point(
    aes(mle, -log10(p.value), colour = p.value < 0.01),
    size = 0.5, alpha = 0.25
  ) +
  geom_hline(yintercept = -log10(0.01), linetype = "dashed") +
  geom_text_repel(
    aes(mle, -log10(p.value), label = gene_name),
    geneData, size = 2, min.segment.length = unit(0, "mm"), fontface = "bold"
  ) +
  scale_x_continuous(limits = mleRange) + theme_bw() +
  ggtitle("4h: cluster 2 - cluster 1")
ggsave("13_out/4h_volcano_Shalek.pdf", height = 5, width = 6.5)
```

## 6h (2 vs 1)

```{r volcano_6h_2v1, echo=FALSE}
v.data <- scde.res[["6h_cluster2-cluster1"]][,c("mle", "Z", "cZ")]
v.data <- convert.z.score(v.data)
ggplot(v.data) +
  geom_point(aes(mle, -log10(p.value), colour = (cZ != 0)), size = 0.5) +
  geom_hline(
    aes(yintercept = -log10(P), linetype = level), data = volcano.sig
  ) + scale_x_continuous(limits = mleRange) + theme_bw() +
  ggtitle("6h: cluster 2 - cluster 1")
```

```{r fig1_volcano_6h_2v1, echo=FALSE, eval=FALSE}
geneData <- subset(
  mutate(v.data,gene_name = fData(sce.ifm.4h)[rownames(v.data),"gene_name"]),
  gene_name %in% c(
    "CTSB","CTSC","CTSD","CTSL", # 4
    "DEFB1","DUOX","ATF3", # 3
    "IL12B","IL23A","IL1A","EBI3","IL6", # 5
    "MAFF","CREBFR","NFKB1","NFKB2") # 4
)
ggplot(v.data) +
  geom_point(aes(Z, -log10(p.value), colour = p.value < 0.01), size = 0.5) +
  geom_hline(yintercept = -log10(0.01), linetype = "dashed", alpha = 0.25) +
  geom_text_repel(
    aes(Z, -log10(p.value), label = gene_name), geneData,
    size = 2, min.segment.length = unit(0, "mm"), alpha = 0.75,
    nudge_x = ifelse(geneData$Z > 0, 1, -1), nudge_y = 0, segment.size = 0.2,
    fontface = "bold"
  ) +
  scale_x_continuous(limits = mleRange) + theme_bw() +
  ggtitle("6h: cluster 2 - cluster 1")
ggsave("13_out/6h_volcano_Vogel.pdf", height = 5, width = 6.5)
```

```{r fig2_volcano_6h_2v1, echo=FALSE, eval=FALSE}
ggplot(v.data) +
  geom_point(
    aes(mle, -log10(p.value), colour = p.value < 0.01),
    size = 0.5, alpha = 0.25
  ) +
  geom_hline(yintercept = -log10(0.01), linetype = "dashed") +
  geom_text_repel(
    aes(mle, -log10(p.value), label = gene_name),
    geneData, size = 2, min.segment.length = unit(0, "mm"), fontface = "bold"
  ) +
  scale_x_continuous(limits = mleRange) + theme_bw() +
  ggtitle("6h: cluster 2 - cluster 1")
ggsave("13_out/6h_volcano_Shalek.pdf", height = 5, width = 6.5)
```

# Count DE genes at various cut-offs {.tabset}

## 4h (2 vs 1)

```{r countDE_4h_2v1, echo=FALSE, results='asis'}
v.data <- scde.res[["4h_cluster2-cluster1"]][,c("Z", "cZ")]
v.data <- convert.z.score(v.data)
v.data <- data.frame(
  P.01 = sum(v.data$p.value < 0.01),
  P.05 = sum(v.data$p.value < 0.05),
  cZ = sum(v.data$cZ != 0),
  row.names = "4h: cluster2 - cluster1"
)
pander::pandoc.table(v.data)
```

## 6h (2 vs 1)

```{r countDE_6h_2v1, echo=FALSE, results='asis'}
v.data <- scde.res[["6h_cluster2-cluster1"]][,c("Z", "cZ")]
v.data <- convert.z.score(v.data)
v.data <- data.frame(
  P.01 = sum(v.data$p.value < 0.01),
  P.05 = sum(v.data$p.value < 0.05),
  cZ = sum(v.data$cZ != 0),
  row.names = "6h: cluster2 - cluster1"
)
pander::pandoc.table(v.data)
```

# Gene ontology

Let us use the [goseq](http://bioconductor.org/packages/goseq) package to
identify the most enriched gene ontologies among the various lists of DE genes.
Note that we restrict the results to GO categories associated with at least
**10** genes, for robustness.

Let us first define:

* a list to store the GO enrichment tables returned by *goseq*

```{r goseq.res_init, eval=FALSE}
goseq.res <- list()
```

* the gene length information reported by *featureCounts*:

```{r geneLengths}
geneLengths <- fData(sce.endo)[,"width", drop = FALSE]
```

Let us identify over- and under-represented GO categories in DE genes
identified in each contrast,
using as background the list of genes detected at each time point:

## Estimate enrichment {.tabset}

### 4h (2 vs 1)

```{r GO_4h_DE, eval=FALSE}
contrastName <- "4h_cluster2-cluster1"
bg.4h <- featureNames(sce.ifm.4h)
scde.table <- scde.res[[contrastName]]
scde.table <- convert.z.score(scde.table)
scde.table <- subset(scde.table, p.value < 0.01)
de.genes <- (bg.4h %in% rownames(scde.table))
names(de.genes) <- bg.4h; table(de.genes)
pwf <- nullp(de.genes, bias.data = geneLengths[names(de.genes),])
go.res <- goseq(pwf, "hg38", "ensGene")
goseq.res[[contrastName]] <- go.res
```

### 6h (2 vs 1)

```{r GO_6h_DE, eval=FALSE}
contrastName <- "6h_cluster2-cluster1"
bg.6h <- featureNames(sce.ifm.6h)
scde.table <- scde.res[[contrastName]]
scde.table <- convert.z.score(scde.table)
scde.table <- subset(scde.table, p.value < 0.01)
de.genes <- (bg.6h %in% rownames(scde.table))
names(de.genes) <- bg.6h; table(de.genes)
pwf <- nullp(de.genes, bias.data = geneLengths[names(de.genes),])
go.res <- goseq(pwf, "hg38", "ensGene")
goseq.res[[contrastName]] <- go.res
```

```{r goseq.res, echo=FALSE, eval=FALSE}
saveRDS(goseq.res, "rds/13_cluster_goseq.res_v2.rds")
for (contrastName in names(goseq.res)){
  goseq.table <- goseq.res[[contrastName]]
  write.csv(goseq.table, sprintf("SCDE_cluster_v2/GO_%s.csv", contrastName))
}
```

## Tables {.tabset}

### Setup

Let us define a function to display the **50** most significantly
`"over"`- or `"under"`-represented GO categories
annotated to at least **10** genes (in the entire genome):

```{r showTopGO}
showTopGO <- function(x, direction = "over", cutoff = 0.05, min = 10){
  goCols <- c(
  "category","over_represented_pvalue","under_represented_pvalue",
  "numDEInCat","numInCat","ontology","term")
  directionCol <- sprintf("%s_represented_pvalue", direction)
  tmp.go <- x[,goCols]
  tmp.sig <- (tmp.go[,directionCol] < cutoff); tmp.go <- tmp.go[tmp.sig,]
  tmp.min <- (tmp.go[,"numInCat"] >= min); tmp.go <- tmp.go[tmp.min,]
  tmp.go <- tmp.go[order(tmp.go[,directionCol]),]
  tmp.go <- head(tmp.go, 50)
  DT::datatable(
    tmp.go, list(pageLength = 10, searching = TRUE), filter = "top",
    rownames = FALSE
  )
}
```

```{r, echo=FALSE}
write.csv(goseq.res[[1]], "13_out/4h_cluster2-1.csv")
write.csv(goseq.res[[2]], "13_out/6h_cluster2-1.csv")
```


### Over-represented {.tabset}

#### 4h_cluster2-cluster1

```{r goseq.over_1, echo=FALSE}
showTopGO(goseq.res[["4h_cluster2-cluster1"]])
```

#### 6h_cluster2-cluster1

```{r goseq.over_2, echo=FALSE}
showTopGO(goseq.res[["6h_cluster2-cluster1"]])
```

### Under-represented {.tabset}

#### `r names(goseq.res)[1]`

```{r goseq.under_1, echo=FALSE}
showTopGO(goseq.res[[1]], "under")
```

#### `r names(goseq.res)[2]`

```{r goseq.under_2, echo=FALSE}
showTopGO(goseq.res[[2]], "under")
```

# Comparison of fold-change between clusters and experimental groups {.tabset}

Let us first define:

* functions that displays as violins the expression level of a given
  gene in each group at 4h and 6h, respectively (only 4h shown)

```{r normExprsByName.4h}
normExprsByName.4h <- function(geneId){
  sc <- sce.ifm.4h[,sce.ifm.4h$Infection != "Mock"]
  gdata <- data.frame(
    norm_exprs = norm_exprs(sc)[geneId,],
    pData(sc)[,c("Infection","Status","quickCluster")],
    row.names = sampleNames(sc)
  )
  gdata$Infection <- paste("STM", gdata$Infection)
  geneName <- subset(fData(sc),gene_id==geneId,"gene_name",drop=TRUE)
  ggplot(gdata, aes(Status, norm_exprs, fill = Status)) + 
    geom_violin(alpha = 0.25) +
    geom_jitter(aes(colour = quickCluster), width = 0.25, size = 3) +
    facet_grid(~Infection) +
    ggtitle(sprintf("%s - %s", geneName, geneId)) +
    scale_y_continuous(limits=range(norm_exprs(sc))) +
    scale_colour_manual(values = col.cluster) +
    theme_minimal()
}
```

```{r normExprsByName.6h, echo=FALSE}
normExprsByName.6h <- function(geneId){
  sc <- sce.ifm.6h[,sce.ifm.6h$Infection != "Mock"]
  gdata <- data.frame(
    norm_exprs = norm_exprs(sc)[geneId,],
    pData(sc)[,c("Infection","Status","quickCluster")],
    row.names = sampleNames(sc)
  )
  gdata$Infection <- paste("STM", gdata$Infection)
  geneName <- subset(fData(sc),gene_id==geneId,"gene_name",drop=TRUE)
  ggplot(gdata, aes(Status, norm_exprs, fill = Status)) + 
    geom_violin(alpha = 0.25) +
    geom_jitter(aes(colour = quickCluster), width = 0.25, size = 3) +
    facet_grid(~Infection) + ggtitle(sprintf("%s - %s", geneName, geneId)) +
    scale_y_continuous(limits=range(norm_exprs(sc))) +
    scale_colour_manual(values = col.cluster) + theme_minimal()
}
```

```{r normExprsByName.4h.Status}
normExprsByName.4h.Status <- function(geneId){ # facet by status
  sc <- sce.ifm.4h[,sce.ifm.4h$Infection != "Mock"]
  gdata <- data.frame(
    norm_exprs = norm_exprs(sc)[geneId,],
    pData(sc)[,c("Infection","Status","quickCluster")],
    row.names = sampleNames(sc)
  )
  gdata$Infection <- paste("STM", gdata$Infection)
  geneName <- subset(fData(sc),gene_id==geneId,"gene_name",drop=TRUE)
  ggplot(gdata, aes(Infection, norm_exprs, fill = Infection)) + 
    geom_violin(alpha = 0.25) +
    geom_jitter(aes(colour = quickCluster), width = 0.25, size = 3) +
    facet_grid(~Status) + ggtitle(sprintf("%s - %s", geneName, geneId)) +
    scale_y_continuous(limits=range(norm_exprs(sc))) +
    scale_colour_manual(values = col.cluster) + theme_minimal()
}
```

```{r normExprsByName.6h.Status, echo=FALSE}
normExprsByName.6h.Status <- function(geneId){ # facet by status
  sc <- sce.ifm.6h[,sce.ifm.6h$Infection != "Mock"]
  gdata <- data.frame(
    norm_exprs = norm_exprs(sc)[geneId,],
    pData(sc)[,c("Infection","Status","quickCluster")],
    row.names = sampleNames(sc)
  )
  gdata$Infection <- paste("STM", gdata$Infection)
  geneName <- subset(fData(sc),gene_id==geneId,"gene_name",drop=TRUE)
  ggplot(gdata, aes(Infection, norm_exprs, fill = Infection)) + 
    geom_violin(alpha = 0.25) +
    geom_jitter(aes(colour = quickCluster), width = 0.25, size = 3) +
    facet_grid(~Status) + ggtitle(sprintf("%s - %s", geneName, geneId)) +
    scale_y_continuous(limits=range(norm_exprs(sc))) +
    scale_colour_manual(values = col.cluster) + theme_minimal()
}
```

```{r mleRange2, include=FALSE}
scde.4h.D23580 <- scde.groupRes[["4h_D23580_infected-4h_D23580_exposed"]]
scde.4h.LT2 <- scde.groupRes[["4h_LT2_infected-4h_LT2_exposed"]]
scde.6h.D23580 <- scde.groupRes[["6h_D23580_infected-6h_D23580_exposed"]]
scde.6h.LT2 <- scde.groupRes[["6h_LT2_infected-6h_LT2_exposed"]]
mleRange2 <-
  max(abs(do.call("c", lapply(
    append(
      scde.res,list(scde.4h.D23580,scde.4h.LT2,scde.6h.D23580,scde.6h.LT2)
    ),
    function(x){return(x$mle)}
  )))) * c(-1,1)
```


## 4h {.tabset}

### Plot

```{r 4h_MLE_cluster_group, echo=FALSE, warning=FALSE, fig.height=5}
scde.4h.cluster <- scde.res[["4h_cluster2-cluster1"]]
scde.4h.D23580 <- convert.z.score(scde.4h.D23580)
scde.4h.LT2 <- convert.z.score(scde.4h.LT2)
scde.4h.cluster <- convert.z.score(scde.4h.cluster)
fn <- rownames(scde.4h.cluster)
scde.table <- data.frame(
  mle.cluster = rep(scde.4h.cluster[fn,"mle"], 2),
  mle.group = c(scde.4h.D23580[fn,"mle"], scde.4h.LT2[fn,"mle"]),
  Group = c(rep("STM-D23580",length(fn)), rep("STM-LT2",length(fn))),
  GENEID = rep(fn, 2),
  stringsAsFactors = FALSE
)
scde.table$GENENAME <-
  with(fData(sce.endo), gene_name[match(scde.table$GENEID, gene_id)])
scde.genes <- subset(scde.table, GENEID %in% c(
  # Cluster 1
  "ENSG00000108688", # CCL7
  "ENSG00000163220", # S100A9
  "ENSG00000135047", # CTSL
  "ENSG00000117984", # CTSD
  # Cluster 2
  "ENSG00000173757", # STAT5B
  "ENSG00000108702", # CCL1
  "ENSG00000101017", # CD40
  "ENSG00000100906" # NFKBIA
))
ggplot(scde.table,aes(-mle.cluster,mle.group))+
  geom_point(size = 0.5, colour = "turquoise3", alpha = 0.2) +
  theme_minimal() + facet_grid(~ Group) +
  scale_x_continuous(limits=mleRange2) + scale_y_continuous(limits=mleRange2) +
  geom_text_repel(
    aes(label = GENENAME),data=scde.genes,min.segment.length=unit(0,"mm")) +
  labs(y="MLE infected / exposed",x="MLE cluster 1 / cluster 2",title="4h")
ggsave("13_out/4h_correlation_cluster.pdf", width = 6.5, height = 5)
```

```{r 4h_MLE_cluster_group_csv, echo=FALSE, eval=FALSE}
scde.table$mle.p <- scde.table$mle.cluster * scde.table$mle.group
write.csv(scde.table, "SCDE_cluster_v2/4h_MLE_cluster_group.csv")
```

### CCL7

<!-- From the highest cluster 1 genes -->

```{r CCL7_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000108688")
```

### S100A9

```{r S100A9_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000163220")
```

### CTSL

```{r CTSL_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000135047")
```

### CTSD

```{r CTSD_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000117984")
```

### CTSS

```{r CTSS_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000163131")
```

### STAB1

```{r STAB1_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000010327")
```

### STAT5B

<!-- From the highest cluster 2 genes -->

```{r STAT5B_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000173757")
```

### CCL1

```{r CCL1_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000108702")
```

### CD40

```{r CD40_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000101017")
```

### IL1A

```{r IL1A_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000115008")
```

### IL1B

```{r IL1B_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000125538")
```

### NFKBIA

```{r NFKBIA_normExprsByName.4h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.4h("ENSG00000100906")
```

```{r fig.violin.4h.D23580.EBI3, echo=FALSE, eval=FALSE}
geneName <- "EBI3"
geneId<-subset(fData(sce.ifm.4h),gene_name==geneName,"gene_id",drop=TRUE)
sampleIdx <- with(pData(sce.ifm.4h), Infection == "STM-D23580" & Time == "4h")
gdata <- data.frame(
  norm_exprs = norm_exprs(sce.ifm.4h)[geneId, sampleIdx],
  pData(sce.ifm.4h)[sampleIdx,c("Infection","Status","quickCluster")],
  row.names = sampleNames(sce.ifm.4h)[sampleIdx]
)
ggplot(gdata, aes(Status, norm_exprs, fill = Status)) + 
  geom_violin(alpha = 0.25) +
  geom_jitter(aes(colour = quickCluster), width = 0.25, size = 3) +
  facet_grid(~Infection) + ggtitle(sprintf("%s - %s", geneName, geneId)) +
  scale_y_continuous(limits=range(norm_exprs(sce.ifm.4h))) +
  scale_colour_manual(values = col.cluster) + theme_minimal()
ggsave(sprintf("13_out/violin_4h_D23580_%s.pdf", geneName),height=4,width=5)
```

```{r fig.violin.4h.D23580.NIT1, echo=FALSE, eval=FALSE}
geneName <- "NIT1"
geneId<-subset(fData(sce.ifm.4h),gene_name==geneName,"gene_id",drop=TRUE)
sampleIdx <- with(pData(sce.ifm.4h), Infection == "STM-D23580" & Time == "4h")
gdata <- data.frame(
  norm_exprs = norm_exprs(sce.ifm.4h)[geneId, sampleIdx],
  pData(sce.ifm.4h)[sampleIdx,c("Infection","Status","quickCluster")],
  row.names = sampleNames(sce.ifm.4h)[sampleIdx]
)
ggplot(gdata, aes(Status, norm_exprs, fill = Status)) + 
  geom_violin(alpha = 0.25) +
  geom_jitter(aes(colour = quickCluster), width = 0.25, size = 3) +
  facet_grid(~Infection) + ggtitle(sprintf("%s - %s", geneName, geneId)) +
  scale_y_continuous(limits=range(norm_exprs(sce.ifm.4h))) +
  scale_colour_manual(values = col.cluster) + theme_minimal()
ggsave(sprintf("13_out/violin_4h_D23580_%s.pdf", geneName),height=4,width=5)
```

```{r fig.violin.4h.LT2.VPS25, echo=FALSE, eval=FALSE}
geneName <- "VPS25"
geneId<-subset(fData(sce.ifm.4h),gene_name==geneName,"gene_id",drop=TRUE)
sampleIdx <- with(pData(sce.ifm.4h), Infection == "STM-LT2" & Time == "4h")
gdata <- data.frame(
  norm_exprs = norm_exprs(sce.ifm.4h)[geneId, sampleIdx],
  pData(sce.ifm.4h)[sampleIdx,c("Infection","Status","quickCluster")],
  row.names = sampleNames(sce.ifm.4h)[sampleIdx]
)
ggplot(gdata, aes(Status, norm_exprs, fill = Status)) + 
  geom_violin(alpha = 0.25) +
  geom_jitter(aes(colour = quickCluster), width = 0.25, size = 3) +
  facet_grid(~Infection) + ggtitle(sprintf("%s - %s", geneName, geneId)) +
  scale_y_continuous(limits=range(norm_exprs(sce.ifm.4h))) +
  scale_colour_manual(values = col.cluster) + theme_minimal()
ggsave(sprintf("13_out/violin_4h_LT2_%s.pdf", geneName),height=4,width=5)
```

## 6h {.tabset}

### Plot

```{r 6h_MLE_cluster_group, echo=FALSE, warning=FALSE, fig.height=5}
scde.6h.cluster <- scde.res[["6h_cluster2-cluster1"]]
scde.6h.D23580 <- convert.z.score(scde.6h.D23580)
scde.6h.LT2 <- convert.z.score(scde.6h.LT2)
scde.6h.cluster <- convert.z.score(scde.6h.cluster)
fn <- rownames(scde.6h.cluster)
scde.table <- data.frame(
  mle.cluster = rep(scde.6h.cluster[fn,"mle"], 2),
  mle.group = c(scde.6h.D23580[fn,"mle"], scde.6h.LT2[fn,"mle"]),
  Group = c(rep("STM-D23580",length(fn)), rep("STM-LT2",length(fn))),
  GENEID = rep(fn, 2),
  stringsAsFactors = FALSE
)
scde.table$GENENAME <-
  with(fData(sce.endo), gene_name[match(scde.table$GENEID, gene_id)])
scde.genes <- subset(scde.table, GENENAME %in% c(
  "IL23A","STAT5B","EBI3","LAMP3",
  "IL12B","IL36B","IL1A","IL6"
  ))
ggplot(scde.table,aes(-mle.cluster,mle.group))+
  geom_point(size = 0.5, colour = "turquoise3", alpha = 0.2) +
  theme_minimal() + facet_grid(~ Group) +
  scale_x_continuous(limits=mleRange2) + scale_y_continuous(limits=mleRange2) +
  geom_text_repel(
    aes(label = GENENAME),data=scde.genes,min.segment.length=unit(0,"mm")) +
  labs(y="MLE infected / exposed",x="MLE cluster 1 / cluster 2",title="6h")
```

```{r 6h_MLE_cluster_group_csv, echo=FALSE, eval=FALSE}
scde.table$mle.p <- scde.table$mle.cluster * scde.table$mle.group
write.csv(scde.table, "SCDE_cluster_v2/6h_MLE_cluster_group.csv")
```

### IL23A

```{r IL23A_normExprsByName.6h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.6h("ENSG00000110944")
```

### IL12B

```{r IL12B_normExprsByName.6h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.6h("ENSG00000110944")
```

### EBI3

```{r EBI3_normExprsByName.6h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.6h("ENSG00000105246")
```

### IL1A

```{r IL1A_normExprsByName.6h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.6h("ENSG00000115008")
```

### IL6

```{r IL6_normExprsByName.6h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.6h("ENSG00000136244")
```

### LAMP3

```{r LAMP3_normExprsByName.6h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.6h("ENSG00000078081")
```

### STAT5B

```{r STAT5B_normExprsByName.6h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.6h("ENSG00000173757")
```

### IL36B

```{r IL36B_normExprsByName.6h, echo=FALSE, warning=FALSE, fig.height=5}
normExprsByName.6h("ENSG00000136696")
```

### Heat map

```{r heatmap, echo=FALSE}
exprsRange <- range(norm_exprs(sce.ifm.6h))
colorMap <- circlize::colorRamp2(
  c(min(exprsRange), median(exprsRange), max(exprsRange)),
  c("blue","white","red"))
geneNames <- c(
  "IL12B","IL23A","IL1A","IL1B","IL6","EBI3",
  "DEFB1","CTSL")
geneIds <- with(fData(sce.endo), gene_id[match(geneNames, gene_name)])
stopifnot(length(geneIds) == length(geneNames))
norm.mat <- norm_exprs(sce.ifm.6h)[geneIds,]
norm.d23exp <- norm.mat[,sce.ifm.6h$Treatment == "STM-D23580_Exposed"]
norm.d23inf <- norm.mat[,sce.ifm.6h$Treatment == "STM-D23580_Violet +"]
norm.lt2exp <- norm.mat[,sce.ifm.6h$Treatment == "STM-LT2_Exposed"]
norm.lt2inf <- norm.mat[,sce.ifm.6h$Treatment == "STM-LT2_Violet +"]
ha.d23.exp <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.d23exp), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster)
)
ha.d23.inf <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.d23inf), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
ha.lt2.exp <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.lt2exp), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
ha.lt2.inf <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.lt2inf), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
rownames(norm.d23exp) <- geneNames
hm.split <- c(rep(" ",6), rep("  ",2))
ht_list <-
  Heatmap(
    norm.d23exp, colorMap, "Normalised\nexpression", row_names_side = "left",
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), top_annotation = ha.d23.exp,
    column_title = "STM-D23580\nexposed") +
  Heatmap(
    norm.d23inf, colorMap, "hm2", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-D23580\ninfected",
    top_annotation = ha.d23.inf, show_heatmap_legend = FALSE) + 
  Heatmap(
    norm.lt2exp, colorMap, "hm3", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-LT2\nexposed",
    top_annotation = ha.lt2.inf, show_heatmap_legend = FALSE) + 
  Heatmap(
    norm.lt2inf, colorMap, "hm4", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-LT2\ninfected",
    top_annotation = ha.lt2.inf, show_heatmap_legend = FALSE)
draw(ht_list, split = hm.split)
```

```{r heatmap_2, echo=FALSE, include=FALSE}
exprsRange <- range(norm_exprs(sce.ifm.6h))
colorMap <- circlize::colorRamp2(
  c(min(exprsRange), median(exprsRange), max(exprsRange)),
  c("blue","white","red"))
geneNames <- c(
  "IL10","STAT3","MARCH1","SOCS1","SOCS2","SOCS3","CD86","CD83",
  "DNAJC13","BLOC1S3")
geneIds <- with(fData(sce.endo), gene_id[match(geneNames, gene_name)])
stopifnot(nrow(geneIds) == length(geneNames))
norm.mat <-norm_exprs(sce.ifm.6h)[geneIds$gene_id,]
norm.d23exp <- norm.mat[,sce.ifm.6h$Treatment == "STM-D23580_Exposed"]
norm.d23inf <- norm.mat[,sce.ifm.6h$Treatment == "STM-D23580_Violet +"]
norm.lt2exp <- norm.mat[,sce.ifm.6h$Treatment == "STM-LT2_Exposed"]
norm.lt2inf <- norm.mat[,sce.ifm.6h$Treatment == "STM-LT2_Violet +"]
ha.d23.exp <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.d23exp), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster)
)
ha.d23.inf <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.d23inf), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
ha.lt2.exp <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.lt2exp), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
ha.lt2.inf <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.lt2inf), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
rownames(norm.d23exp) <- geneNames
ht_list <-
  Heatmap(
    norm.d23exp, colorMap, "Normalised\nexpression", row_names_side = "left",
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), top_annotation = ha.d23.exp,
    column_title = "STM-D23580\nexposed") +
  Heatmap(
    norm.d23inf, colorMap, "hm2", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-D23580\ninfected",
    top_annotation = ha.d23.inf, show_heatmap_legend = FALSE) + 
  Heatmap(
    norm.lt2exp, colorMap, "hm3", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-LT2\nexposed",
    top_annotation = ha.lt2.inf, show_heatmap_legend = FALSE) + 
  Heatmap(
    norm.lt2inf, colorMap, "hm4", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-LT2\ninfected",
    top_annotation = ha.lt2.inf, show_heatmap_legend = FALSE)
pdf("13_out/heatmap_2.pdf", height = 5, width = 7)
draw(ht_list)
dev.off()
```

```{r heatmap_3, echo=FALSE, include=FALSE}
exprsRange <- range(norm_exprs(sce.ifm.6h))
colorMap <- circlize::colorRamp2(
  c(min(exprsRange), median(exprsRange), max(exprsRange)),
  c("blue","white","red"))
geneNames <- c(
  "IL6","IL6R","IL6ST","IL10","IL10RA","IL10RB","TYK2",
  "JAK1","JAK3","STAT3","MARCH1")
geneIds <- with(fData(sce.endo), gene_id[match(geneNames, gene_name)])
stopifnot(nrow(geneIds) == length(geneNames))
norm.mat <-norm_exprs(sce.ifm.6h)[geneIds$gene_id,]
norm.d23exp <- norm.mat[,sce.ifm.6h$Treatment == "STM-D23580_Exposed"]
norm.d23inf <- norm.mat[,sce.ifm.6h$Treatment == "STM-D23580_Violet +"]
norm.lt2exp <- norm.mat[,sce.ifm.6h$Treatment == "STM-LT2_Exposed"]
norm.lt2inf <- norm.mat[,sce.ifm.6h$Treatment == "STM-LT2_Violet +"]
ha.d23.exp <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.d23exp), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster)
)
ha.d23.inf <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.d23inf), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
ha.lt2.exp <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.lt2exp), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
ha.lt2.inf <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.lt2inf), "quickCluster", drop = FALSE],
  col = list(quickCluster = col.cluster),
  show_legend = FALSE
)
rownames(norm.d23exp) <- geneNames
ht_list <-
  Heatmap(
    norm.d23exp, colorMap, "Normalised\nexpression", row_names_side = "left",
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), top_annotation = ha.d23.exp,
    column_title = "STM-D23580\nexposed") +
  Heatmap(
    norm.d23inf, colorMap, "hm2", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-D23580\ninfected",
    top_annotation = ha.d23.inf, show_heatmap_legend = FALSE) + 
  Heatmap(
    norm.lt2exp, colorMap, "hm3", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-LT2\nexposed",
    top_annotation = ha.lt2.inf, show_heatmap_legend = FALSE) + 
  Heatmap(
    norm.lt2inf, colorMap, "hm4", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-LT2\ninfected",
    top_annotation = ha.lt2.inf, show_heatmap_legend = FALSE)
pdf("13_out/heatmap_3.pdf", height = 5, width = 8)
draw(ht_list)
dev.off()
```

```{r heatmap_4, echo=FALSE, include=FALSE}
exprsRange <- range(norm_exprs(sce.ifm.6h))
colorMap <- circlize::colorRamp2(
  c(min(exprsRange), median(exprsRange), max(exprsRange)),
  c("blue","white","red"))
geneNames <- c(
  "IL6","IL6R","IL6ST","IL10","IL10RA","IL10RB","TYK2",
  "JAK1","JAK3","STAT3","MARCH1")
geneIds <- with(fData(sce.endo), gene_id[match(geneNames, gene_name)])
stopifnot(nrow(geneIds) == length(geneNames))
norm.mat <- norm_exprs(sce.ifm.6h)[geneIds,]
norm.d23 <- norm.mat[,sce.ifm.6h$Infection == "STM-D23580"]
norm.lt2 <- norm.mat[,sce.ifm.6h$Infection == "STM-LT2"]
ha.d23 <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.d23), "Status", drop = FALSE],
  col = list(Status = col.status)
)
ha.lt2 <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[colnames(norm.lt2), "Status", drop = FALSE],
  col = list(Status = col.status),
  show_legend = FALSE
)
rownames(norm.d23) <- geneNames
ht_list <-
  Heatmap(
    norm.d23, colorMap, "Normalised\nexpression", row_names_side = "left",
    cluster_rows=FALSE,show_column_names=FALSE,#show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), top_annotation = ha.d23,
    column_title = "STM-D23580"
  ) +
  Heatmap(
    norm.lt2, colorMap, "hm4", show_row_names = FALSE,
    cluster_rows=FALSE,show_column_names=FALSE,#show_column_dend=FALSE,
    column_title_gp = gpar(fontsize = 10), column_title = "STM-LT2",
    top_annotation = ha.lt2, show_heatmap_legend = FALSE)
pdf("13_out/heatmap_4.pdf", height = 4, width = 8)
draw(ht_list)
dev.off()
```

```{r MARCH1_normExprsByName.6h.Status, include=FALSE}
normExprsByName.6h.Status("ENSG00000145416")
ggsave(sprintf("20170525_lab-meeting/%s_colour.pdf", "MARCH1"), width = 6, height = 4)
```

```{r CD86_normExprsByName.6h.Status, include=FALSE}
normExprsByName.6h.Status("ENSG00000114013")
ggsave(sprintf("20170525_lab-meeting/%s_colour.pdf", "CD86"), width = 6, height = 4)
```

```{r CD83_normExprsByName.6h.Status, include=FALSE}
normExprsByName.6h.Status("ENSG00000112149")
ggsave(sprintf("20170525_lab-meeting/%s_colour.pdf", "CD83"), width = 6, height = 4)
```

```{r IL10_normExprsByName.6h.Status, include=FALSE}
normExprsByName.6h.Status("ENSG00000136634")
ggsave(sprintf("20170525_lab-meeting/%s_colour.pdf", "IL10"), width = 6, height = 4)
```


# Heat map {#HeatMap .tabset}

In this section, the differentially expressed genes identified above
between clusters are displayed in heat maps.

## 4h

```{r HeatMap_DE_4h, echo=FALSE, fig.height=7}
# Expression matrix
scde.table <- convert.z.score(scde.res[["4h_cluster2-cluster1"]])
scde.genes <- rownames(scde.table)[scde.table$p.value < 0.01]
scde.cells <- sampleNames(sce.ifm.4h)[
  order(factor(sce.ifm.4h$quickCluster,1:2),na.last=NA)]
scde.norm <- norm_exprs(sce.ifm.4h)[scde.genes, scde.cells]
# Sample annotations
haNames <- c("quickCluster", "Status", "Infection")
ha <- HeatmapAnnotation(
  df = pData(sce.ifm.4h)[scde.cells, haNames],
  col = list(
    quickCluster = col.cluster,
    Status = col.status,
    Infection = col.infection
  )
)
# Gene annotations
anno.names <- c(
  "CTSS","CTSD","CTSL","CTSC","SPSB1","CTSB","GRN", # cathepsin
  "S100A9","STAB1", # antimicrobial
  "SPIRE1","VAMP4","LRP1","PSAP", # vesicles
  "NFKBIA","IRF8","STAT5A","CEBPA", # transcription factors
  "IL1B","IL1A","EBI3" # pro-inflammatory
)
anno.ids <- with(fData(sce.endo), gene_id[match(anno.names, gene_name)])
stopifnot(length(anno.ids) == length(anno.names))
gene.idx = match(anno.ids, rownames(scde.norm))
ra <- rowAnnotation(
  link = row_anno_link(at = gene.idx, labels = anno.names),
  width = unit(0.25, "char") + max_text_width(labels))
# Heat map
hm <- Heatmap(
  scde.norm,
  name = "norm_exprs", column_title = "4h", cluster_columns = FALSE,
  top_annotation = ha, show_row_names = FALSE, show_column_names = FALSE
)
draw(hm + ra)
```

## 6h

```{r HeatMap_DE_6h, echo=FALSE, fig.height=7}
# Expression matrix
scde.table <- convert.z.score(scde.res[["6h_cluster2-cluster1"]])
scde.genes <- rownames(scde.table)[scde.table$p.value < 0.01]
scde.cells <- sampleNames(sce.ifm.6h)[
  order(factor(sce.ifm.6h$quickCluster,1:2),na.last=NA)]
scde.norm <- norm_exprs(sce.ifm.6h)[scde.genes, scde.cells]
# Sample annotations
haNames <- c("quickCluster", "Status", "Infection")
ha <- HeatmapAnnotation(
  df = pData(sce.ifm.6h)[scde.cells, haNames],
  col = list(
    quickCluster = col.cluster,
    Status = col.status,
    Infection = col.infection
  )
)
# Gene annotations
anno.names <- c(
  "IL12B","IL23A","IL1A","EBI3","IL6","LITAF","NLRP1", # pro-inflammatory
  "CREBRF","IRF8","STAT5","NFKB1","NFKB2","NFKBIA","MAFF", # TFs
  "CTSB","CTSC","CTSD","CTSL","CAPN1", # proteases
  "DEFB1","DUOX", # antimicrobial
  "CORO1A" # vesicles
)
anno.ids <- with(fData(sce.endo), gene_id[match(anno.names, gene_name)])
stopifnot(length(anno.ids) == length(anno.names))
gene.idx = match(anno.ids, rownames(scde.norm))
ra <- rowAnnotation(
  link = row_anno_link(at = gene.idx, labels = anno.names),
  width = unit(0.25, "char") + max_text_width(labels))
# Heat map
hm <- Heatmap(
  scde.norm,
  name = "norm_exprs", column_title = "6h", cluster_columns = FALSE,
  top_annotation = ha, show_row_names = FALSE, show_column_names = FALSE
)
draw(hm + ra)
```

# Expression profile of selected genes {.tabset}

## IL6

Up-regulated in cluster 2 relative to cluster 1 at the 6h time point:

```{r single.scde.6h_D23580_IL6, echo=FALSE}
scde.id <- mapIds(EnsDb.Hsapiens.v79,"IL6","GENEID","GENENAME")
single.scde.6h(scde.id, 2, 1)
```


