---
title: "Differential expression using scde"
bibliography:
  bibtex.bib
---

<!-- As HTML pages are built independently, large objects need to be
reimported from disk, and smaller ones quickly recomputed. -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(scater)
library(scran)
library(scde)
library(ensembldb)
library(dplyr)
library(ggplot2)
sce.norm <- readRDS("rds/sce.norm.tSNE.rds")
EnsDb.Hsapiens.v79 <- EnsDb.Hsapiens.v79::EnsDb.Hsapiens.v79
```

<!-- Cleaner code for v2
Remove ERCC spike-ins
!! Split cells by time
!! Retain genes detected at each time
Error model: {cells @ time; genes @ time}
Prior @ time
DE: {group on all cells with NA; no bat}
-->

# Remove spike-in features

The analysis starts with a matrix of read counts,
filtered based on gene and cell requirements.
In this case, let us use the raw numbers of reads mapped to each endogenous
feature, recast as `integer` type:

```{r table_isSpike}
table(isSpike(sce.norm))
```

```{r sce.endo}
sce.endo <- sce.norm[!isSpike(sce.norm),]
```

```{r sce.endo_show, echo=FALSE}
dim(sce.endo)
```


# Experimental groups

Let us also define a factor that separate cells from each experimental group:

```{r sg}
sg <- sce.norm$Group
names(sg) <- colnames(sce.norm)
```

```{r sg_pandoc, echo=FALSE, results='asis'}
pander::pandoc.table(as.data.frame(summary(sg)))
```

# Data set-wide error models

## Compute

Error models were previously computed using raw counts for
all endogenous features detected as per feature QC,
for all the cells that passed QC:

```{r oifm, eval=FALSE}
cd <- counts(sce.endo); storage.mode(cd) <- "integer"
o.ifm <- scde.error.models(counts = cd, groups = sg, n.cores = 4)
```

```{r o.ifm_save, echo=FALSE, eval=FALSE}
saveRDS(o.ifm, "rds/o.ifm.rds")
```

Let us load those precomputed error models:

```{r o.ifm_read}
o.ifm <- readRDS("rds/o.ifm.rds")
```

## Remove poor fits (if any)

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed:

```{r validCells}
valid.cells <- (o.ifm$corr.a > 0)
o.ifm <- o.ifm[valid.cells, ]
```

```{r validCells_show, echo=FALSE}
table(valid.cells)
```


## Reorder counts

The `scde.error.models` produces error models with cells reordered 
by experimental group.
For clarity, let us prepare a `SCESet` and count matrix that match this order:

```{r sce.ifm}
sce.ifm <- sce.endo[,rownames(o.ifm)]
cd.ifm <- counts(sce.ifm); storage.mode(cd.ifm) <- "integer"
```

# Data set-wide prior

Finally, we need to define an expression magnitude prior for the genes.
Its main function, however, is to define a grid of expression magnitude values
on which the numerical calculations will be carried out.

```{r oPrior}
o.prior <- scde.expression.prior(o.ifm, cd.ifm, show.plot = TRUE)
```

# Subset data by time point (6h)

## Subset cells

```{r sce.6h}
sce.6h <- sce.endo[,sce.norm$Time == '6h']
```

```{r sce.6h_show, echo=FALSE}
dim(sce.6h)
```


## Subset detected genes

Let us first define a function to apply the detection cutoff:

```{r filterCounts}
filterCounts <- function(m, g, counts = 10, cells = 5, groups = 1){
  apply(m, 1, function(e){
    sum(tapply(e >= counts, g, "sum") >= cells) >= 1
  })
}
```

Apply the function:

```{r cleanCounts_6h}
sg.6h <- droplevels(sce.6h$Group)
keep.6h <- filterCounts(counts(sce.6h), sg.6h)
cd.6h <- counts(sce.6h)[keep.6h,]; storage.mode(cd.6h) <- 'integer'
```

```{r cd.6h_show, echo=FALSE}
dim(cd.6h)
```


# Error models by time point (6h)

## Compute

Error models were previously computed using raw counts for
all endogenous features detected as per feature QC,
for all the cells that passed QC:

```{r o.ifm.6h, eval=FALSE}
o.ifm.6h <- scde.error.models(counts = cd.6h, groups = sg.6h, n.cores = 4)
```

```{r o.ifm.6h_save, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.6h, "rds/o.ifm.6h.rds")
```

Let us load those precomputed error models:

```{r o.ifm.6h_read}
o.ifm.6h <- readRDS("rds/o.ifm.6h.rds")
```

## Remove poor fits (if any)

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed:

```{r valid.cells.6h}
valid.cells <- o.ifm.6h$corr.a > 0
o.ifm.6h <- o.ifm.6h[valid.cells, ]
```

```{r valid.cells.6h_show}
table(valid.cells)
```


## Reorder counts

The `scde.error.models` produces error models with cells reordered 
by experimental group.
For clarity, let us prepare a `SCESet` and count matrix that match this order:

```{r sce.ifm.6h}
sce.ifm.6h <- sce.6h[,rownames(o.ifm.6h)]
cd.ifm.6h <- counts(sce.ifm.6h); storage.mode(cd.ifm.6h) <- "integer"
```

# Prior by time point (6h)

Finally, we need to define an expression magnitude prior for the genes.
Its main function, however, is to define a grid of expression magnitude values
on which the numerical calculations will be carried out.

```{r o.prior.6h}
o.prior.6h <- scde.expression.prior(o.ifm.6h, cd.ifm.6h, show.plot = TRUE)
```

# DE {.tabset}

Let us define one contrast that will be used to compare all methods:

```{r targetGroup}
targetGroup <- "6h_D23580_infected"; refGroup <- "6h_Mock_uninfected"
contrastName <- sprintf("%s-%s", targetGroup, refGroup)
```

## All cells; all genes

```{r res.1.allCells.allGenes}
sg.test<-factor(sce.ifm$Group,levels=c(targetGroup, refGroup));summary(sg.test)
batch.test <- sce.ifm$Plate; summary(batch.test)
res.1.allCells.allGenes <- scde.expression.difference(
  o.ifm, cd.ifm, o.prior, sg.test, batch.test, n.cores = 4, verbose = 1
)
```

## 6h cells; 6h genes

```{r res.2.6hCells.6hGenes}
sg.6h<-factor(sce.ifm.6h$Group,levels=c(targetGroup,refGroup));summary(sg.6h)
batch.6h <- droplevels(sce.ifm.6h$Plate); summary(batch.6h)
res.2.6hCells.6hGenes <- scde.expression.difference(
  o.ifm.6h, cd.ifm.6h, o.prior.6h, sg.6h, batch.6h, n.cores = 4, verbose = 1
)
```

# Volcano plots

Let us define:

* a function used to convert the Z-score computed by *scde* to a empirical
P-value ([reference](https://www.biostars.org/p/17227/)):

```{r convert.z.score}
convert.z.score <- function(x, one.sided = NULL) {
  z <- x$Z
  if(is.null(one.sided)) {
    pval = pnorm(-abs(z));
    pval = 2 * pval
  } else if(one.sided=="-") {
    pval = pnorm(z);
  } else {
    pval = pnorm(-z);
  }
    x <- cbind(
      x,
      p.value = pval
  )
  return(x);
}   
```

* a function used to annotate the tables of results returned by *scde*:

```{r addGENENAME}
addGENENAME <- function(x){
  x <- cbind(
    GENENAME=mapIds(EnsDb.Hsapiens.v79, rownames(x), 'GENENAME', 'GENEID'),
    x
  )
  return(x)
}
```

* a function to order results by decreasing absolute
corrected Z-score and subsequently absolute Z-score:

```{r orderResults}
orderResults <- function(x){
  x <- x[with(x, order(abs(cZ), abs(Z), decreasing = TRUE)),]
  return(x)
}
```

* various significance levels:

```{r volcano.sig}
sig.levels <- c(0.05, 0.01)
volcano.sig <- data.frame(
  p.value = sig.levels,
  level = as.character(sig.levels)
)
```

* a function to visualise *scde* differential expression statistics, and
  return a table with those statistics augmented by a an empirical *P* value
  computed from the *Z* score returned by *scde*:

```{r volcano.mle}
volcano.mle <- function(x, sub = NULL){
  xadj <- convert.z.score(addGENENAME(orderResults(x)))
  varName <- deparse(substitute(x))
  gg <- ggplot(xadj, aes(mle, -log10(p.value))) +
    geom_point(aes(colour = (cZ != 0))) +
    geom_hline(aes(yintercept=-log10(p.value),linetype=level),volcano.sig) +
    ggtitle(varName, sub)
  print(gg)
  return(xadj)
}
```

## Raw {.tabset}

### All cells, all genes

```{r res.1.adj_volcano_raw, echo=FALSE}
res.1.adj <- volcano.mle(
  res.1.allCells.allGenes$results, "All cells and all genes used for error models")
```

### 6h cells, 6h genes

```{r res.2.adj_volcano_raw, echo=FALSE}
res.2.adj <- volcano.mle(
  res.2.6hCells.6hGenes$results, "6h cells and 6h genes used for error models")
```

## Batch-adjusted {.tabset}

### All cells, all genes

```{r res.1.adj_volcano_adjusted, echo=FALSE}
res.1.adj <- volcano.mle(
  res.1.allCells.allGenes$batch.adjusted,
  "All cells and all genes used for error models")
```

### 6h cells, 6h genes

```{r res.2.adj_volcano_adjusted, echo=FALSE}
res.2.adj <- volcano.mle(
  res.2.6hCells.6hGenes$batch.adjusted,
  "6h cells and 6h genes used for error models")
```
