---
title: "Differential expression using scde"
bibliography:
  bibtex.bib
---

<!-- As HTML pages are built independently, large objects need to be
reimported from disk, and smaller ones quickly recomputed. -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(scater)
library(scran)
library(scde)
library(ensembldb)
library(RColorBrewer)
library(goseq)
library(dplyr)
library(ComplexHeatmap)
sce.norm <- readRDS("rds/sce.norm.rds")
EnsDb.Hsapiens.v79 <- EnsDb.Hsapiens.v79::EnsDb.Hsapiens.v79
```

<!--
Remove ERCC spike-ins
Split cells by time !!
Retain genes detected at each time !!
Error model: {time cells; time genes}
Prior
DE: {group on time cells with NA; batch on time cells w/out NA}
-->

# Plotting themes

First of all, let us define theme elements used throughout various figures
in the following sections:

```{r exprsViolinTheme}
exprsViolinTheme <- theme(
    panel.grid.major.x = element_blank(),
    legend.position = "bottom", legend.box = "vertical")
```

Let us also define a function that displays normalised (*scran*) expression
data in each group:

```{r normExprsByName}
normExprsById <- function(geneId){
  geneName <- subset(fData(sce.norm), gene_id == geneId, "gene_name", drop = TRUE)
  stopifnot(length(geneName) == 1)
  gdata <- data.frame(
    norm_exprs = norm_exprs(sce.norm)[geneId,],
    pData(sce.norm)[,c("Infection","Status","Time")],
    row.names = sampleNames(sce.norm)
  )
  ggplot(gdata, aes(Infection, norm_exprs)) + 
    geom_violin() + geom_jitter(width = 0.1) +
    facet_grid(Time ~ Status) +
    ggtitle(sprintf("%s - %s", geneId, geneName))
}
```

# Prepare data

## Remove spike-in features

```{r table_iSpike, include=FALSE}
table(isSpike(sce.norm))
```

```{r remove_isSpike}
sce.norm <- sce.norm[!isSpike(sce.norm),]
dim(sce.norm)
```

## Reorder samples by experimental group

Let us first reorder and group samples by their experimental group;
this mimics the behavour of the
[scde](https://github.com/Bioconductor-mirror/scde) `scde.error.models`
function, and avoid later confusion and superfluous work to ensure
the compatibility of sample order between individual analysis objects:

```{r order_group_sce.norm}
sce.norm <- sce.norm[,order(sce.norm$Group)]
```

## Subset data by time point {.tabset}

Considering the demonstrated impact on the `Time` factor on gene expression
profiles, and the fact that differential expression will *not* be assessed
between time points, individual time points will be processed separately for
the remainder of this *scde* analysis.

### 2h

```{r sce.2h}
sce.2h <- sce.norm[,sce.norm$Time == '2h']
dim(sce.2h)
```

### 4h

```{r sce.4h}
sce.4h <- sce.norm[,sce.norm$Time == '4h']
dim(sce.4h)
```

### 6h

```{r sce.6h}
sce.6h <- sce.norm[,sce.norm$Time == '6h']
dim(sce.6h)
```

## Extract and filter count matrices {.tabset}

The analysis starts with a matrix of read counts,
filtered based on gene and cell requirements.

In this case, let us use the raw numbers of reads mapped to each endogenous
feature, recast as `integer` type. In addition, *for each time point*,
let us retain only features detected with at least **10** counts in at least
**5** cells of any of the five experimental groups.

Let us first define a function to apply the detection cutoff:

```{r filterCounts}
filterCounts <- function(m, g, counts = 10, cells = 5, groups = 1){
  apply(m, 1, function(e){
    sum(tapply(e >= counts, g, "sum") >= cells) >= 1
  })
}
```

### 2h

```{r cleanCounts_2h}
sg.2h <- droplevels(sce.2h$Group)
keep.2h <- filterCounts(counts(sce.2h), sg.2h)
cd.2h <- counts(sce.2h)[keep.2h,]
storage.mode(cd.2h) <- 'integer'
dim(cd.2h)
```

### 4h

```{r cleanCounts_4h}
sg.4h <- droplevels(sce.4h$Group)
keep.4h <- filterCounts(counts(sce.4h), sg.4h)
cd.4h <- counts(sce.4h)[keep.4h,]
storage.mode(cd.4h) <- 'integer'
dim(cd.4h)
```

### 6h

```{r cleanCounts_6h}
sg.6h <- droplevels(sce.6h$Group)
keep.6h <- filterCounts(counts(sce.6h), sg.6h)
cd.6h <- counts(sce.6h)[keep.6h,]
storage.mode(cd.6h) <- 'integer'
dim(cd.6h)
```

# Error models {.tabset}

Here, we fit the error models on which all subsequent calculations
will rely. The fitting process relies on a subset of robust genes that are
detected in multiple cross-cell comparisons. Here we supply the
`groups` argument, so that the error models for each
experimental group of cells are fit independently.
If the groups argument is omitted, the models would be fit using a common set.

## 2h

```{r o.ifm.2h, eval=FALSE}
o.ifm.2h <- scde.error.models(
  counts = cd.2h, groups = sg.2h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.2h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.2h, "rds/o.ifm.2h.rds")
```

```{r oIfmTrim.2h, include=FALSE}
write.csv(o.ifm.2h, "SCDE_v2/o.ifm.2h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.2h}
valid.cells <- o.ifm.2h$corr.a > 0; table(valid.cells)
o.ifm.2h <- o.ifm.2h[valid.cells, ]
dim(o.ifm.2h)
```

## 4h

```{r o.ifm.4h, eval=FALSE}
o.ifm.4h <- scde.error.models(
  counts = cd.4h, groups = sg.4h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.4h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.4h, "rds/o.ifm.4h.rds")
```

```{r read_o.ifm.4h, include=FALSE}
o.ifm.4h <- readRDS("rds/o.ifm.4h.rds")
```

```{r oIfmTrim.4h, include=FALSE}
write.csv(o.ifm.4h, "SCDE_v2/o.ifm.4h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.4h}
valid.cells <- o.ifm.4h$corr.a > 0; table(valid.cells)
o.ifm.4h <- o.ifm.4h[valid.cells, ]
dim(o.ifm.4h)
```

## 6h

```{r o.ifm.6h, eval=FALSE}
o.ifm.6h <- scde.error.models(
  counts = cd.6h, groups = sg.6h, n.cores = 4, verbose = 1
)
```

```{r save_o.ifm.6h, echo=FALSE, eval=FALSE}
saveRDS(o.ifm.6h, "rds/o.ifm.6h.rds")
```

```{r read_o.ifm.6h, include=FALSE}
o.ifm.6h <- readRDS("rds/o.ifm.6h.rds")
```

```{r oIfmTrim.6h, include=FALSE}
write.csv(o.ifm.6h, "SCDE_v2/o.ifm.6h.csv")
```

Particularly poor cells may result in abnormal fits, most commonly showing
negative `corr.a`, and should be removed.

```{r validCells.6h}
valid.cells <- o.ifm.6h$corr.a > 0; table(valid.cells)
o.ifm.6h <- o.ifm.6h[valid.cells, ]
dim(o.ifm.6h)
```

# Prior distribution for gene expression magnitudes {.tabset}

Finally, we need to define an expression magnitude prior for the genes.
Its main function, however, is to define a grid of expression magnitude values
on which the numerical calculations will be carried out:

## 2h

```{r oPrior_2h}
o.prior.2h <- scde.expression.prior(
  models = o.ifm.2h, counts = cd.2h, show.plot = TRUE
)
```

Here we used a grid of **400** points, and let the maximum expression magnitude
be determined by the default **0.999** quantile
(default `max.quantile` setting;
use `max.value` parameter to specify the maximum expression magnitude
explicitly --- on log~10~ scale).

```{r stopifnot_order_2h}
stopifnot(all(rownames(o.ifm.2h) == colnames(cd.2h)))
```

## 4h

```{r oPrior_4h}
o.prior.4h <- scde.expression.prior(
  models = o.ifm.4h, counts = cd.4h, show.plot = TRUE
)
```

Here we used a grid of **400** points, and let the maximum expression magnitude
be determined by the default **0.999** quantile
(default `max.quantile` setting;
use `max.value` parameter to specify the maximum expression magnitude
explicitly --- on log~10~ scale).

```{r stopifnot_order_4h}
stopifnot(all(rownames(o.ifm.4h) == colnames(cd.4h)))
```

## 6h

```{r oPrior_6h}
o.prior.6h <- scde.expression.prior(
  models = o.ifm.6h, counts = cd.6h, show.plot = TRUE
)
```

Here we used a grid of **400** points, and let the maximum expression magnitude
be determined by the default **0.999** quantile
(default `max.quantile` setting;
use `max.value` parameter to specify the maximum expression magnitude
explicitly --- on log~10~ scale).

```{r stopifnot_order_6h}
stopifnot(all(rownames(o.ifm.6h) == colnames(cd.6h)))
```

# Differential expression

Let us define a function that will be used to annotate the results returned
by *scde*:

```{r addGENENAME}
addGENENAME <- function(x){
  x <- cbind(
    GENENAME=mapIds(EnsDb.Hsapiens.v79, rownames(x), 'GENENAME', 'GENEID'),
    x
  )
  return(x)
}
```

A separate function to order results by decreasing absolute
corrected Z-score and subsequently absolute Z-score:

```{r orderResults}
orderResults <- function(x){
  x <- x[with(x, order(abs(cZ), abs(Z), decreasing = TRUE)),]
  return(x)
}
```

Let us also initialise lists to store differential expression results:

* default results data frame
* batch-corrected results
* differences explained by batch effects alone

```{r scde_lists}
scde.results <- list()
scde.adjusted <- list()
scde.batch <- list()
```

## Relative to control {#DEcontrol}

For each group of stimulated cells, compute differential expression
statistics relative to the time-matched group of `uninfected` cell.

## 2h {.tabset}

### D23580 infected

```{r scde_2h_D23580_infected_control}
sg.target <- "2h_D23580_infected"; sg.reference <- "2h_Mock_uninfected"
contrastName <- sprintf("%s-vs-%s", sg.target, sg.reference)
sg.d23.inf.2h.un <- factor(sg.2h, levels = c(sg.target, sg.reference))
summary(sg.d23.inf.2h.un)
```

```{r, eval=FALSE}
ediff <- scde.expression.difference(
  o.ifm.2h, cd.2h, o.prior.2h, groups = sg.d23.inf.2h.un,
  batch = sce.2h$Plate, n.cores  =  4, verbose = 1
)
```

```{r, echo=FALSE, eval=FALSE}
saveRDS(ediff, sprintf("rds/scde.%s.rds", contrastName))
```

```{r, include=FALSE}
ediff <- readRDS(sprintf("rds/scde.%s.rds", contrastName))
```

Extract and store results in separate lists:

```{r}
scde.results[[contrastName]] <- ediff$results
scde.adjusted[[contrastName]] <- ediff$batch.adjusted
scde.batch[[contrastName]] <- ediff$batch.effect
```

Count of significantly DE genes (*i.e.*, corrected Z-score different from 0):

```{r}
table(ediff$batch.adjusted$cZ != 0)
```

```{r, include=FALSE}
write.csv(
  addGENENAME(orderResults(ediff$results)),
  sprintf('SCDE_v2/results_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.adjusted)),
  sprintf('SCDE_v2/adjusted_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.effect)),
  sprintf('SCDE_v2/batch_%s-vs-%s.csv', sg.target, sg.reference)
)
```

Display the *scde* maximum likelihood estimate values for expression
and differential expression:

```{r}
scde.test.gene.expression.difference(
  "ENSG00000158050",
  models = o.ifm.2h, counts = cd.2h, prior = o.prior.2h,
  groups = sg.d23.inf.2h.un, batch = sce.2h$Plate
)
```

Display the *scran*-normalised expression data:

```{r}
normExprsById("ENSG00000158050")
```


## 4h {.tabset}

### D23580 infected

```{r scde_4h_D23580_infected_control}
sg.target <- "4h_D23580_infected"; sg.reference <- "4h_Mock_uninfected"
contrastName <- sprintf("%s-vs-%s", sg.target, sg.reference)
sg.d23.inf.4h.un <- factor(sg.4h, levels = c(sg.target, sg.reference))
summary(sg.d23.inf.4h.un)
```

```{r, eval=FALSE}
ediff <- scde.expression.difference(
  o.ifm.4h, cd.4h, o.prior.4h, groups = sg.d23.inf.4h.un,
  batch = sce.4h$Plate, n.cores  =  4, verbose = 1
)
```

```{r, echo=FALSE, eval=FALSE}
saveRDS(ediff, sprintf("rds/scde.%s.rds", contrastName))
```

```{r, include=FALSE}
ediff <- readRDS(sprintf("rds/scde.%s.rds", contrastName))
```

Extract and store results in separate lists:

```{r}
scde.results[[contrastName]] <- ediff$results
scde.adjusted[[contrastName]] <- ediff$batch.adjusted
scde.batch[[contrastName]] <- ediff$batch.effect
```

Count of significantly DE genes (*i.e.*, corrected Z-score different from 0):

```{r}
table(ediff$batch.adjusted$cZ != 0)
```

```{r, include=FALSE}
write.csv(
  addGENENAME(orderResults(ediff$results)),
  sprintf('SCDE_v2/results_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.adjusted)),
  sprintf('SCDE_v2/adjusted_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.effect)),
  sprintf('SCDE_v2/batch_%s-vs-%s.csv', sg.target, sg.reference)
)
```

Display the *scde* maximum likelihood estimate values for expression
and differential expression:

```{r}
scde.test.gene.expression.difference(
  "ENSG00000110218",
  models = o.ifm.4h, counts = cd.4h, prior = o.prior.4h,
  groups = sg.d23.inf.4h.un, batch = sce.4h$Plate
)
```

Display the *scran*-normalised expression data:

```{r}
normExprsById("ENSG00000110218")
```


## 6h {.tabset}

### D23580 exposed

```{r scde_6h_D23580_exposed_control}
sg.target <- "6h_D23580_exposed"; sg.reference <- "6h_Mock_uninfected"
contrastName <- sprintf("%s-vs-%s", sg.target, sg.reference)
sg.d23.exp.6h.un <- factor(sg.6h, levels = c(sg.target, sg.reference))
table(sg.d23.exp.6h.un, useNA = "ifany")
```

```{r, eval=FALSE}
ediff <- scde.expression.difference(
  o.ifm.6h, cd.6h, o.prior.6h, groups = sg.d23.exp.6h.un,
  batch = sce.6h$Plate, n.cores  =  4, verbose = 1
)
```

```{r, echo=FALSE, eval=FALSE}
saveRDS(ediff, sprintf("rds/scde.%s.rds", contrastName))
```

```{r, include=FALSE}
ediff <- readRDS(sprintf("rds/scde.%s.rds", contrastName))
```

Extract and store results in separate lists:

```{r}
scde.results[[contrastName]] <- ediff$results
scde.adjusted[[contrastName]] <- ediff$batch.adjusted
scde.batch[[contrastName]] <- ediff$batch.effect
```

Count of significantly DE genes (*i.e.*, corrected Z-score different from 0):

```{r}
table(ediff$results$cZ != 0)
table(ediff$batch.adjusted$cZ != 0)
```

```{r, include=FALSE}
write.csv(
  addGENENAME(orderResults(ediff$results)),
  sprintf('SCDE_v2/results_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.adjusted)),
  sprintf('SCDE_v2/adjusted_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.effect)),
  sprintf('SCDE_v2/batch_%s-vs-%s.csv', sg.target, sg.reference)
)
```

Display the *scde* maximum likelihood estimate values for expression
and differential expression:

```{r}
scde.test.gene.expression.difference(
  rownames(orderResults(ediff$batch.adjusted))[1],
  models = o.ifm.6h, counts = cd.6h, prior = o.prior.6h,
  groups = sg.d23.exp.6h.un, batch = sce.6h$Plate
)
```

Display the *scran*-normalised expression data:

```{r}
normExprsById(rownames(orderResults(ediff$batch.adjusted))[1])
```

### D23580 infected

```{r scde_6h_D23580_infected_control}
sg.target <- "6h_D23580_infected"; sg.reference <- "6h_Mock_uninfected"
contrastName <- sprintf("%s-vs-%s", sg.target, sg.reference)
sg.d23.inf.6h.un <- factor(sg.6h, levels = c(sg.target, sg.reference))
summary(sg.d23.inf.6h.un)
```

```{r, eval=FALSE}
ediff <- scde.expression.difference(
  o.ifm.6h, cd.6h, o.prior.6h, groups = sg.d23.inf.6h.un,
  batch = sce.6h$Plate, n.cores  =  4, verbose = 1
)
```

```{r, echo=FALSE, eval=FALSE}
saveRDS(ediff, sprintf("rds/scde.%s.rds", contrastName))
```

```{r, include=FALSE}
ediff <- readRDS(sprintf("rds/scde.%s.rds", contrastName))
```

Extract and store results in separate lists:

```{r}
scde.results[[contrastName]] <- ediff$results
scde.adjusted[[contrastName]] <- ediff$batch.adjusted
scde.batch[[contrastName]] <- ediff$batch.effect
```

Count of significantly DE genes (*i.e.*, corrected Z-score different from 0):

```{r}
table(ediff$batch.adjusted$cZ != 0)
```

```{r, include=FALSE}
write.csv(
  addGENENAME(orderResults(ediff$results)),
  sprintf('SCDE_v2/results_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.adjusted)),
  sprintf('SCDE_v2/adjusted_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.effect)),
  sprintf('SCDE_v2/batch_%s-vs-%s.csv', sg.target, sg.reference)
)
```

Display the *scde* maximum likelihood estimate values for expression
and differential expression:

```{r}
scde.test.gene.expression.difference(
  rownames(orderResults(ediff$batch.adjusted))[1],
  models = o.ifm.6h, counts = cd.6h, prior = o.prior.6h,
  groups = sg.d23.exp.6h.un, batch = sce.6h$Plate
)
```

Display the *scran*-normalised expression data:

```{r}
normExprsById(rownames(orderResults(ediff$batch.adjusted))[1])
```

## Direct comparisons

For each time point, compute differential expression
between groups of stimulated cells that differ
either by `Infection` or `Status`:

## 6h {.tabset}

### D23580 exposed - LT2 exposed

```{r scde_6h_D23580_exposed_vs_6h_LT2_exposed}
sg.target <- "6h_D23580_exposed"; sg.reference <- "6h_LT2_exposed"
contrastName <- sprintf("%s-vs-%s", sg.target, sg.reference)
sg.d23.lt2.exp.6h <- factor(sg.6h, levels = c(sg.target, sg.reference))
table(sg.d23.lt2.exp.6h, useNA = "ifany")
```

```{r, eval=FALSE}
ediff <- scde.expression.difference(
  o.ifm.6h, cd.6h, o.prior.6h, groups = sg.d23.lt2.exp.6h, batch = sce.6h$Plate,
  n.cores  =  4, verbose = 1
)
```

```{r, echo=FALSE, eval=FALSE}
saveRDS(ediff, sprintf("rds/scde.%s.rds", contrastName))
```

```{r, include=FALSE}
ediff <- readRDS(sprintf("rds/scde.%s.rds", contrastName))
```

Extract and store results in separate lists:

```{r}
scde.results[[contrastName]] <- ediff$results
scde.adjusted[[contrastName]] <- ediff$batch.adjusted
scde.batch[[contrastName]] <- ediff$batch.effect
```

Count of significantly DE genes (*i.e.*, corrected Z-score different from 0):

```{r}
table(ediff$batch.adjusted$cZ != 0)
```

```{r, include=FALSE}
write.csv(
  addGENENAME(orderResults(ediff$results)),
  sprintf('SCDE_v2/results_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.adjusted)),
  sprintf('SCDE_v2/adjusted_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.effect)),
  sprintf('SCDE_v2/batch_%s-vs-%s.csv', sg.target, sg.reference)
)
```

Display the *scde* maximum likelihood estimate values for expression
and differential expression:

```{r}
scde.test.gene.expression.difference(
  rownames(orderResults(ediff$batch.adjusted))[1],
  models = o.ifm.6h, counts = cd.6h, prior = o.prior.6h,
  groups = sg.d23.exp.6h.un, batch = sce.6h$Plate
)
```

Display the *scran*-normalised expression data:

```{r}
normExprsById(rownames(orderResults(ediff$batch.adjusted))[1])
```

### D23580 infected - LT2 infected

```{r scde_6h_D23580_infected_vs_6h_LT2_infected}
sg.target <- "6h_D23580_infected"; sg.reference <- "6h_LT2_infected"
contrastName <- sprintf("%s-vs-%s", sg.target, sg.reference)
sg.d23.lt2.inf.6h <- factor(sg.6h, levels = c(sg.target, sg.reference))
table(sg.d23.lt2.inf.6h, useNA = "ifany")
```

```{r, eval=FALSE}
ediff <- scde.expression.difference(
  o.ifm.6h, cd.6h, o.prior.6h, groups = sg.d23.lt2.inf.6h, batch = sce.6h$Plate,
  n.cores  =  4, verbose = 1
)
```

```{r, echo=FALSE, eval=FALSE}
saveRDS(ediff, sprintf("rds/scde.%s.rds", contrastName))
```

```{r, include=FALSE}
ediff <- readRDS(sprintf("rds/scde.%s.rds", contrastName))
```

Extract and store results in separate lists:

```{r}
scde.results[[contrastName]] <- ediff$results
scde.adjusted[[contrastName]] <- ediff$batch.adjusted
scde.batch[[contrastName]] <- ediff$batch.effect
```

Count of significantly DE genes (*i.e.*, corrected Z-score different from 0):

```{r}
table(ediff$batch.adjusted$cZ != 0)
```

```{r, include=FALSE}
write.csv(
  addGENENAME(orderResults(ediff$results)),
  sprintf('SCDE_v2/results_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.adjusted)),
  sprintf('SCDE_v2/adjusted_%s-vs-%s.csv', sg.target, sg.reference)
)
write.csv(
  addGENENAME(orderResults(ediff$batch.effect)),
  sprintf('SCDE_v2/batch_%s-vs-%s.csv', sg.target, sg.reference)
)
```

Display the *scde* maximum likelihood estimate values for expression
and differential expression:

```{r}
scde.test.gene.expression.difference(
  rownames(orderResults(ediff$batch.adjusted))[1],
  models = o.ifm.6h, counts = cd.6h, prior = o.prior.6h,
  groups = sg.d23.exp.6h.un, batch = sce.6h$Plate
)
```

Display the *scran*-normalised expression data:

```{r}
normExprsById(rownames(orderResults(ediff$batch.adjusted))[1])
```
